// @trezor/connect-web/libESM/popup/abstract.mjs
import EventEmitter from "events";
import * as ERRORS from "@trezor/connect-common/libESM/constants/errors.mjs";
import { DEVICE_EVENT, POPUP } from "@trezor/connect-common/libESM/events/index.mjs";
import { getOrigin } from "@trezor/connect-common/libESM/utils/urlUtils.mjs";
import { createDeferred } from "@trezor/utils";
const POPUP_CLOSE_INTERVAL = 500;
export class Popup extends EventEmitter {
  channel;
  handshakePromise;
  logger;
  popupSrc;
  manifest;
  version;
  closeInterval;
  locked = false;
  closedEmitted = false;
  pendingFocusOrCreate = Promise.resolve();
  constructor({
    popupSrc,
    manifest,
    version,
    logger
  }) {
    super();
    this.logger = logger;
    this.popupSrc = popupSrc;
    this.manifest = manifest;
    this.version = version;
    this.channel = this.createChannel(getOrigin(popupSrc));
    this.handshakePromise = createDeferred();
    this.handshakePromise.promise.catch(() => {});
    this.channel.on('message', this.handleCoreMessage.bind(this));
  }
  async focusOrCreate() {
    this.pendingFocusOrCreate = this.pendingFocusOrCreate.catch(() => {}).then(() => this._focusOrCreate());
    await this.pendingFocusOrCreate;
  }
  async _focusOrCreate() {
    const isOpen = await this.isOpen();
    if (!isOpen && this.locked) {
      this.reset(false);
    }
    if (this.locked) {
      this.focusPopup();
      return;
    }
    if (isOpen) {
      await this.close();
    }
    this.locked = true;
    this.closedEmitted = false;
    return this.open();
  }
  buildPopupUrl(src) {
    return src;
  }
  focusPopup() {}
  startCloseMonitoring() {
    this.closeInterval = setInterval(async () => {
      if (!(await this.isOpen())) {
        this.handlePopupClosed();
      }
    }, POPUP_CLOSE_INTERVAL);
  }
  async handleCoreMessage(message) {
    this.logger.debug('handleCoreMessage', message);
    if (message.type === POPUP.CORE_LOADED) {
      this.channel.postMessage({
        type: POPUP.HANDSHAKE,
        payload: {
          manifest: this.manifest,
          version: this.version
        }
      });
      this.handshakePromise?.resolve();
    } else if (message.type === POPUP.CLOSED) {
      await this.close();
      this.handlePopupClosed();
    } else if (message.event === DEVICE_EVENT) {
      this.emit(DEVICE_EVENT, message);
    }
  }
  handlePopupClosed() {
    this.emitClosed();
    this.reset();
  }
  reset(focus = true) {
    if (!this.locked) return;
    this.locked = false;
    this.handshakePromise?.reject(ERRORS.TypedError('Method_Interrupted'));
    this.handshakePromise = createDeferred();
    this.handshakePromise.promise.catch(() => {});
    this.channel.abortHandshake('Popup closed');
    this.channel.disconnect();
    if (this.closeInterval) {
      clearInterval(this.closeInterval);
      this.closeInterval = undefined;
    }
    this.onReset(focus);
  }
  handleOpenFailure(reason) {
    this.logger.error('Failed to open popup:', reason);
    const error = new Error(reason);
    this.channel.abortHandshake(reason);
    this.handshakePromise?.reject(error);
    this.handlePopupClosed();
  }
  async close() {
    if (!(await this.isOpen())) return;
    this.logger.debug('closing popup');
    this.closePopup();
    this.channel.clear();
  }
  emitClosed() {
    if (this.closedEmitted) return;
    this.closedEmitted = true;
    this.channel.resolveMessagePromises({
      success: false,
      error: ERRORS.TypedError('Method_Interrupted')
    });
    this.emit(POPUP.CLOSED);
  }
}
