// @trezor/connect-web/libESM/popup/web.mjs
import * as ERRORS from "@trezor/connect-common/libESM/constants/errors.mjs";
import { WindowWindowChannel } from "@trezor/connect-common/libESM/messageChannel/window-window.mjs";
import { getOrigin } from "@trezor/connect-common/libESM/utils/urlUtils.mjs";
import { getWeakRandomId } from "@trezor/utils/libESM/getWeakRandomId.mjs";
import { Popup } from "./abstract.mjs";
import { getIframeInstance } from "./iframe.mjs";
import { BootstrapError } from "../bootstrap/bootstrap-errors.mjs";
export class WebPopup extends Popup {
  popupWindow;
  iframe = getIframeInstance();
  channelId = getWeakRandomId(16);
  createChannel() {
    return new WindowWindowChannel({
      windowHere: window,
      windowPeer: () => this.iframe.get()?.contentWindow || undefined,
      channel: {
        here: '@trezor/connect-web',
        peer: '@trezor/connect-popup'
      },
      logger: this.logger,
      origin: getOrigin(this.popupSrc)
    });
  }
  async open() {
    const url = this.buildPopupUrl(this.popupSrc);
    const query = `connect-popup-req=${this.channelId}`;
    const debug = this.logger.enabled ? '&debug=1' : '';
    const popupUrl = `${url}/bootstrap.html?${query}${debug}`;
    const popupOrigin = getOrigin(this.popupSrc);
    const windowResult = window.open('about:blank', '_blank');
    if (!windowResult) {
      const error = ERRORS.TypedError('Handshake_Error', 'popup-blocked');
      this.handleOpenFailure(error.message);
      return Promise.reject(error);
    }
    try {
      await this.iframe.create(popupUrl);
    } catch (error) {
      windowResult.close();
      this.handleOpenFailure(error.message);
      return Promise.reject(error);
    }
    windowResult.location.href = popupUrl;
    const iframeWindowChannel = new WindowWindowChannel({
      windowHere: window,
      windowPeer: () => this.iframe.get()?.contentWindow || undefined,
      channel: {
        here: '@trezor/connect-web',
        peer: '@trezor/connect-bootstrap-iframe'
      },
      logger: this.logger,
      origin: popupOrigin
    });
    iframeWindowChannel.on('message', message => {
      if (message.type === 'channel-handshake-error') {
        windowResult.postMessage({
          type: 'channel-handshake-error',
          error: message.error
        }, popupOrigin);
        this.handleOpenFailure(message.error);
        iframeWindowChannel.abortHandshake(message.error);
      }
    });
    try {
      await iframeWindowChannel.init();
      iframeWindowChannel.disconnect();
    } catch (error) {
      this.handleOpenFailure(error.message);
      iframeWindowChannel.disconnect();
      const isBootstrapError = Object.values(BootstrapError).includes(error.message);
      if (isBootstrapError) {
        return Promise.reject(ERRORS.TypedError('Handshake_Error', error.message));
      }
      return Promise.reject(error);
    }
    if (!this.channel.isConnected) {
      this.channel.connect();
    }
    return Promise.resolve();
  }
  focusPopup() {
    this.popupWindow?.focus();
  }
  closePopup() {
    this.popupWindow?.close();
    this.popupWindow = undefined;
  }
  isOpen() {
    return Promise.resolve(this.popupWindow !== undefined);
  }
  onReset() {}
}
