// @trezor/connect-web/libESM/popup/webextension.mjs
import { ServiceWorkerWindowExtConnectableChannel } from "@trezor/connect-common/libESM/messageChannel/serviceworker-window-ext-connectable.mjs";
import { createDeferred } from "@trezor/utils";
import { Popup } from "./abstract.mjs";
export class WebExtensionPopup extends Popup {
  popupWindow;
  popupWindowPromise = createDeferred();
  extensionTabId = 0;
  buildPopupUrl(src) {
    if (typeof chrome !== 'undefined' && chrome.runtime?.id) {
      const hashParams = new URLSearchParams();
      hashParams.set('extension-id', chrome.runtime.id);
      return src + '#' + hashParams.toString();
    }
    return src;
  }
  checkIfTabExists(tabId) {
    return new Promise(resolve => {
      chrome.tabs.get(tabId, () => {
        resolve(!chrome.runtime.lastError);
      });
    });
  }
  logChromeError(operation, error) {
    if (error) {
      this.logger.error(`Chrome ${operation} error:`, error);
    }
  }
  createChannel(origin) {
    return new ServiceWorkerWindowExtConnectableChannel({
      channel: {
        here: '@trezor/connect-webextension-externally-connectable',
        peer: '@trezor/suite-web'
      },
      currentId: () => this.popupWindowPromise?.promise.then(tab => tab.id),
      logger: this.logger,
      allowedOrigin: origin
    });
  }
  open() {
    this.popupWindowPromise = createDeferred();
    this.popupWindowPromise.promise.catch(() => {});
    const url = this.buildPopupUrl(this.popupSrc);
    chrome.windows.getCurrent(currentWindow => {
      this.logger.debug('opening popup. currentWindow type:', currentWindow.type);
      if (currentWindow.type !== 'normal') {
        this.openPopupInNewWindow(url);
      } else {
        this.openPopupInNewTab(url);
      }
    });
    if (!this.channel.isConnected) {
      this.channel.connect();
      this.channel.init().catch(error => {
        this.logger.error('Channel handshake failed:', error);
      });
    }
    return Promise.resolve();
  }
  openPopupInNewWindow(url) {
    chrome.windows.create({
      url
    }, newWindow => {
      if (!newWindow) {
        this.popupWindowPromise?.reject(new Error('Failed to create popup window'));
        this.handleOpenFailure('Failed to create popup window');
        return;
      }
      chrome.tabs.query({
        windowId: newWindow.id,
        active: true
      }, tabs => this.onPopupTabResolved(tabs[0]));
    });
  }
  openPopupInNewTab(url) {
    chrome.tabs.query({
      currentWindow: true,
      active: true
    }, tabs => {
      if (!tabs[0]?.id) {
        this.popupWindowPromise?.reject(new Error('No active tab found'));
        this.handleOpenFailure('No active tab found');
        return;
      }
      this.extensionTabId = tabs[0].id;
      chrome.tabs.create({
        url,
        index: tabs[0].index + 1
      }, tab => this.onPopupTabResolved(tab));
    });
  }
  onPopupTabResolved(tab) {
    this.popupWindow = tab;
    this.logger.debug('popup tab resolved:', tab.id);
    this.popupWindowPromise?.resolve(tab);
    this.startCloseMonitoring();
  }
  focusPopup() {
    if (this.popupWindow?.id) {
      chrome.tabs.update(this.popupWindow.id, {
        active: true
      });
    }
  }
  closePopup() {
    const tabId = this.popupWindow?.id;
    if (!tabId) return;
    chrome.tabs.remove(tabId, () => {
      this.logChromeError('close', chrome.runtime.lastError);
    });
    this.popupWindow = undefined;
  }
  isOpen() {
    if (!this.popupWindow?.id) return Promise.resolve(false);
    return this.checkIfTabExists(this.popupWindow.id);
  }
  onReset(focus = true) {
    if (!focus || !this.extensionTabId) return;
    this.logger.debug('Focusing back to extension tab:', this.extensionTabId);
    chrome.tabs.update(this.extensionTabId, {
      active: true
    }, () => {
      this.logChromeError('focus', chrome.runtime.lastError);
    });
    this.extensionTabId = 0;
  }
}
