// @trezor/protobuf/libESM/definitions/messages-evolu.mjs
import { Type } from "@trezor/schema-utils";
export const EvoluDelegatedIdentityKey = Type.Object({
  private_key: Type.String(),
  rotation_index: Type.Optional(Type.Number())
}, {
  $id: 'EvoluDelegatedIdentityKey'
});
export const EvoluGetDelegatedIdentityKey = Type.Object({
  thp_credential: Type.Optional(Type.String()),
  rotation_index: Type.Optional(Type.Number()),
  rotate: Type.Optional(Type.Boolean())
}, {
  $id: 'EvoluGetDelegatedIdentityKey'
});
export const EvoluGetNode = Type.Object({
  proof_of_delegated_identity: Type.String(),
  node_rotation_index: Type.Optional(Type.Number())
}, {
  $id: 'EvoluGetNode'
});
export const EvoluIndexManagement = Type.Object({
  rotation_index: Type.Optional(Type.Number())
}, {
  $id: 'EvoluIndexManagement'
});
export const EvoluIndexManagementResponse = Type.Object({
  rotation_index: Type.Optional(Type.Number())
}, {
  $id: 'EvoluIndexManagementResponse'
});
export const EvoluNode = Type.Object({
  data: Type.String()
}, {
  $id: 'EvoluNode'
});
export const EvoluRegistrationRequest = Type.Object({
  certificate_chain: Type.Array(Type.String()),
  signature: Type.String()
}, {
  $id: 'EvoluRegistrationRequest'
});
export const EvoluSignRegistrationRequest = Type.Object({
  challenge_from_server: Type.String(),
  size_to_acquire: Type.Number(),
  proof_of_delegated_identity: Type.String()
}, {
  $id: 'EvoluSignRegistrationRequest'
});
