// @trezor/protobuf/libESM/definitions/messages-management.mjs
import { Type } from "@trezor/schema-utils";
export var Enum_BackupAvailability;
(function (Enum_BackupAvailability) {
  Enum_BackupAvailability[Enum_BackupAvailability["NotAvailable"] = 0] = "NotAvailable";
  Enum_BackupAvailability[Enum_BackupAvailability["Required"] = 1] = "Required";
  Enum_BackupAvailability[Enum_BackupAvailability["Available"] = 2] = "Available";
})(Enum_BackupAvailability || (Enum_BackupAvailability = {}));
export const EnumEnum_BackupAvailability = Type.Enum(Enum_BackupAvailability);
export const BackupAvailability = Type.KeyOfEnum(Enum_BackupAvailability, {
  $id: 'BackupAvailability'
});
export var BackupMethod;
(function (BackupMethod) {
  BackupMethod[BackupMethod["Display"] = 0] = "Display";
  BackupMethod[BackupMethod["N4W1"] = 1] = "N4W1";
})(BackupMethod || (BackupMethod = {}));
export const EnumBackupMethod = Type.Enum(BackupMethod);
export var Enum_BackupType;
(function (Enum_BackupType) {
  Enum_BackupType[Enum_BackupType["Bip39"] = 0] = "Bip39";
  Enum_BackupType[Enum_BackupType["Slip39_Basic"] = 1] = "Slip39_Basic";
  Enum_BackupType[Enum_BackupType["Slip39_Advanced"] = 2] = "Slip39_Advanced";
  Enum_BackupType[Enum_BackupType["Slip39_Single_Extendable"] = 3] = "Slip39_Single_Extendable";
  Enum_BackupType[Enum_BackupType["Slip39_Basic_Extendable"] = 4] = "Slip39_Basic_Extendable";
  Enum_BackupType[Enum_BackupType["Slip39_Advanced_Extendable"] = 5] = "Slip39_Advanced_Extendable";
})(Enum_BackupType || (Enum_BackupType = {}));
export const EnumEnum_BackupType = Type.Enum(Enum_BackupType);
export const BackupType = Type.KeyOfEnum(Enum_BackupType, {
  $id: 'BackupType'
});
export var BootCommand;
(function (BootCommand) {
  BootCommand[BootCommand["STOP_AND_WAIT"] = 0] = "STOP_AND_WAIT";
  BootCommand[BootCommand["INSTALL_UPGRADE"] = 1] = "INSTALL_UPGRADE";
})(BootCommand || (BootCommand = {}));
export const EnumBootCommand = Type.Enum(BootCommand);
export var Enum_Capability;
(function (Enum_Capability) {
  Enum_Capability[Enum_Capability["Capability_Bitcoin"] = 1] = "Capability_Bitcoin";
  Enum_Capability[Enum_Capability["Capability_Bitcoin_like"] = 2] = "Capability_Bitcoin_like";
  Enum_Capability[Enum_Capability["Capability_Binance"] = 3] = "Capability_Binance";
  Enum_Capability[Enum_Capability["Capability_Cardano"] = 4] = "Capability_Cardano";
  Enum_Capability[Enum_Capability["Capability_Crypto"] = 5] = "Capability_Crypto";
  Enum_Capability[Enum_Capability["Capability_EOS"] = 6] = "Capability_EOS";
  Enum_Capability[Enum_Capability["Capability_Ethereum"] = 7] = "Capability_Ethereum";
  Enum_Capability[Enum_Capability["Capability_Lisk"] = 8] = "Capability_Lisk";
  Enum_Capability[Enum_Capability["Capability_Monero"] = 9] = "Capability_Monero";
  Enum_Capability[Enum_Capability["Capability_NEM"] = 10] = "Capability_NEM";
  Enum_Capability[Enum_Capability["Capability_Ripple"] = 11] = "Capability_Ripple";
  Enum_Capability[Enum_Capability["Capability_Stellar"] = 12] = "Capability_Stellar";
  Enum_Capability[Enum_Capability["Capability_Tezos"] = 13] = "Capability_Tezos";
  Enum_Capability[Enum_Capability["Capability_U2F"] = 14] = "Capability_U2F";
  Enum_Capability[Enum_Capability["Capability_Shamir"] = 15] = "Capability_Shamir";
  Enum_Capability[Enum_Capability["Capability_ShamirGroups"] = 16] = "Capability_ShamirGroups";
  Enum_Capability[Enum_Capability["Capability_PassphraseEntry"] = 17] = "Capability_PassphraseEntry";
  Enum_Capability[Enum_Capability["Capability_Solana"] = 18] = "Capability_Solana";
  Enum_Capability[Enum_Capability["Capability_Translations"] = 19] = "Capability_Translations";
  Enum_Capability[Enum_Capability["Capability_Brightness"] = 20] = "Capability_Brightness";
  Enum_Capability[Enum_Capability["Capability_Haptic"] = 21] = "Capability_Haptic";
  Enum_Capability[Enum_Capability["Capability_BLE"] = 22] = "Capability_BLE";
  Enum_Capability[Enum_Capability["Capability_NFC"] = 23] = "Capability_NFC";
  Enum_Capability[Enum_Capability["Capability_Tron"] = 24] = "Capability_Tron";
  Enum_Capability[Enum_Capability["Capability_N4W1"] = 25] = "Capability_N4W1";
})(Enum_Capability || (Enum_Capability = {}));
export const EnumEnum_Capability = Type.Enum(Enum_Capability);
export const Capability = Type.KeyOfEnum(Enum_Capability, {
  $id: 'Capability'
});
export var DeviceModelInternal;
(function (DeviceModelInternal) {
  DeviceModelInternal["T1B1"] = "T1B1";
  DeviceModelInternal["T2T1"] = "T2T1";
  DeviceModelInternal["T2B1"] = "T2B1";
  DeviceModelInternal["T3B1"] = "T3B1";
  DeviceModelInternal["T3T1"] = "T3T1";
  DeviceModelInternal["T3W1"] = "T3W1";
  DeviceModelInternal["UNKNOWN"] = "UNKNOWN";
})(DeviceModelInternal || (DeviceModelInternal = {}));
export const EnumDeviceModelInternal = Type.Enum(DeviceModelInternal);
export var Enum_DisplayRotation;
(function (Enum_DisplayRotation) {
  Enum_DisplayRotation[Enum_DisplayRotation["North"] = 0] = "North";
  Enum_DisplayRotation[Enum_DisplayRotation["East"] = 90] = "East";
  Enum_DisplayRotation[Enum_DisplayRotation["South"] = 180] = "South";
  Enum_DisplayRotation[Enum_DisplayRotation["West"] = 270] = "West";
})(Enum_DisplayRotation || (Enum_DisplayRotation = {}));
export const EnumEnum_DisplayRotation = Type.Enum(Enum_DisplayRotation);
export const DisplayRotation = Type.KeyOfEnum(Enum_DisplayRotation, {
  $id: 'DisplayRotation'
});
export var Enum_HomescreenFormat;
(function (Enum_HomescreenFormat) {
  Enum_HomescreenFormat[Enum_HomescreenFormat["Toif"] = 1] = "Toif";
  Enum_HomescreenFormat[Enum_HomescreenFormat["Jpeg"] = 2] = "Jpeg";
  Enum_HomescreenFormat[Enum_HomescreenFormat["ToiG"] = 3] = "ToiG";
})(Enum_HomescreenFormat || (Enum_HomescreenFormat = {}));
export const EnumEnum_HomescreenFormat = Type.Enum(Enum_HomescreenFormat);
export const HomescreenFormat = Type.KeyOfEnum(Enum_HomescreenFormat, {
  $id: 'HomescreenFormat'
});
export var RecoveryDeviceInputMethod;
(function (RecoveryDeviceInputMethod) {
  RecoveryDeviceInputMethod[RecoveryDeviceInputMethod["ScrambledWords"] = 0] = "ScrambledWords";
  RecoveryDeviceInputMethod[RecoveryDeviceInputMethod["Matrix"] = 1] = "Matrix";
})(RecoveryDeviceInputMethod || (RecoveryDeviceInputMethod = {}));
export const EnumRecoveryDeviceInputMethod = Type.Enum(RecoveryDeviceInputMethod);
export var Enum_RecoveryStatus;
(function (Enum_RecoveryStatus) {
  Enum_RecoveryStatus[Enum_RecoveryStatus["Nothing"] = 0] = "Nothing";
  Enum_RecoveryStatus[Enum_RecoveryStatus["Recovery"] = 1] = "Recovery";
  Enum_RecoveryStatus[Enum_RecoveryStatus["Backup"] = 2] = "Backup";
})(Enum_RecoveryStatus || (Enum_RecoveryStatus = {}));
export const EnumEnum_RecoveryStatus = Type.Enum(Enum_RecoveryStatus);
export const RecoveryStatus = Type.KeyOfEnum(Enum_RecoveryStatus, {
  $id: 'RecoveryStatus'
});
export var Enum_RecoveryType;
(function (Enum_RecoveryType) {
  Enum_RecoveryType[Enum_RecoveryType["NormalRecovery"] = 0] = "NormalRecovery";
  Enum_RecoveryType[Enum_RecoveryType["DryRun"] = 1] = "DryRun";
  Enum_RecoveryType[Enum_RecoveryType["UnlockRepeatedBackup"] = 2] = "UnlockRepeatedBackup";
})(Enum_RecoveryType || (Enum_RecoveryType = {}));
export const EnumEnum_RecoveryType = Type.Enum(Enum_RecoveryType);
export const RecoveryType = Type.KeyOfEnum(Enum_RecoveryType, {
  $id: 'RecoveryType'
});
export var Enum_SafetyCheckLevel;
(function (Enum_SafetyCheckLevel) {
  Enum_SafetyCheckLevel[Enum_SafetyCheckLevel["Strict"] = 0] = "Strict";
  Enum_SafetyCheckLevel[Enum_SafetyCheckLevel["PromptAlways"] = 1] = "PromptAlways";
  Enum_SafetyCheckLevel[Enum_SafetyCheckLevel["PromptTemporarily"] = 2] = "PromptTemporarily";
})(Enum_SafetyCheckLevel || (Enum_SafetyCheckLevel = {}));
export const EnumEnum_SafetyCheckLevel = Type.Enum(Enum_SafetyCheckLevel);
export const SafetyCheckLevel = Type.KeyOfEnum(Enum_SafetyCheckLevel, {
  $id: 'SafetyCheckLevel'
});
export var SdProtectOperationType;
(function (SdProtectOperationType) {
  SdProtectOperationType[SdProtectOperationType["DISABLE"] = 0] = "DISABLE";
  SdProtectOperationType[SdProtectOperationType["ENABLE"] = 1] = "ENABLE";
  SdProtectOperationType[SdProtectOperationType["REFRESH"] = 2] = "REFRESH";
})(SdProtectOperationType || (SdProtectOperationType = {}));
export const EnumSdProtectOperationType = Type.Enum(SdProtectOperationType);
export var Enum_WordRequestType;
(function (Enum_WordRequestType) {
  Enum_WordRequestType[Enum_WordRequestType["WordRequestType_Plain"] = 0] = "WordRequestType_Plain";
  Enum_WordRequestType[Enum_WordRequestType["WordRequestType_Matrix9"] = 1] = "WordRequestType_Matrix9";
  Enum_WordRequestType[Enum_WordRequestType["WordRequestType_Matrix6"] = 2] = "WordRequestType_Matrix6";
})(Enum_WordRequestType || (Enum_WordRequestType = {}));
export const EnumEnum_WordRequestType = Type.Enum(Enum_WordRequestType);
export const WordRequestType = Type.KeyOfEnum(Enum_WordRequestType, {
  $id: 'WordRequestType'
});
export const ApplyFlags = Type.Object({
  flags: Type.Number()
}, {
  $id: 'ApplyFlags'
});
export const ApplySettings = Type.Object({
  language: Type.Optional(Type.String()),
  label: Type.Optional(Type.String()),
  use_passphrase: Type.Optional(Type.Boolean()),
  homescreen: Type.Optional(Type.String()),
  _passphrase_source: Type.Optional(Type.Number()),
  auto_lock_delay_ms: Type.Optional(Type.Number()),
  display_rotation: Type.Optional(Type.Union([DisplayRotation, EnumEnum_DisplayRotation])),
  passphrase_always_on_device: Type.Optional(Type.Boolean()),
  safety_checks: Type.Optional(SafetyCheckLevel),
  experimental_features: Type.Optional(Type.Boolean()),
  hide_passphrase_from_host: Type.Optional(Type.Boolean()),
  haptic_feedback: Type.Optional(Type.Boolean()),
  homescreen_length: Type.Optional(Type.Number()),
  auto_lock_delay_battery_ms: Type.Optional(Type.Number())
}, {
  $id: 'ApplySettings'
});
export const AuthenticateDevice = Type.Object({
  challenge: Type.String()
}, {
  $id: 'AuthenticateDevice'
});
export const AuthenticityProof = Type.Object({
  optiga_certificates: Type.Array(Type.String()),
  optiga_signature: Type.String(),
  tropic_certificates: Type.Array(Type.String()),
  tropic_signature: Type.Optional(Type.String()),
  mcu_certificates: Type.Array(Type.String()),
  mcu_signature: Type.Optional(Type.String())
}, {
  $id: 'AuthenticityProof'
});
export const Slip39Group = Type.Object({
  member_threshold: Type.Number(),
  member_count: Type.Number()
}, {
  $id: 'Slip39Group'
});
export const BackupDevice = Type.Object({
  group_threshold: Type.Optional(Type.Number()),
  groups: Type.Optional(Type.Array(Slip39Group)),
  backup_method: Type.Optional(EnumBackupMethod)
}, {
  $id: 'BackupDevice'
});
export const Cancel = Type.Object({}, {
  $id: 'Cancel'
});
export const CancelAuthorization = Type.Object({}, {
  $id: 'CancelAuthorization'
});
export const ChangeLanguage = Type.Object({
  data_length: Type.Number(),
  show_display: Type.Optional(Type.Boolean())
}, {
  $id: 'ChangeLanguage'
});
export const ChangePin = Type.Object({
  remove: Type.Optional(Type.Boolean())
}, {
  $id: 'ChangePin'
});
export const ChangeWipeCode = Type.Object({
  remove: Type.Optional(Type.Boolean())
}, {
  $id: 'ChangeWipeCode'
});
export const DataChunkAck = Type.Object({
  data_chunk: Type.String()
}, {
  $id: 'DataChunkAck'
});
export const DataChunkRequest = Type.Object({
  data_length: Type.Number(),
  data_offset: Type.Number()
}, {
  $id: 'DataChunkRequest'
});
export const DoPreauthorized = Type.Object({}, {
  $id: 'DoPreauthorized'
});
export const EndSession = Type.Object({}, {
  $id: 'EndSession'
});
export const Entropy = Type.Object({
  entropy: Type.String()
}, {
  $id: 'Entropy'
});
export const EntropyAck = Type.Object({
  entropy: Type.String()
}, {
  $id: 'EntropyAck'
});
export const EntropyCheckContinue = Type.Object({
  finish: Type.Optional(Type.Boolean())
}, {
  $id: 'EntropyCheckContinue'
});
export const EntropyCheckReady = Type.Object({}, {
  $id: 'EntropyCheckReady'
});
export const EntropyRequest = Type.Object({
  entropy_commitment: Type.Optional(Type.String()),
  prev_entropy: Type.Optional(Type.String())
}, {
  $id: 'EntropyRequest'
});
export const Features = Type.Object({
  vendor: Type.String(),
  major_version: Type.Number(),
  minor_version: Type.Number(),
  patch_version: Type.Number(),
  build_version: Type.Optional(Type.Number()),
  bootloader_mode: Type.Union([Type.Boolean(), Type.Null()]),
  device_id: Type.Union([Type.String(), Type.Null()]),
  pin_protection: Type.Union([Type.Boolean(), Type.Null()]),
  passphrase_protection: Type.Union([Type.Boolean(), Type.Null()]),
  language: Type.Union([Type.String(), Type.Null()]),
  label: Type.Union([Type.String(), Type.Null()]),
  initialized: Type.Union([Type.Boolean(), Type.Null()]),
  revision: Type.Union([Type.String(), Type.Null()]),
  bootloader_hash: Type.Union([Type.String(), Type.Null()]),
  imported: Type.Union([Type.Boolean(), Type.Null()]),
  unlocked: Type.Union([Type.Boolean(), Type.Null()]),
  _passphrase_cached: Type.Optional(Type.Boolean()),
  firmware_present: Type.Union([Type.Boolean(), Type.Null()]),
  backup_availability: Type.Union([BackupAvailability, Type.Null()]),
  flags: Type.Union([Type.Number(), Type.Null()]),
  model: Type.String(),
  fw_major: Type.Union([Type.Number(), Type.Null()]),
  fw_minor: Type.Union([Type.Number(), Type.Null()]),
  fw_patch: Type.Union([Type.Number(), Type.Null()]),
  fw_build: Type.Optional(Type.Number()),
  fw_vendor: Type.Union([Type.String(), Type.Null()]),
  unfinished_backup: Type.Union([Type.Boolean(), Type.Null()]),
  no_backup: Type.Union([Type.Boolean(), Type.Null()]),
  recovery_status: Type.Union([RecoveryStatus, Type.Null()]),
  capabilities: Type.Array(Capability),
  backup_type: Type.Union([BackupType, Type.Null()]),
  sd_card_present: Type.Union([Type.Boolean(), Type.Null()]),
  sd_protection: Type.Union([Type.Boolean(), Type.Null()]),
  wipe_code_protection: Type.Union([Type.Boolean(), Type.Null()]),
  session_id: Type.Union([Type.String(), Type.Null()]),
  passphrase_always_on_device: Type.Union([Type.Boolean(), Type.Null()]),
  safety_checks: Type.Union([SafetyCheckLevel, Type.Null()]),
  auto_lock_delay_ms: Type.Union([Type.Number(), Type.Null()]),
  display_rotation: Type.Union([DisplayRotation, Type.Null()]),
  experimental_features: Type.Union([Type.Boolean(), Type.Null()]),
  busy: Type.Optional(Type.Boolean()),
  homescreen_format: Type.Optional(HomescreenFormat),
  hide_passphrase_from_host: Type.Optional(Type.Boolean()),
  internal_model: EnumDeviceModelInternal,
  unit_color: Type.Optional(Type.Number()),
  unit_btconly: Type.Optional(Type.Boolean()),
  homescreen_width: Type.Optional(Type.Number()),
  homescreen_height: Type.Optional(Type.Number()),
  bootloader_locked: Type.Optional(Type.Boolean()),
  language_version_matches: Type.Optional(Type.Boolean()),
  unit_packaging: Type.Optional(Type.Number()),
  haptic_feedback: Type.Optional(Type.Boolean()),
  recovery_type: Type.Optional(RecoveryType),
  optiga_sec: Type.Optional(Type.Number()),
  soc: Type.Optional(Type.Number()),
  firmware_corrupted: Type.Optional(Type.Boolean()),
  auto_lock_delay_battery_ms: Type.Optional(Type.Number()),
  led: Type.Optional(Type.Boolean()),
  usb_connected: Type.Optional(Type.Boolean()),
  wireless_connected: Type.Optional(Type.Boolean())
}, {
  $id: 'Features'
});
export const FirmwareHash = Type.Object({
  hash: Type.String()
}, {
  $id: 'FirmwareHash'
});
export const GetEntropy = Type.Object({
  size: Type.Number()
}, {
  $id: 'GetEntropy'
});
export const GetFeatures = Type.Object({}, {
  $id: 'GetFeatures'
});
export const GetFirmwareHash = Type.Object({
  challenge: Type.Optional(Type.String())
}, {
  $id: 'GetFirmwareHash'
});
export const GetNextU2FCounter = Type.Object({}, {
  $id: 'GetNextU2FCounter'
});
export const GetNonce = Type.Object({}, {
  $id: 'GetNonce'
});
export const GetSerialNumber = Type.Object({}, {
  $id: 'GetSerialNumber'
});
export const Initialize = Type.Object({
  session_id: Type.Optional(Type.String()),
  _skip_passphrase: Type.Optional(Type.Boolean()),
  derive_cardano: Type.Optional(Type.Boolean())
}, {
  $id: 'Initialize'
});
export const LoadDevice = Type.Object({
  mnemonics: Type.Array(Type.String()),
  pin: Type.Optional(Type.String()),
  passphrase_protection: Type.Optional(Type.Boolean()),
  language: Type.Optional(Type.String()),
  label: Type.Optional(Type.String()),
  skip_checksum: Type.Optional(Type.Boolean()),
  u2f_counter: Type.Optional(Type.Number()),
  needs_backup: Type.Optional(Type.Boolean()),
  no_backup: Type.Optional(Type.Boolean()),
  unfinished_backup: Type.Optional(Type.Boolean())
}, {
  $id: 'LoadDevice'
});
export const LockDevice = Type.Object({}, {
  $id: 'LockDevice'
});
export const NextU2FCounter = Type.Object({
  u2f_counter: Type.Number()
}, {
  $id: 'NextU2FCounter'
});
export const Nonce = Type.Object({
  nonce: Type.String()
}, {
  $id: 'Nonce'
});
export const Ping = Type.Object({
  message: Type.Optional(Type.String()),
  button_protection: Type.Optional(Type.Boolean())
}, {
  $id: 'Ping'
});
export const PreauthorizedRequest = Type.Object({}, {
  $id: 'PreauthorizedRequest'
});
export const RebootToBootloader = Type.Object({
  boot_command: Type.Optional(EnumBootCommand),
  firmware_header: Type.Optional(Type.String())
}, {
  $id: 'RebootToBootloader'
});
export const RecoveryDevice = Type.Object({
  word_count: Type.Optional(Type.Number()),
  passphrase_protection: Type.Optional(Type.Boolean()),
  pin_protection: Type.Optional(Type.Boolean()),
  language: Type.Optional(Type.String()),
  label: Type.Optional(Type.String()),
  enforce_wordlist: Type.Optional(Type.Boolean()),
  input_method: Type.Optional(EnumRecoveryDeviceInputMethod),
  u2f_counter: Type.Optional(Type.Number()),
  type: Type.Optional(RecoveryType),
  backup_method: Type.Optional(EnumBackupMethod)
}, {
  $id: 'RecoveryDevice'
});
export const ResetDevice = Type.Object({
  strength: Type.Optional(Type.Number()),
  passphrase_protection: Type.Optional(Type.Boolean()),
  pin_protection: Type.Optional(Type.Boolean()),
  language: Type.Optional(Type.String()),
  label: Type.Optional(Type.String()),
  u2f_counter: Type.Optional(Type.Number()),
  skip_backup: Type.Optional(Type.Boolean()),
  no_backup: Type.Optional(Type.Boolean()),
  backup_type: Type.Optional(EnumEnum_BackupType),
  entropy_check: Type.Optional(Type.Boolean()),
  backup_method: Type.Optional(EnumBackupMethod)
}, {
  $id: 'ResetDevice'
});
export const SdProtect = Type.Object({
  operation: EnumSdProtectOperationType
}, {
  $id: 'SdProtect'
});
export const SerialNumber = Type.Object({
  serial_number: Type.String()
}, {
  $id: 'SerialNumber'
});
export const SetBrightness = Type.Object({
  value: Type.Optional(Type.Number())
}, {
  $id: 'SetBrightness'
});
export const SetBusy = Type.Object({
  expiry_ms: Type.Optional(Type.Number())
}, {
  $id: 'SetBusy'
});
export const SetU2FCounter = Type.Object({
  u2f_counter: Type.Number()
}, {
  $id: 'SetU2FCounter'
});
export const ShowDeviceTutorial = Type.Object({}, {
  $id: 'ShowDeviceTutorial'
});
export const UnlockBootloader = Type.Object({}, {
  $id: 'UnlockBootloader'
});
export const UnlockedPathRequest = Type.Object({
  mac: Type.String()
}, {
  $id: 'UnlockedPathRequest'
});
export const UnlockPath = Type.Object({
  address_n: Type.Array(Type.Number()),
  mac: Type.Optional(Type.String())
}, {
  $id: 'UnlockPath'
});
export const WipeDevice = Type.Object({}, {
  $id: 'WipeDevice'
});
export const WordAck = Type.Object({
  word: Type.String()
}, {
  $id: 'WordAck'
});
export const WordRequest = Type.Object({
  type: WordRequestType
}, {
  $id: 'WordRequest'
});
