// @trezor/protobuf/libESM/definitions/messages-eos.mjs
import { Type } from "@trezor/schema-utils";
export const EosAsset = Type.Object({
  amount: Type.Uint(),
  symbol: Type.String()
}, {
  $id: 'EosAsset'
});
export const EosActionBuyRam = Type.Object({
  payer: Type.String(),
  receiver: Type.String(),
  quantity: EosAsset
}, {
  $id: 'EosActionBuyRam'
});
export const EosActionBuyRamBytes = Type.Object({
  payer: Type.String(),
  receiver: Type.String(),
  bytes: Type.Number()
}, {
  $id: 'EosActionBuyRamBytes'
});
export const EosPermissionLevel = Type.Object({
  actor: Type.String(),
  permission: Type.String()
}, {
  $id: 'EosPermissionLevel'
});
export const EosActionCommon = Type.Object({
  account: Type.String(),
  name: Type.String(),
  authorization: Type.Array(EosPermissionLevel)
}, {
  $id: 'EosActionCommon'
});
export const EosActionDelegate = Type.Object({
  sender: Type.String(),
  receiver: Type.String(),
  net_quantity: EosAsset,
  cpu_quantity: EosAsset,
  transfer: Type.Boolean()
}, {
  $id: 'EosActionDelegate'
});
export const EosActionDeleteAuth = Type.Object({
  account: Type.String(),
  permission: Type.String()
}, {
  $id: 'EosActionDeleteAuth'
});
export const EosActionLinkAuth = Type.Object({
  account: Type.String(),
  code: Type.String(),
  type: Type.String(),
  requirement: Type.String()
}, {
  $id: 'EosActionLinkAuth'
});
export const EosAuthorizationKey = Type.Object({
  type: Type.Optional(Type.Number()),
  key: Type.String(),
  address_n: Type.Optional(Type.Array(Type.Number())),
  weight: Type.Number()
}, {
  $id: 'EosAuthorizationKey'
});
export const EosAuthorizationAccount = Type.Object({
  account: EosPermissionLevel,
  weight: Type.Number()
}, {
  $id: 'EosAuthorizationAccount'
});
export const EosAuthorizationWait = Type.Object({
  wait_sec: Type.Number(),
  weight: Type.Number()
}, {
  $id: 'EosAuthorizationWait'
});
export const EosAuthorization = Type.Object({
  threshold: Type.Number(),
  keys: Type.Array(EosAuthorizationKey),
  accounts: Type.Array(EosAuthorizationAccount),
  waits: Type.Array(EosAuthorizationWait)
}, {
  $id: 'EosAuthorization'
});
export const EosActionNewAccount = Type.Object({
  creator: Type.String(),
  name: Type.String(),
  owner: EosAuthorization,
  active: EosAuthorization
}, {
  $id: 'EosActionNewAccount'
});
export const EosActionRefund = Type.Object({
  owner: Type.String()
}, {
  $id: 'EosActionRefund'
});
export const EosActionSellRam = Type.Object({
  account: Type.String(),
  bytes: Type.Number()
}, {
  $id: 'EosActionSellRam'
});
export const EosActionTransfer = Type.Object({
  sender: Type.String(),
  receiver: Type.String(),
  quantity: EosAsset,
  memo: Type.String()
}, {
  $id: 'EosActionTransfer'
});
export const EosActionUndelegate = Type.Object({
  sender: Type.String(),
  receiver: Type.String(),
  net_quantity: EosAsset,
  cpu_quantity: EosAsset
}, {
  $id: 'EosActionUndelegate'
});
export const EosActionUnknown = Type.Object({
  data_size: Type.Number(),
  data_chunk: Type.String()
}, {
  $id: 'EosActionUnknown'
});
export const EosActionUnlinkAuth = Type.Object({
  account: Type.String(),
  code: Type.String(),
  type: Type.String()
}, {
  $id: 'EosActionUnlinkAuth'
});
export const EosActionUpdateAuth = Type.Object({
  account: Type.String(),
  permission: Type.String(),
  parent: Type.String(),
  auth: EosAuthorization
}, {
  $id: 'EosActionUpdateAuth'
});
export const EosActionVoteProducer = Type.Object({
  voter: Type.String(),
  proxy: Type.String(),
  producers: Type.Array(Type.String())
}, {
  $id: 'EosActionVoteProducer'
});
export const EosGetPublicKey = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'EosGetPublicKey'
});
export const EosPublicKey = Type.Object({
  wif_public_key: Type.String(),
  raw_public_key: Type.String()
}, {
  $id: 'EosPublicKey'
});
export const EosSignedTx = Type.Object({
  signature: Type.String()
}, {
  $id: 'EosSignedTx'
});
export const EosTxHeader = Type.Object({
  expiration: Type.Number(),
  ref_block_num: Type.Number(),
  ref_block_prefix: Type.Number(),
  max_net_usage_words: Type.Number(),
  max_cpu_usage_ms: Type.Number(),
  delay_sec: Type.Number()
}, {
  $id: 'EosTxHeader'
});
export const EosSignTx = Type.Object({
  address_n: Type.Array(Type.Number()),
  chain_id: Type.String(),
  header: EosTxHeader,
  num_actions: Type.Number(),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'EosSignTx'
});
export const EosTxActionAck = Type.Object({
  common: EosActionCommon,
  transfer: Type.Optional(EosActionTransfer),
  delegate: Type.Optional(EosActionDelegate),
  undelegate: Type.Optional(EosActionUndelegate),
  refund: Type.Optional(EosActionRefund),
  buy_ram: Type.Optional(EosActionBuyRam),
  buy_ram_bytes: Type.Optional(EosActionBuyRamBytes),
  sell_ram: Type.Optional(EosActionSellRam),
  vote_producer: Type.Optional(EosActionVoteProducer),
  update_auth: Type.Optional(EosActionUpdateAuth),
  delete_auth: Type.Optional(EosActionDeleteAuth),
  link_auth: Type.Optional(EosActionLinkAuth),
  unlink_auth: Type.Optional(EosActionUnlinkAuth),
  new_account: Type.Optional(EosActionNewAccount),
  unknown: Type.Optional(EosActionUnknown)
}, {
  $id: 'EosTxActionAck'
});
export const EosTxActionRequest = Type.Object({
  data_size: Type.Optional(Type.Number())
}, {
  $id: 'EosTxActionRequest'
});
