// @trezor/protobuf/libESM/definitions/messages-monero.mjs
import { Type } from "@trezor/schema-utils";
export var MoneroNetworkType;
(function (MoneroNetworkType) {
  MoneroNetworkType[MoneroNetworkType["MAINNET"] = 0] = "MAINNET";
  MoneroNetworkType[MoneroNetworkType["TESTNET"] = 1] = "TESTNET";
  MoneroNetworkType[MoneroNetworkType["STAGENET"] = 2] = "STAGENET";
  MoneroNetworkType[MoneroNetworkType["FAKECHAIN"] = 3] = "FAKECHAIN";
})(MoneroNetworkType || (MoneroNetworkType = {}));
export const EnumMoneroNetworkType = Type.Enum(MoneroNetworkType);
export const DebugMoneroDiagAck = Type.Object({
  ins: Type.Optional(Type.Number()),
  p1: Type.Optional(Type.Number()),
  p2: Type.Optional(Type.Number()),
  pd: Type.Array(Type.Number()),
  data1: Type.Optional(Type.String()),
  data2: Type.Optional(Type.String())
}, {
  $id: 'DebugMoneroDiagAck'
});
export const DebugMoneroDiagRequest = Type.Object({
  ins: Type.Optional(Type.Number()),
  p1: Type.Optional(Type.Number()),
  p2: Type.Optional(Type.Number()),
  pd: Type.Array(Type.Number()),
  data1: Type.Optional(Type.String()),
  data2: Type.Optional(Type.String())
}, {
  $id: 'DebugMoneroDiagRequest'
});
export const MoneroAccountPublicAddress = Type.Object({
  spend_public_key: Type.Optional(Type.String()),
  view_public_key: Type.Optional(Type.String())
}, {
  $id: 'MoneroAccountPublicAddress'
});
export const MoneroAddress = Type.Object({
  address: Type.String()
}, {
  $id: 'MoneroAddress'
});
export const MoneroExportedKeyImage = Type.Object({
  iv: Type.Optional(Type.String()),
  blob: Type.Optional(Type.String())
}, {
  $id: 'MoneroExportedKeyImage'
});
export const MoneroGetAddress = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  network_type: Type.Optional(EnumMoneroNetworkType),
  account: Type.Optional(Type.Number()),
  minor: Type.Optional(Type.Number()),
  payment_id: Type.Optional(Type.String()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'MoneroGetAddress'
});
export const MoneroGetTxKeyAck = Type.Object({
  salt: Type.Optional(Type.String()),
  tx_keys: Type.Optional(Type.String()),
  tx_derivations: Type.Optional(Type.String())
}, {
  $id: 'MoneroGetTxKeyAck'
});
export const MoneroGetTxKeyRequest = Type.Object({
  address_n: Type.Array(Type.Number()),
  network_type: Type.Optional(EnumMoneroNetworkType),
  salt1: Type.String(),
  salt2: Type.String(),
  tx_enc_keys: Type.String(),
  tx_prefix_hash: Type.String(),
  reason: Type.Optional(Type.Number()),
  view_public_key: Type.Optional(Type.String())
}, {
  $id: 'MoneroGetTxKeyRequest'
});
export const MoneroGetWatchKey = Type.Object({
  address_n: Type.Array(Type.Number()),
  network_type: Type.Optional(EnumMoneroNetworkType)
}, {
  $id: 'MoneroGetWatchKey'
});
export const MoneroKeyImageExportInitAck = Type.Object({}, {
  $id: 'MoneroKeyImageExportInitAck'
});
export const MoneroSubAddressIndicesList = Type.Object({
  account: Type.Number(),
  minor_indices: Type.Array(Type.Number())
}, {
  $id: 'MoneroSubAddressIndicesList'
});
export const MoneroKeyImageExportInitRequest = Type.Object({
  num: Type.Number(),
  hash: Type.Uint8Array(),
  address_n: Type.Array(Type.Number()),
  network_type: Type.Optional(EnumMoneroNetworkType),
  subs: Type.Array(MoneroSubAddressIndicesList)
}, {
  $id: 'MoneroKeyImageExportInitRequest'
});
export const MoneroKeyImageSyncFinalAck = Type.Object({
  enc_key: Type.Optional(Type.String())
}, {
  $id: 'MoneroKeyImageSyncFinalAck'
});
export const MoneroKeyImageSyncFinalRequest = Type.Object({}, {
  $id: 'MoneroKeyImageSyncFinalRequest'
});
export const MoneroKeyImageSyncStepAck = Type.Object({
  kis: Type.Array(MoneroExportedKeyImage)
}, {
  $id: 'MoneroKeyImageSyncStepAck'
});
export const MoneroTransferDetails = Type.Object({
  out_key: Type.Uint8Array(),
  tx_pub_key: Type.Uint8Array(),
  additional_tx_pub_keys: Type.Array(Type.Uint8Array()),
  internal_output_index: Type.Number(),
  sub_addr_major: Type.Optional(Type.Number()),
  sub_addr_minor: Type.Optional(Type.Number())
}, {
  $id: 'MoneroTransferDetails'
});
export const MoneroKeyImageSyncStepRequest = Type.Object({
  tdis: Type.Array(MoneroTransferDetails)
}, {
  $id: 'MoneroKeyImageSyncStepRequest'
});
export const MoneroLiveRefreshFinalAck = Type.Object({}, {
  $id: 'MoneroLiveRefreshFinalAck'
});
export const MoneroLiveRefreshFinalRequest = Type.Object({}, {
  $id: 'MoneroLiveRefreshFinalRequest'
});
export const MoneroLiveRefreshStartAck = Type.Object({}, {
  $id: 'MoneroLiveRefreshStartAck'
});
export const MoneroLiveRefreshStartRequest = Type.Object({
  address_n: Type.Array(Type.Number()),
  network_type: Type.Optional(EnumMoneroNetworkType)
}, {
  $id: 'MoneroLiveRefreshStartRequest'
});
export const MoneroLiveRefreshStepAck = Type.Object({
  salt: Type.Optional(Type.String()),
  key_image: Type.Optional(Type.String())
}, {
  $id: 'MoneroLiveRefreshStepAck'
});
export const MoneroLiveRefreshStepRequest = Type.Object({
  out_key: Type.String(),
  recv_deriv: Type.String(),
  real_out_idx: Type.Number(),
  sub_addr_major: Type.Number(),
  sub_addr_minor: Type.Number()
}, {
  $id: 'MoneroLiveRefreshStepRequest'
});
export const MoneroMultisigKLRki = Type.Object({
  K: Type.Optional(Type.String()),
  L: Type.Optional(Type.String()),
  R: Type.Optional(Type.String()),
  ki: Type.Optional(Type.String())
}, {
  $id: 'MoneroMultisigKLRki'
});
export const MoneroRctKeyPublic = Type.Object({
  dest: Type.String(),
  commitment: Type.String()
}, {
  $id: 'MoneroRctKeyPublic'
});
export const MoneroOutputEntry = Type.Object({
  idx: Type.Optional(Type.Number()),
  key: Type.Optional(MoneroRctKeyPublic)
}, {
  $id: 'MoneroOutputEntry'
});
export const MoneroRingCtSig = Type.Object({
  txn_fee: Type.Optional(Type.Number()),
  message: Type.Optional(Type.String()),
  rv_type: Type.Optional(Type.Number())
}, {
  $id: 'MoneroRingCtSig'
});
export const MoneroTransactionRsigData = Type.Object({
  rsig_type: Type.Optional(Type.Number()),
  offload_type: Type.Optional(Type.Number()),
  grouping: Type.Array(Type.Number()),
  mask: Type.Optional(Type.String()),
  rsig: Type.Optional(Type.String()),
  rsig_parts: Type.Array(Type.String()),
  bp_version: Type.Optional(Type.Number())
}, {
  $id: 'MoneroTransactionRsigData'
});
export const MoneroTransactionAllInputsSetAck = Type.Object({
  rsig_data: Type.Optional(MoneroTransactionRsigData)
}, {
  $id: 'MoneroTransactionAllInputsSetAck'
});
export const MoneroTransactionAllInputsSetRequest = Type.Object({}, {
  $id: 'MoneroTransactionAllInputsSetRequest'
});
export const MoneroTransactionAllOutSetAck = Type.Object({
  extra: Type.Optional(Type.String()),
  tx_prefix_hash: Type.String(),
  rv: MoneroRingCtSig,
  full_message_hash: Type.Optional(Type.String())
}, {
  $id: 'MoneroTransactionAllOutSetAck'
});
export const MoneroTransactionAllOutSetRequest = Type.Object({
  rsig_data: Type.Optional(MoneroTransactionRsigData)
}, {
  $id: 'MoneroTransactionAllOutSetRequest'
});
export const MoneroTransactionDestinationEntry = Type.Object({
  amount: Type.Optional(Type.Uint()),
  addr: Type.Optional(MoneroAccountPublicAddress),
  is_subaddress: Type.Optional(Type.Boolean()),
  original: Type.Optional(Type.String()),
  is_integrated: Type.Optional(Type.Boolean())
}, {
  $id: 'MoneroTransactionDestinationEntry'
});
export const MoneroTransactionData = Type.Object({
  version: Type.Optional(Type.Number()),
  payment_id: Type.Optional(Type.String()),
  unlock_time: Type.Optional(Type.Number()),
  outputs: Type.Array(MoneroTransactionDestinationEntry),
  change_dts: Type.Optional(MoneroTransactionDestinationEntry),
  num_inputs: Type.Optional(Type.Number()),
  mixin: Type.Optional(Type.Number()),
  fee: Type.Optional(Type.Uint()),
  account: Type.Optional(Type.Number()),
  minor_indices: Type.Array(Type.Number()),
  rsig_data: Type.Optional(MoneroTransactionRsigData),
  integrated_indices: Type.Array(Type.Number()),
  client_version: Type.Optional(Type.Number()),
  hard_fork: Type.Optional(Type.Number()),
  monero_version: Type.Optional(Type.String()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'MoneroTransactionData'
});
export const MoneroTransactionFinalAck = Type.Object({
  cout_key: Type.Optional(Type.String()),
  salt: Type.Optional(Type.String()),
  rand_mult: Type.Optional(Type.String()),
  tx_enc_keys: Type.Optional(Type.String()),
  opening_key: Type.Optional(Type.String())
}, {
  $id: 'MoneroTransactionFinalAck'
});
export const MoneroTransactionFinalRequest = Type.Object({}, {
  $id: 'MoneroTransactionFinalRequest'
});
export const MoneroTransactionInitAck = Type.Object({
  hmacs: Type.Array(Type.String()),
  rsig_data: Type.Optional(MoneroTransactionRsigData)
}, {
  $id: 'MoneroTransactionInitAck'
});
export const MoneroTransactionInitRequest = Type.Object({
  version: Type.Optional(Type.Number()),
  address_n: Type.Array(Type.Number()),
  network_type: Type.Optional(EnumMoneroNetworkType),
  tsx_data: Type.Optional(MoneroTransactionData)
}, {
  $id: 'MoneroTransactionInitRequest'
});
export const MoneroTransactionInputViniAck = Type.Object({}, {
  $id: 'MoneroTransactionInputViniAck'
});
export const MoneroTransactionSourceEntry = Type.Object({
  outputs: Type.Array(MoneroOutputEntry),
  real_output: Type.Optional(Type.Number()),
  real_out_tx_key: Type.Optional(Type.String()),
  real_out_additional_tx_keys: Type.Array(Type.String()),
  real_output_in_tx_index: Type.Optional(Type.Number()),
  amount: Type.Optional(Type.Uint()),
  rct: Type.Optional(Type.Boolean()),
  mask: Type.Optional(Type.String()),
  multisig_kLRki: Type.Optional(MoneroMultisigKLRki),
  subaddr_minor: Type.Optional(Type.Number())
}, {
  $id: 'MoneroTransactionSourceEntry'
});
export const MoneroTransactionInputViniRequest = Type.Object({
  src_entr: Type.Optional(MoneroTransactionSourceEntry),
  vini: Type.Optional(Type.String()),
  vini_hmac: Type.Optional(Type.String()),
  pseudo_out: Type.Optional(Type.String()),
  pseudo_out_hmac: Type.Optional(Type.String()),
  orig_idx: Type.Optional(Type.Number())
}, {
  $id: 'MoneroTransactionInputViniRequest'
});
export const MoneroTransactionSetInputAck = Type.Object({
  vini: Type.String(),
  vini_hmac: Type.String(),
  pseudo_out: Type.String(),
  pseudo_out_hmac: Type.String(),
  pseudo_out_alpha: Type.String(),
  spend_key: Type.String()
}, {
  $id: 'MoneroTransactionSetInputAck'
});
export const MoneroTransactionSetInputRequest = Type.Object({
  src_entr: Type.Optional(MoneroTransactionSourceEntry)
}, {
  $id: 'MoneroTransactionSetInputRequest'
});
export const MoneroTransactionSetOutputAck = Type.Object({
  tx_out: Type.Optional(Type.String()),
  vouti_hmac: Type.Optional(Type.String()),
  rsig_data: Type.Optional(MoneroTransactionRsigData),
  out_pk: Type.Optional(Type.String()),
  ecdh_info: Type.Optional(Type.String())
}, {
  $id: 'MoneroTransactionSetOutputAck'
});
export const MoneroTransactionSetOutputRequest = Type.Object({
  dst_entr: Type.Optional(MoneroTransactionDestinationEntry),
  dst_entr_hmac: Type.Optional(Type.String()),
  rsig_data: Type.Optional(MoneroTransactionRsigData),
  is_offloaded_bp: Type.Optional(Type.Boolean())
}, {
  $id: 'MoneroTransactionSetOutputRequest'
});
export const MoneroTransactionSignInputAck = Type.Object({
  signature: Type.Optional(Type.String()),
  pseudo_out: Type.Optional(Type.String())
}, {
  $id: 'MoneroTransactionSignInputAck'
});
export const MoneroTransactionSignInputRequest = Type.Object({
  src_entr: Type.Optional(MoneroTransactionSourceEntry),
  vini: Type.Optional(Type.String()),
  vini_hmac: Type.Optional(Type.String()),
  pseudo_out: Type.Optional(Type.String()),
  pseudo_out_hmac: Type.Optional(Type.String()),
  pseudo_out_alpha: Type.Optional(Type.String()),
  spend_key: Type.Optional(Type.String()),
  orig_idx: Type.Optional(Type.Number())
}, {
  $id: 'MoneroTransactionSignInputRequest'
});
export const MoneroWatchKey = Type.Object({
  watch_key: Type.String(),
  address: Type.String()
}, {
  $id: 'MoneroWatchKey'
});
