// @trezor/protobuf/libESM/definitions/messages-crypto.mjs
import { Type } from "@trezor/schema-utils";
import { PaymentRequest } from "./messages-common.mjs";
export const CipheredKeyValue = Type.Object({
  value: Type.String()
}, {
  $id: 'CipheredKeyValue'
});
export const CipherKeyValue = Type.Object({
  address_n: Type.Array(Type.Number()),
  key: Type.String(),
  value: Type.String(),
  encrypt: Type.Optional(Type.Boolean()),
  ask_on_encrypt: Type.Optional(Type.Boolean()),
  ask_on_decrypt: Type.Optional(Type.Boolean()),
  iv: Type.Optional(Type.String())
}, {
  $id: 'CipherKeyValue'
});
export const ECDHSessionKey = Type.Object({
  session_key: Type.String(),
  public_key: Type.Optional(Type.String())
}, {
  $id: 'ECDHSessionKey'
});
export const IdentityType = Type.Object({
  proto: Type.Optional(Type.String()),
  user: Type.Optional(Type.String()),
  host: Type.Optional(Type.String()),
  port: Type.Optional(Type.String()),
  path: Type.Optional(Type.String()),
  index: Type.Optional(Type.Number())
}, {
  $id: 'IdentityType'
});
export const GetECDHSessionKey = Type.Object({
  identity: IdentityType,
  peer_public_key: Type.String(),
  ecdsa_curve_name: Type.Optional(Type.String())
}, {
  $id: 'GetECDHSessionKey'
});
export const PaymentNotification = Type.Object({
  payment_req: Type.Optional(PaymentRequest)
}, {
  $id: 'PaymentNotification'
});
export const SignedIdentity = Type.Object({
  address: Type.String(),
  public_key: Type.String(),
  signature: Type.String()
}, {
  $id: 'SignedIdentity'
});
export const SignIdentity = Type.Object({
  identity: IdentityType,
  challenge_hidden: Type.Optional(Type.String()),
  challenge_visual: Type.Optional(Type.String()),
  ecdsa_curve_name: Type.Optional(Type.String())
}, {
  $id: 'SignIdentity'
});
