// @trezor/protobuf/libESM/definitions/messages-tron.mjs
import { Type } from "@trezor/schema-utils";
export var TronRawContractType;
(function (TronRawContractType) {
  TronRawContractType[TronRawContractType["TransferContract"] = 1] = "TransferContract";
  TronRawContractType[TronRawContractType["VoteWitnessContract"] = 4] = "VoteWitnessContract";
  TronRawContractType[TronRawContractType["TriggerSmartContract"] = 31] = "TriggerSmartContract";
  TronRawContractType[TronRawContractType["FreezeBalanceV2Contract"] = 54] = "FreezeBalanceV2Contract";
  TronRawContractType[TronRawContractType["UnfreezeBalanceV2Contract"] = 55] = "UnfreezeBalanceV2Contract";
  TronRawContractType[TronRawContractType["WithdrawExpireUnfreezeContract"] = 56] = "WithdrawExpireUnfreezeContract";
})(TronRawContractType || (TronRawContractType = {}));
export const EnumTronRawContractType = Type.Enum(TronRawContractType);
export var TronResourceCode;
(function (TronResourceCode) {
  TronResourceCode[TronResourceCode["BANDWIDTH"] = 0] = "BANDWIDTH";
  TronResourceCode[TronResourceCode["ENERGY"] = 1] = "ENERGY";
})(TronResourceCode || (TronResourceCode = {}));
export const EnumTronResourceCode = Type.Enum(TronResourceCode);
export const TronAddress = Type.Object({
  address: Type.String(),
  mac: Type.Optional(Type.String())
}, {
  $id: 'TronAddress'
});
export const TronContractRequest = Type.Object({}, {
  $id: 'TronContractRequest'
});
export const TronFreezeBalanceV2Contract = Type.Object({
  owner_address: Type.String(),
  balance: Type.Number(),
  resource: Type.Optional(EnumTronResourceCode)
}, {
  $id: 'TronFreezeBalanceV2Contract'
});
export const TronGetAddress = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'TronGetAddress'
});
export const TronRawParameter = Type.Object({
  type_url: Type.String(),
  value: Type.String()
}, {
  $id: 'TronRawParameter'
});
export const TronRawContract = Type.Object({
  type: EnumTronRawContractType,
  parameter: TronRawParameter
}, {
  $id: 'TronRawContract'
});
export const TronRawTransaction = Type.Object({
  ref_block_bytes: Type.String(),
  ref_block_hash: Type.String(),
  expiration: Type.Number(),
  data: Type.Optional(Type.String()),
  contract: Type.Array(TronRawContract),
  timestamp: Type.Number(),
  fee_limit: Type.Optional(Type.Number())
}, {
  $id: 'TronRawTransaction'
});
export const TronSignature = Type.Object({
  signature: Type.String()
}, {
  $id: 'TronSignature'
});
export const TronSignTx = Type.Object({
  address_n: Type.Array(Type.Number()),
  ref_block_bytes: Type.String(),
  ref_block_hash: Type.String(),
  expiration: Type.Number(),
  data: Type.Optional(Type.String()),
  timestamp: Type.Number(),
  fee_limit: Type.Optional(Type.Number())
}, {
  $id: 'TronSignTx'
});
export const TronTransferContract = Type.Object({
  owner_address: Type.String(),
  to_address: Type.String(),
  amount: Type.Uint()
}, {
  $id: 'TronTransferContract'
});
export const TronTriggerSmartContract = Type.Object({
  owner_address: Type.String(),
  contract_address: Type.String(),
  data: Type.String()
}, {
  $id: 'TronTriggerSmartContract'
});
export const TronUnfreezeBalanceV2Contract = Type.Object({
  owner_address: Type.String(),
  balance: Type.Number(),
  resource: Type.Optional(EnumTronResourceCode)
}, {
  $id: 'TronUnfreezeBalanceV2Contract'
});
export const TronVote = Type.Object({
  address: Type.String(),
  count: Type.Number()
}, {
  $id: 'TronVote'
});
export const TronVoteWitnessContract = Type.Object({
  owner_address: Type.String(),
  votes: Type.Array(TronVote)
}, {
  $id: 'TronVoteWitnessContract'
});
export const TronWithdrawUnfreeze = Type.Object({
  owner_address: Type.String()
}, {
  $id: 'TronWithdrawUnfreeze'
});
