// @trezor/protobuf/libESM/definitions/messages-bitcoin.mjs
import { CloneType, Type } from "@trezor/schema-utils";
import { HDNodeType } from "./messages-common.mjs";
export var AmountUnit;
(function (AmountUnit) {
  AmountUnit[AmountUnit["BITCOIN"] = 0] = "BITCOIN";
  AmountUnit[AmountUnit["MILLIBITCOIN"] = 1] = "MILLIBITCOIN";
  AmountUnit[AmountUnit["MICROBITCOIN"] = 2] = "MICROBITCOIN";
  AmountUnit[AmountUnit["SATOSHI"] = 3] = "SATOSHI";
})(AmountUnit || (AmountUnit = {}));
export const EnumAmountUnit = Type.Enum(AmountUnit);
export var DecredStakingSpendType;
(function (DecredStakingSpendType) {
  DecredStakingSpendType[DecredStakingSpendType["SSGen"] = 0] = "SSGen";
  DecredStakingSpendType[DecredStakingSpendType["SSRTX"] = 1] = "SSRTX";
})(DecredStakingSpendType || (DecredStakingSpendType = {}));
export const EnumDecredStakingSpendType = Type.Enum(DecredStakingSpendType);
export var Enum_InputScriptType;
(function (Enum_InputScriptType) {
  Enum_InputScriptType[Enum_InputScriptType["SPENDADDRESS"] = 0] = "SPENDADDRESS";
  Enum_InputScriptType[Enum_InputScriptType["SPENDMULTISIG"] = 1] = "SPENDMULTISIG";
  Enum_InputScriptType[Enum_InputScriptType["EXTERNAL"] = 2] = "EXTERNAL";
  Enum_InputScriptType[Enum_InputScriptType["SPENDWITNESS"] = 3] = "SPENDWITNESS";
  Enum_InputScriptType[Enum_InputScriptType["SPENDP2SHWITNESS"] = 4] = "SPENDP2SHWITNESS";
  Enum_InputScriptType[Enum_InputScriptType["SPENDTAPROOT"] = 5] = "SPENDTAPROOT";
})(Enum_InputScriptType || (Enum_InputScriptType = {}));
export const EnumEnum_InputScriptType = Type.Enum(Enum_InputScriptType);
export const InputScriptType = Type.KeyOfEnum(Enum_InputScriptType, {
  $id: 'InputScriptType'
});
export var MultisigPubkeysOrder;
(function (MultisigPubkeysOrder) {
  MultisigPubkeysOrder[MultisigPubkeysOrder["PRESERVED"] = 0] = "PRESERVED";
  MultisigPubkeysOrder[MultisigPubkeysOrder["LEXICOGRAPHIC"] = 1] = "LEXICOGRAPHIC";
})(MultisigPubkeysOrder || (MultisigPubkeysOrder = {}));
export const EnumMultisigPubkeysOrder = Type.Enum(MultisigPubkeysOrder);
export var Enum_OutputScriptType;
(function (Enum_OutputScriptType) {
  Enum_OutputScriptType[Enum_OutputScriptType["PAYTOADDRESS"] = 0] = "PAYTOADDRESS";
  Enum_OutputScriptType[Enum_OutputScriptType["PAYTOSCRIPTHASH"] = 1] = "PAYTOSCRIPTHASH";
  Enum_OutputScriptType[Enum_OutputScriptType["PAYTOMULTISIG"] = 2] = "PAYTOMULTISIG";
  Enum_OutputScriptType[Enum_OutputScriptType["PAYTOOPRETURN"] = 3] = "PAYTOOPRETURN";
  Enum_OutputScriptType[Enum_OutputScriptType["PAYTOWITNESS"] = 4] = "PAYTOWITNESS";
  Enum_OutputScriptType[Enum_OutputScriptType["PAYTOP2SHWITNESS"] = 5] = "PAYTOP2SHWITNESS";
  Enum_OutputScriptType[Enum_OutputScriptType["PAYTOTAPROOT"] = 6] = "PAYTOTAPROOT";
})(Enum_OutputScriptType || (Enum_OutputScriptType = {}));
export const EnumEnum_OutputScriptType = Type.Enum(Enum_OutputScriptType);
export const OutputScriptType = Type.KeyOfEnum(Enum_OutputScriptType, {
  $id: 'OutputScriptType'
});
export var Enum_RequestType;
(function (Enum_RequestType) {
  Enum_RequestType[Enum_RequestType["TXINPUT"] = 0] = "TXINPUT";
  Enum_RequestType[Enum_RequestType["TXOUTPUT"] = 1] = "TXOUTPUT";
  Enum_RequestType[Enum_RequestType["TXMETA"] = 2] = "TXMETA";
  Enum_RequestType[Enum_RequestType["TXFINISHED"] = 3] = "TXFINISHED";
  Enum_RequestType[Enum_RequestType["TXEXTRADATA"] = 4] = "TXEXTRADATA";
  Enum_RequestType[Enum_RequestType["TXORIGINPUT"] = 5] = "TXORIGINPUT";
  Enum_RequestType[Enum_RequestType["TXORIGOUTPUT"] = 6] = "TXORIGOUTPUT";
  Enum_RequestType[Enum_RequestType["TXPAYMENTREQ"] = 7] = "TXPAYMENTREQ";
})(Enum_RequestType || (Enum_RequestType = {}));
export const EnumEnum_RequestType = Type.Enum(Enum_RequestType);
export const RequestType = Type.KeyOfEnum(Enum_RequestType, {
  $id: 'RequestType'
});
export const Address = Type.Object({
  address: Type.String(),
  mac: Type.Optional(Type.String())
}, {
  $id: 'Address'
});
export const AuthorizeCoinJoin = Type.Object({
  coordinator: Type.String(),
  max_rounds: Type.Number(),
  max_coordinator_fee_rate: Type.Number(),
  max_fee_per_kvbyte: Type.Number(),
  address_n: Type.Array(Type.Number()),
  coin_name: Type.Optional(Type.String()),
  script_type: Type.Optional(InputScriptType),
  amount_unit: Type.Optional(EnumAmountUnit)
}, {
  $id: 'AuthorizeCoinJoin'
});
export const CoinJoinRequest = Type.Object({
  fee_rate: Type.Number(),
  no_fee_threshold: Type.Number(),
  min_registrable_amount: Type.Number(),
  mask_public_key: Type.Optional(Type.String()),
  signature: Type.Optional(Type.String())
}, {
  $id: 'CoinJoinRequest'
});
export const HDNodePathType = Type.Object({
  node: Type.Union([HDNodeType, Type.String()]),
  address_n: Type.Array(Type.Number())
}, {
  $id: 'HDNodePathType'
});
export const MultisigRedeemScriptType = Type.Object({
  pubkeys: Type.Array(HDNodePathType),
  signatures: Type.Array(Type.String()),
  m: Type.Number(),
  nodes: Type.Optional(Type.Array(HDNodeType)),
  address_n: Type.Optional(Type.Array(Type.Number())),
  pubkeys_order: Type.Optional(EnumMultisigPubkeysOrder)
}, {
  $id: 'MultisigRedeemScriptType'
});
export const GetAddress = Type.Object({
  address_n: Type.Array(Type.Number()),
  coin_name: Type.Optional(Type.String()),
  show_display: Type.Optional(Type.Boolean()),
  multisig: Type.Optional(MultisigRedeemScriptType),
  script_type: Type.Optional(InputScriptType),
  ignore_xpub_magic: Type.Optional(Type.Boolean()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'GetAddress'
});
export const GetOwnershipId = Type.Object({
  address_n: Type.Array(Type.Number()),
  coin_name: Type.Optional(Type.String()),
  multisig: Type.Optional(MultisigRedeemScriptType),
  script_type: Type.Optional(InputScriptType)
}, {
  $id: 'GetOwnershipId'
});
export const GetOwnershipProof = Type.Object({
  address_n: Type.Array(Type.Number()),
  coin_name: Type.Optional(Type.String()),
  script_type: Type.Optional(InputScriptType),
  multisig: Type.Optional(MultisigRedeemScriptType),
  user_confirmation: Type.Optional(Type.Boolean()),
  ownership_ids: Type.Optional(Type.Array(Type.String())),
  commitment_data: Type.Optional(Type.String())
}, {
  $id: 'GetOwnershipProof'
});
export const GetPublicKey = Type.Object({
  address_n: Type.Array(Type.Number()),
  ecdsa_curve_name: Type.Optional(Type.String()),
  show_display: Type.Optional(Type.Boolean()),
  coin_name: Type.Optional(Type.String()),
  script_type: Type.Optional(InputScriptType),
  ignore_xpub_magic: Type.Optional(Type.Boolean())
}, {
  $id: 'GetPublicKey'
});
export const MessageSignature = Type.Object({
  address: Type.String(),
  signature: Type.String()
}, {
  $id: 'MessageSignature'
});
export const OwnershipId = Type.Object({
  ownership_id: Type.String()
}, {
  $id: 'OwnershipId'
});
export const OwnershipProof = Type.Object({
  ownership_proof: Type.String(),
  signature: Type.String()
}, {
  $id: 'OwnershipProof'
});
export const PrevInput = Type.Object({
  prev_hash: Type.String(),
  prev_index: Type.Number(),
  script_sig: Type.String(),
  sequence: Type.Number(),
  decred_tree: Type.Optional(Type.Number())
}, {
  $id: 'PrevInput'
});
export const PrevOutput = Type.Object({
  amount: Type.Uint(),
  script_pubkey: Type.String(),
  decred_script_version: Type.Optional(Type.Number())
}, {
  $id: 'PrevOutput'
});
export const PrevTx = Type.Object({
  version: Type.Number(),
  lock_time: Type.Number(),
  inputs_count: Type.Number(),
  outputs_count: Type.Number(),
  extra_data_len: Type.Optional(Type.Number()),
  expiry: Type.Optional(Type.Number()),
  version_group_id: Type.Optional(Type.Number()),
  timestamp: Type.Optional(Type.Number()),
  branch_id: Type.Optional(Type.Number())
}, {
  $id: 'PrevTx'
});
export const PublicKey = Type.Object({
  node: HDNodeType,
  xpub: Type.String(),
  root_fingerprint: Type.Optional(Type.Number()),
  descriptor: Type.Optional(Type.String())
}, {
  $id: 'PublicKey'
});
export const SignMessage = Type.Object({
  address_n: Type.Array(Type.Number()),
  message: Type.String(),
  coin_name: Type.Optional(Type.String()),
  script_type: Type.Optional(InputScriptType),
  no_script_type: Type.Optional(Type.Boolean()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'SignMessage'
});
export const SignTx = Type.Object({
  outputs_count: Type.Number(),
  inputs_count: Type.Number(),
  coin_name: Type.Optional(Type.String()),
  version: Type.Optional(Type.Number()),
  lock_time: Type.Optional(Type.Number()),
  expiry: Type.Optional(Type.Number()),
  overwintered: Type.Optional(Type.Boolean()),
  version_group_id: Type.Optional(Type.Number()),
  timestamp: Type.Optional(Type.Number()),
  branch_id: Type.Optional(Type.Number()),
  amount_unit: Type.Optional(EnumAmountUnit),
  decred_staking_ticket: Type.Optional(Type.Boolean()),
  serialize: Type.Optional(Type.Boolean()),
  coinjoin_request: Type.Optional(CoinJoinRequest),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'SignTx'
});
export const InternalInputScriptType = Type.Exclude(InputScriptType, Type.Literal('EXTERNAL'), {
  $id: 'InternalInputScriptType'
});
const CommonTxInputType = Type.Object({
  prev_hash: Type.String(),
  prev_index: Type.Number(),
  amount: Type.Uint(),
  sequence: Type.Optional(Type.Number()),
  multisig: Type.Optional(MultisigRedeemScriptType),
  decred_tree: Type.Optional(Type.Number()),
  orig_hash: Type.Optional(Type.String()),
  orig_index: Type.Optional(Type.Number()),
  decred_staking_spend: Type.Optional(EnumDecredStakingSpendType),
  script_pubkey: Type.Optional(Type.String()),
  coinjoin_flags: Type.Optional(Type.Number()),
  script_sig: Type.Optional(Type.String()),
  witness: Type.Optional(Type.String()),
  ownership_proof: Type.Optional(Type.String()),
  commitment_data: Type.Optional(Type.String())
}, {
  $id: 'CommonTxInputType'
});
export const TxInputType = Type.Union([Type.Intersect([CommonTxInputType, Type.Object({
  address_n: Type.Array(Type.Number()),
  script_type: Type.Optional(InternalInputScriptType)
})]), Type.Intersect([CommonTxInputType, Type.Object({
  address_n: Type.Optional(Type.Undefined()),
  script_type: Type.Literal('EXTERNAL'),
  script_pubkey: Type.String()
})])], {
  $id: 'TxInputType'
});
export const TxInput = CloneType(TxInputType, {
  $id: 'TxInput'
});
export const TxOutputBinType = Type.Object({
  amount: Type.Uint(),
  script_pubkey: Type.String(),
  decred_script_version: Type.Optional(Type.Number())
}, {
  $id: 'TxOutputBinType'
});
export const ChangeOutputScriptType = Type.Exclude(OutputScriptType, Type.Literal('PAYTOOPRETURN'), {
  $id: 'ChangeOutputScriptType'
});
export const TxOutputType = Type.Union([Type.Object({
  address: Type.String(),
  address_n: Type.Optional(Type.Undefined()),
  script_type: Type.Optional(Type.Literal('PAYTOADDRESS')),
  amount: Type.Uint(),
  multisig: Type.Optional(MultisigRedeemScriptType),
  orig_hash: Type.Optional(Type.String()),
  orig_index: Type.Optional(Type.Number()),
  payment_req_index: Type.Optional(Type.Number())
}), Type.Object({
  address: Type.Optional(Type.Undefined()),
  address_n: Type.Array(Type.Number()),
  script_type: Type.Optional(ChangeOutputScriptType),
  amount: Type.Uint(),
  multisig: Type.Optional(MultisigRedeemScriptType),
  orig_hash: Type.Optional(Type.String()),
  orig_index: Type.Optional(Type.Number()),
  payment_req_index: Type.Optional(Type.Number())
}), Type.Object({
  address: Type.Optional(Type.Undefined()),
  address_n: Type.Optional(Type.Undefined()),
  amount: Type.Union([Type.Literal('0'), Type.Literal(0)]),
  op_return_data: Type.String(),
  script_type: Type.Literal('PAYTOOPRETURN'),
  orig_hash: Type.Optional(Type.String()),
  orig_index: Type.Optional(Type.Number()),
  payment_req_index: Type.Optional(Type.Number())
})], {
  $id: 'TxOutputType'
});
export const TxOutput = CloneType(TxOutputType, {
  $id: 'TxOutput'
});
export const TxAckResponse = Type.Union([Type.Object({
  inputs: Type.Array(Type.Union([TxInputType, PrevInput]))
}), Type.Object({
  bin_outputs: Type.Array(TxOutputBinType)
}), Type.Object({
  outputs: Type.Array(TxOutputType)
}), Type.Object({
  extra_data: Type.String()
}), Type.Object({
  version: Type.Optional(Type.Number()),
  lock_time: Type.Optional(Type.Number()),
  inputs_cnt: Type.Number(),
  outputs_cnt: Type.Number(),
  extra_data: Type.Optional(Type.String()),
  extra_data_len: Type.Optional(Type.Number()),
  timestamp: Type.Optional(Type.Number()),
  version_group_id: Type.Optional(Type.Number()),
  expiry: Type.Optional(Type.Number()),
  branch_id: Type.Optional(Type.Number())
})], {
  $id: 'TxAckResponse'
});
export const TxAck = Type.Object({
  tx: TxAckResponse
}, {
  $id: 'TxAck'
});
export const TxAckInputWrapper = Type.Object({
  input: TxInput
}, {
  $id: 'TxAckInputWrapper'
});
export const TxAckInput = Type.Object({
  tx: TxAckInputWrapper
}, {
  $id: 'TxAckInput'
});
export const TxAckOutputWrapper = Type.Object({
  output: TxOutput
}, {
  $id: 'TxAckOutputWrapper'
});
export const TxAckOutput = Type.Object({
  tx: TxAckOutputWrapper
}, {
  $id: 'TxAckOutput'
});
export const TxAckPrevExtraDataWrapper = Type.Object({
  extra_data_chunk: Type.String()
}, {
  $id: 'TxAckPrevExtraDataWrapper'
});
export const TxAckPrevExtraData = Type.Object({
  tx: TxAckPrevExtraDataWrapper
}, {
  $id: 'TxAckPrevExtraData'
});
export const TxAckPrevInputWrapper = Type.Object({
  input: PrevInput
}, {
  $id: 'TxAckPrevInputWrapper'
});
export const TxAckPrevInput = Type.Object({
  tx: TxAckPrevInputWrapper
}, {
  $id: 'TxAckPrevInput'
});
export const TxAckPrevMeta = Type.Object({
  tx: PrevTx
}, {
  $id: 'TxAckPrevMeta'
});
export const TxAckPrevOutputWrapper = Type.Object({
  output: PrevOutput
}, {
  $id: 'TxAckPrevOutputWrapper'
});
export const TxAckPrevOutput = Type.Object({
  tx: TxAckPrevOutputWrapper
}, {
  $id: 'TxAckPrevOutput'
});
export const TxRequestDetailsType = Type.Object({
  request_index: Type.Number(),
  tx_hash: Type.Optional(Type.String()),
  extra_data_len: Type.Optional(Type.Number()),
  extra_data_offset: Type.Optional(Type.Number())
}, {
  $id: 'TxRequestDetailsType'
});
export const TxRequestSerializedType = Type.Object({
  signature_index: Type.Optional(Type.Number()),
  signature: Type.Optional(Type.String()),
  serialized_tx: Type.Optional(Type.String())
}, {
  $id: 'TxRequestSerializedType'
});
export const TxRequest = Type.Object({
  request_type: RequestType,
  details: TxRequestDetailsType,
  serialized: Type.Optional(TxRequestSerializedType)
}, {
  $id: 'TxRequest'
});
export const VerifyMessage = Type.Object({
  address: Type.String(),
  signature: Type.String(),
  message: Type.String(),
  coin_name: Type.Optional(Type.String()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'VerifyMessage'
});
