// @trezor/protobuf/libESM/definitions/messages-debug.mjs
import { Type } from "@trezor/schema-utils";
export var DebugButton;
(function (DebugButton) {
  DebugButton[DebugButton["NO"] = 0] = "NO";
  DebugButton[DebugButton["YES"] = 1] = "YES";
  DebugButton[DebugButton["INFO"] = 2] = "INFO";
})(DebugButton || (DebugButton = {}));
export const EnumDebugButton = Type.Enum(DebugButton);
export var DebugPhysicalButton;
(function (DebugPhysicalButton) {
  DebugPhysicalButton[DebugPhysicalButton["LEFT_BTN"] = 0] = "LEFT_BTN";
  DebugPhysicalButton[DebugPhysicalButton["MIDDLE_BTN"] = 1] = "MIDDLE_BTN";
  DebugPhysicalButton[DebugPhysicalButton["RIGHT_BTN"] = 2] = "RIGHT_BTN";
})(DebugPhysicalButton || (DebugPhysicalButton = {}));
export const EnumDebugPhysicalButton = Type.Enum(DebugPhysicalButton);
export var DebugTouchEventType;
(function (DebugTouchEventType) {
  DebugTouchEventType[DebugTouchEventType["TOUCH_FULL_CLICK"] = 0] = "TOUCH_FULL_CLICK";
  DebugTouchEventType[DebugTouchEventType["TOUCH_START"] = 1] = "TOUCH_START";
  DebugTouchEventType[DebugTouchEventType["TOUCH_END"] = 2] = "TOUCH_END";
})(DebugTouchEventType || (DebugTouchEventType = {}));
export const EnumDebugTouchEventType = Type.Enum(DebugTouchEventType);
export var DebugWaitType;
(function (DebugWaitType) {
  DebugWaitType[DebugWaitType["IMMEDIATE"] = 0] = "IMMEDIATE";
  DebugWaitType[DebugWaitType["NEXT_LAYOUT"] = 1] = "NEXT_LAYOUT";
  DebugWaitType[DebugWaitType["CURRENT_LAYOUT"] = 2] = "CURRENT_LAYOUT";
})(DebugWaitType || (DebugWaitType = {}));
export const EnumDebugWaitType = Type.Enum(DebugWaitType);
export const DebugLinkGcInfoItem = Type.Object({
  name: Type.String(),
  value: Type.Number()
}, {
  $id: 'DebugLinkGcInfoItem'
});
export const DebugLinkGcInfo = Type.Object({
  items: Type.Array(DebugLinkGcInfoItem)
}, {
  $id: 'DebugLinkGcInfo'
});
export const DebugLinkGetGcInfo = Type.Object({}, {
  $id: 'DebugLinkGetGcInfo'
});
export const DebugLinkGetPairingInfo = Type.Object({
  channel_id: Type.Optional(Type.String()),
  handshake_hash: Type.Optional(Type.String()),
  nfc_secret_host: Type.Optional(Type.String())
}, {
  $id: 'DebugLinkGetPairingInfo'
});
export const DebugLinkN4W1Connected = Type.Object({}, {
  $id: 'DebugLinkN4W1Connected'
});
export const DebugLinkN4W1Read = Type.Object({
  key: Type.Optional(Type.String())
}, {
  $id: 'DebugLinkN4W1Read'
});
export const DebugLinkN4W1Response = Type.Object({
  value: Type.Optional(Type.String())
}, {
  $id: 'DebugLinkN4W1Response'
});
export const DebugLinkN4W1Write = Type.Object({
  key: Type.Optional(Type.String()),
  value: Type.Optional(Type.String())
}, {
  $id: 'DebugLinkN4W1Write'
});
export const DebugLinkOptigaSetSecMax = Type.Object({}, {
  $id: 'DebugLinkOptigaSetSecMax'
});
export const DebugLinkPairingInfo = Type.Object({
  channel_id: Type.Optional(Type.String()),
  handshake_hash: Type.Optional(Type.String()),
  code_entry_code: Type.Optional(Type.Number()),
  code_qr_code: Type.Optional(Type.String()),
  nfc_secret_trezor: Type.Optional(Type.String())
}, {
  $id: 'DebugLinkPairingInfo'
});
export const DebugLinkResetDebugEvents = Type.Object({}, {
  $id: 'DebugLinkResetDebugEvents'
});
export const DebugLinkSetBatteryState = Type.Object({
  soc: Type.Optional(Type.Number()),
  usb_connected: Type.Optional(Type.Boolean()),
  wireless_connected: Type.Optional(Type.Boolean()),
  ntc_connected: Type.Optional(Type.Boolean()),
  charging_limited: Type.Optional(Type.Boolean()),
  temp_control_active: Type.Optional(Type.Boolean()),
  battery_connected: Type.Optional(Type.Boolean())
}, {
  $id: 'DebugLinkSetBatteryState'
});
export const DebugLinkSetLogFilter = Type.Object({
  filter: Type.Optional(Type.String())
}, {
  $id: 'DebugLinkSetLogFilter'
});
