// @trezor/protobuf/libESM/definitions/messages-thp.mjs
import { Type } from "@trezor/schema-utils";
export var ThpPairingMethod;
(function (ThpPairingMethod) {
  ThpPairingMethod[ThpPairingMethod["SkipPairing"] = 1] = "SkipPairing";
  ThpPairingMethod[ThpPairingMethod["CodeEntry"] = 2] = "CodeEntry";
  ThpPairingMethod[ThpPairingMethod["QrCode"] = 3] = "QrCode";
  ThpPairingMethod[ThpPairingMethod["NFC"] = 4] = "NFC";
})(ThpPairingMethod || (ThpPairingMethod = {}));
export const EnumThpPairingMethod = Type.Enum(ThpPairingMethod);
export const ThpCodeEntryChallenge = Type.Object({
  challenge: Type.String()
}, {
  $id: 'ThpCodeEntryChallenge'
});
export const ThpCodeEntryCommitment = Type.Object({
  commitment: Type.String()
}, {
  $id: 'ThpCodeEntryCommitment'
});
export const ThpCodeEntryCpaceHostTag = Type.Object({
  cpace_host_public_key: Type.String(),
  tag: Type.String()
}, {
  $id: 'ThpCodeEntryCpaceHostTag'
});
export const ThpCodeEntryCpaceTrezor = Type.Object({
  cpace_trezor_public_key: Type.String()
}, {
  $id: 'ThpCodeEntryCpaceTrezor'
});
export const ThpCodeEntrySecret = Type.Object({
  secret: Type.String()
}, {
  $id: 'ThpCodeEntrySecret'
});
export const ThpCreateNewSession = Type.Intersect([Type.Object({
  derive_cardano: Type.Optional(Type.Boolean())
}), Type.Union([Type.Object({
  passphrase: Type.String(),
  on_device: Type.Optional(Type.Undefined())
}), Type.Object({
  passphrase: Type.Optional(Type.Undefined()),
  on_device: Type.Boolean()
})])], {
  $id: 'ThpCreateNewSession'
});
export const ThpCredentialRequest = Type.Object({
  host_static_public_key: Type.String(),
  autoconnect: Type.Optional(Type.Boolean()),
  credential: Type.Optional(Type.String())
}, {
  $id: 'ThpCredentialRequest'
});
export const ThpCredentialResponse = Type.Object({
  trezor_static_public_key: Type.String(),
  credential: Type.String()
}, {
  $id: 'ThpCredentialResponse'
});
export const ThpDeviceProperties = Type.Object({
  internal_model: Type.String(),
  model_variant: Type.Number(),
  protocol_version_major: Type.Number(),
  protocol_version_minor: Type.Number(),
  pairing_methods: Type.Array(Type.KeyOfEnum(ThpPairingMethod))
}, {
  $id: 'ThpDeviceProperties'
});
export const ThpEndRequest = Type.Object({}, {
  $id: 'ThpEndRequest'
});
export const ThpEndResponse = Type.Object({}, {
  $id: 'ThpEndResponse'
});
export const ThpHandshakeCompletionReqNoisePayload = Type.Object({
  host_pairing_credential: Type.Optional(Type.String())
}, {
  $id: 'ThpHandshakeCompletionReqNoisePayload'
});
export const ThpNfcTagHost = Type.Object({
  tag: Type.String()
}, {
  $id: 'ThpNfcTagHost'
});
export const ThpNfcTagTrezor = Type.Object({
  tag: Type.String()
}, {
  $id: 'ThpNfcTagTrezor'
});
export const ThpPairingPreparationsFinished = Type.Object({}, {
  $id: 'ThpPairingPreparationsFinished'
});
export const ThpPairingRequest = Type.Object({
  host_name: Type.String(),
  app_name: Type.String()
}, {
  $id: 'ThpPairingRequest'
});
export const ThpPairingRequestApproved = Type.Object({}, {
  $id: 'ThpPairingRequestApproved'
});
export const ThpQrCodeSecret = Type.Object({
  secret: Type.String()
}, {
  $id: 'ThpQrCodeSecret'
});
export const ThpQrCodeTag = Type.Object({
  tag: Type.String()
}, {
  $id: 'ThpQrCodeTag'
});
export const ThpSelectMethod = Type.Object({
  selected_pairing_method: EnumThpPairingMethod
}, {
  $id: 'ThpSelectMethod'
});
