// @trezor/protobuf/libESM/definitions/messages-common.mjs
import { Type } from "@trezor/schema-utils";
export var Enum_ButtonRequestType;
(function (Enum_ButtonRequestType) {
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_Other"] = 1] = "ButtonRequest_Other";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_FeeOverThreshold"] = 2] = "ButtonRequest_FeeOverThreshold";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_ConfirmOutput"] = 3] = "ButtonRequest_ConfirmOutput";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_ResetDevice"] = 4] = "ButtonRequest_ResetDevice";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_ConfirmWord"] = 5] = "ButtonRequest_ConfirmWord";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_WipeDevice"] = 6] = "ButtonRequest_WipeDevice";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_ProtectCall"] = 7] = "ButtonRequest_ProtectCall";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_SignTx"] = 8] = "ButtonRequest_SignTx";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_FirmwareCheck"] = 9] = "ButtonRequest_FirmwareCheck";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_Address"] = 10] = "ButtonRequest_Address";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_PublicKey"] = 11] = "ButtonRequest_PublicKey";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_MnemonicWordCount"] = 12] = "ButtonRequest_MnemonicWordCount";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_MnemonicInput"] = 13] = "ButtonRequest_MnemonicInput";
  Enum_ButtonRequestType[Enum_ButtonRequestType["_Deprecated_ButtonRequest_PassphraseType"] = 14] = "_Deprecated_ButtonRequest_PassphraseType";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_UnknownDerivationPath"] = 15] = "ButtonRequest_UnknownDerivationPath";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_RecoveryHomepage"] = 16] = "ButtonRequest_RecoveryHomepage";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_Success"] = 17] = "ButtonRequest_Success";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_Warning"] = 18] = "ButtonRequest_Warning";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_PassphraseEntry"] = 19] = "ButtonRequest_PassphraseEntry";
  Enum_ButtonRequestType[Enum_ButtonRequestType["ButtonRequest_PinEntry"] = 20] = "ButtonRequest_PinEntry";
})(Enum_ButtonRequestType || (Enum_ButtonRequestType = {}));
export const EnumEnum_ButtonRequestType = Type.Enum(Enum_ButtonRequestType);
export const ButtonRequestType = Type.KeyOfEnum(Enum_ButtonRequestType, {
  $id: 'ButtonRequestType'
});
export var Enum_FailureType;
(function (Enum_FailureType) {
  Enum_FailureType[Enum_FailureType["Failure_UnexpectedMessage"] = 1] = "Failure_UnexpectedMessage";
  Enum_FailureType[Enum_FailureType["Failure_ButtonExpected"] = 2] = "Failure_ButtonExpected";
  Enum_FailureType[Enum_FailureType["Failure_DataError"] = 3] = "Failure_DataError";
  Enum_FailureType[Enum_FailureType["Failure_ActionCancelled"] = 4] = "Failure_ActionCancelled";
  Enum_FailureType[Enum_FailureType["Failure_PinExpected"] = 5] = "Failure_PinExpected";
  Enum_FailureType[Enum_FailureType["Failure_PinCancelled"] = 6] = "Failure_PinCancelled";
  Enum_FailureType[Enum_FailureType["Failure_PinInvalid"] = 7] = "Failure_PinInvalid";
  Enum_FailureType[Enum_FailureType["Failure_InvalidSignature"] = 8] = "Failure_InvalidSignature";
  Enum_FailureType[Enum_FailureType["Failure_ProcessError"] = 9] = "Failure_ProcessError";
  Enum_FailureType[Enum_FailureType["Failure_NotEnoughFunds"] = 10] = "Failure_NotEnoughFunds";
  Enum_FailureType[Enum_FailureType["Failure_NotInitialized"] = 11] = "Failure_NotInitialized";
  Enum_FailureType[Enum_FailureType["Failure_PinMismatch"] = 12] = "Failure_PinMismatch";
  Enum_FailureType[Enum_FailureType["Failure_WipeCodeMismatch"] = 13] = "Failure_WipeCodeMismatch";
  Enum_FailureType[Enum_FailureType["Failure_InvalidSession"] = 14] = "Failure_InvalidSession";
  Enum_FailureType[Enum_FailureType["Failure_Busy"] = 15] = "Failure_Busy";
  Enum_FailureType[Enum_FailureType["Failure_ThpUnallocatedSession"] = 16] = "Failure_ThpUnallocatedSession";
  Enum_FailureType[Enum_FailureType["Failure_InvalidProtocol"] = 17] = "Failure_InvalidProtocol";
  Enum_FailureType[Enum_FailureType["Failure_InProgress"] = 19] = "Failure_InProgress";
  Enum_FailureType[Enum_FailureType["Failure_FirmwareError"] = 99] = "Failure_FirmwareError";
})(Enum_FailureType || (Enum_FailureType = {}));
export const EnumEnum_FailureType = Type.Enum(Enum_FailureType);
export const FailureType = Type.KeyOfEnum(Enum_FailureType, {
  $id: 'FailureType'
});
export var Enum_PinMatrixRequestType;
(function (Enum_PinMatrixRequestType) {
  Enum_PinMatrixRequestType[Enum_PinMatrixRequestType["PinMatrixRequestType_Current"] = 1] = "PinMatrixRequestType_Current";
  Enum_PinMatrixRequestType[Enum_PinMatrixRequestType["PinMatrixRequestType_NewFirst"] = 2] = "PinMatrixRequestType_NewFirst";
  Enum_PinMatrixRequestType[Enum_PinMatrixRequestType["PinMatrixRequestType_NewSecond"] = 3] = "PinMatrixRequestType_NewSecond";
  Enum_PinMatrixRequestType[Enum_PinMatrixRequestType["PinMatrixRequestType_WipeCodeFirst"] = 4] = "PinMatrixRequestType_WipeCodeFirst";
  Enum_PinMatrixRequestType[Enum_PinMatrixRequestType["PinMatrixRequestType_WipeCodeSecond"] = 5] = "PinMatrixRequestType_WipeCodeSecond";
})(Enum_PinMatrixRequestType || (Enum_PinMatrixRequestType = {}));
export const EnumEnum_PinMatrixRequestType = Type.Enum(Enum_PinMatrixRequestType);
export const PinMatrixRequestType = Type.KeyOfEnum(Enum_PinMatrixRequestType, {
  $id: 'PinMatrixRequestType'
});
export const ButtonAck = Type.Object({}, {
  $id: 'ButtonAck'
});
export const ButtonRequest = Type.Object({
  code: Type.Optional(ButtonRequestType),
  pages: Type.Optional(Type.Number()),
  name: Type.Optional(Type.String())
}, {
  $id: 'ButtonRequest'
});
export const CoinPurchaseMemo = Type.Object({
  coin_type: Type.Number(),
  amount: Type.String(),
  address: Type.String(),
  address_n: Type.Array(Type.Number()),
  mac: Type.String()
}, {
  $id: 'CoinPurchaseMemo'
});
export const Deprecated_PassphraseStateAck = Type.Object({}, {
  $id: 'Deprecated_PassphraseStateAck'
});
export const Deprecated_PassphraseStateRequest = Type.Object({
  state: Type.Optional(Type.String())
}, {
  $id: 'Deprecated_PassphraseStateRequest'
});
export const Failure = Type.Object({
  code: Type.Optional(FailureType),
  message: Type.Optional(Type.String())
}, {
  $id: 'Failure'
});
export const HDNodeType = Type.Object({
  depth: Type.Number(),
  fingerprint: Type.Number(),
  child_num: Type.Number(),
  chain_code: Type.String(),
  private_key: Type.Optional(Type.String()),
  public_key: Type.String()
}, {
  $id: 'HDNodeType'
});
export const PassphraseAck = Type.Object({
  passphrase: Type.Optional(Type.String()),
  _state: Type.Optional(Type.String()),
  on_device: Type.Optional(Type.Boolean())
}, {
  $id: 'PassphraseAck'
});
export const PassphraseRequest = Type.Object({
  _on_device: Type.Optional(Type.Boolean())
}, {
  $id: 'PassphraseRequest'
});
export const TextMemo = Type.Object({
  text: Type.String()
}, {
  $id: 'TextMemo'
});
export const RefundMemo = Type.Object({
  address: Type.String(),
  address_n: Type.Array(Type.Number()),
  mac: Type.String()
}, {
  $id: 'RefundMemo'
});
export const TextDetailsMemo = Type.Object({
  title: Type.String(),
  text: Type.String()
}, {
  $id: 'TextDetailsMemo'
});
export const PaymentRequestMemo = Type.Object({
  text_memo: Type.Optional(TextMemo),
  refund_memo: Type.Optional(RefundMemo),
  coin_purchase_memo: Type.Optional(CoinPurchaseMemo),
  text_details_memo: Type.Optional(TextDetailsMemo)
}, {
  $id: 'PaymentRequestMemo'
});
export const PaymentRequest = Type.Object({
  nonce: Type.Optional(Type.String()),
  recipient_name: Type.String(),
  memos: Type.Optional(Type.Array(PaymentRequestMemo)),
  amount: Type.Optional(Type.String()),
  signature: Type.String()
}, {
  $id: 'PaymentRequest'
});
export const PinMatrixAck = Type.Object({
  pin: Type.String()
}, {
  $id: 'PinMatrixAck'
});
export const PinMatrixRequest = Type.Object({
  type: Type.Optional(PinMatrixRequestType)
}, {
  $id: 'PinMatrixRequest'
});
export const Success = Type.Object({
  message: Type.String()
}, {
  $id: 'Success'
});
