// @trezor/protobuf/libESM/definitions/messages-stellar.mjs
import { Type } from "@trezor/schema-utils";
import { PaymentRequest } from "./messages-common.mjs";
export var StellarAssetType;
(function (StellarAssetType) {
  StellarAssetType[StellarAssetType["NATIVE"] = 0] = "NATIVE";
  StellarAssetType[StellarAssetType["ALPHANUM4"] = 1] = "ALPHANUM4";
  StellarAssetType[StellarAssetType["ALPHANUM12"] = 2] = "ALPHANUM12";
})(StellarAssetType || (StellarAssetType = {}));
export const EnumStellarAssetType = Type.Enum(StellarAssetType);
export var StellarMemoType;
(function (StellarMemoType) {
  StellarMemoType[StellarMemoType["NONE"] = 0] = "NONE";
  StellarMemoType[StellarMemoType["TEXT"] = 1] = "TEXT";
  StellarMemoType[StellarMemoType["ID"] = 2] = "ID";
  StellarMemoType[StellarMemoType["HASH"] = 3] = "HASH";
  StellarMemoType[StellarMemoType["RETURN"] = 4] = "RETURN";
})(StellarMemoType || (StellarMemoType = {}));
export const EnumStellarMemoType = Type.Enum(StellarMemoType);
export var StellarSignerType;
(function (StellarSignerType) {
  StellarSignerType[StellarSignerType["ACCOUNT"] = 0] = "ACCOUNT";
  StellarSignerType[StellarSignerType["PRE_AUTH"] = 1] = "PRE_AUTH";
  StellarSignerType[StellarSignerType["HASH"] = 2] = "HASH";
})(StellarSignerType || (StellarSignerType = {}));
export const EnumStellarSignerType = Type.Enum(StellarSignerType);
export const StellarAccountMergeOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  destination_account: Type.String()
}, {
  $id: 'StellarAccountMergeOp'
});
export const StellarAddress = Type.Object({
  address: Type.String(),
  mac: Type.Optional(Type.String())
}, {
  $id: 'StellarAddress'
});
export const StellarAllowTrustOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  trusted_account: Type.String(),
  asset_type: EnumStellarAssetType,
  asset_code: Type.Optional(Type.String()),
  is_authorized: Type.Boolean()
}, {
  $id: 'StellarAllowTrustOp'
});
export const StellarAsset = Type.Object({
  type: Type.Union([Type.Literal(0), Type.Literal(1), Type.Literal(2), Type.Literal('NATIVE'), Type.Literal('ALPHANUM4'), Type.Literal('ALPHANUM12')]),
  code: Type.Optional(Type.String()),
  issuer: Type.Optional(Type.String())
}, {
  $id: 'StellarAsset'
});
export const StellarBumpSequenceOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  bump_to: Type.Uint()
}, {
  $id: 'StellarBumpSequenceOp'
});
export const StellarChangeTrustOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  asset: StellarAsset,
  limit: Type.Uint()
}, {
  $id: 'StellarChangeTrustOp'
});
export const StellarClaimClaimableBalanceOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  balance_id: Type.String()
}, {
  $id: 'StellarClaimClaimableBalanceOp'
});
export const StellarCreateAccountOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  new_account: Type.String(),
  starting_balance: Type.Uint()
}, {
  $id: 'StellarCreateAccountOp'
});
export const StellarCreatePassiveSellOfferOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  selling_asset: StellarAsset,
  buying_asset: StellarAsset,
  amount: Type.Uint(),
  price_n: Type.Number(),
  price_d: Type.Number()
}, {
  $id: 'StellarCreatePassiveSellOfferOp'
});
export const StellarGetAddress = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'StellarGetAddress'
});
export const StellarManageBuyOfferOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  selling_asset: StellarAsset,
  buying_asset: StellarAsset,
  amount: Type.Uint(),
  price_n: Type.Number(),
  price_d: Type.Number(),
  offer_id: Type.Uint()
}, {
  $id: 'StellarManageBuyOfferOp'
});
export const StellarManageDataOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  key: Type.String(),
  value: Type.Optional(Type.Union([Type.Buffer(), Type.String()]))
}, {
  $id: 'StellarManageDataOp'
});
export const StellarManageSellOfferOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  selling_asset: StellarAsset,
  buying_asset: StellarAsset,
  amount: Type.Uint(),
  price_n: Type.Number(),
  price_d: Type.Number(),
  offer_id: Type.Uint()
}, {
  $id: 'StellarManageSellOfferOp'
});
export const StellarPathPaymentStrictReceiveOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  send_asset: StellarAsset,
  send_max: Type.Uint(),
  destination_account: Type.String(),
  destination_asset: StellarAsset,
  destination_amount: Type.Uint(),
  paths: Type.Optional(Type.Array(StellarAsset))
}, {
  $id: 'StellarPathPaymentStrictReceiveOp'
});
export const StellarPathPaymentStrictSendOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  send_asset: StellarAsset,
  send_amount: Type.Uint(),
  destination_account: Type.String(),
  destination_asset: StellarAsset,
  destination_min: Type.Uint(),
  paths: Type.Optional(Type.Array(StellarAsset))
}, {
  $id: 'StellarPathPaymentStrictSendOp'
});
export const StellarPaymentOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  destination_account: Type.String(),
  asset: StellarAsset,
  amount: Type.Uint()
}, {
  $id: 'StellarPaymentOp'
});
export const StellarSetOptionsOp = Type.Object({
  source_account: Type.Optional(Type.String()),
  inflation_destination_account: Type.Optional(Type.String()),
  clear_flags: Type.Optional(Type.Number()),
  set_flags: Type.Optional(Type.Number()),
  master_weight: Type.Optional(Type.Uint()),
  low_threshold: Type.Optional(Type.Uint()),
  medium_threshold: Type.Optional(Type.Uint()),
  high_threshold: Type.Optional(Type.Uint()),
  home_domain: Type.Optional(Type.String()),
  signer_type: Type.Optional(EnumStellarSignerType),
  signer_key: Type.Optional(Type.Union([Type.Buffer(), Type.String()])),
  signer_weight: Type.Optional(Type.Number())
}, {
  $id: 'StellarSetOptionsOp'
});
export const StellarSignedTx = Type.Object({
  public_key: Type.String(),
  signature: Type.String()
}, {
  $id: 'StellarSignedTx'
});
export const StellarSignTx = Type.Object({
  address_n: Type.Array(Type.Number()),
  network_passphrase: Type.String(),
  source_account: Type.String(),
  fee: Type.Uint(),
  sequence_number: Type.Uint(),
  timebounds_start: Type.Number(),
  timebounds_end: Type.Number(),
  memo_type: EnumStellarMemoType,
  memo_text: Type.Optional(Type.String()),
  memo_id: Type.Optional(Type.Uint()),
  memo_hash: Type.Optional(Type.Union([Type.Buffer(), Type.String()])),
  num_operations: Type.Number(),
  payment_req: Type.Optional(PaymentRequest)
}, {
  $id: 'StellarSignTx'
});
export const StellarTxOpRequest = Type.Object({}, {
  $id: 'StellarTxOpRequest'
});
