// @trezor/protobuf/libESM/definitions/messages-solana.mjs
import { Type } from "@trezor/schema-utils";
import { PaymentRequest } from "./messages-common.mjs";
export const SolanaAddress = Type.Object({
  address: Type.String(),
  mac: Type.Optional(Type.String())
}, {
  $id: 'SolanaAddress'
});
export const SolanaGetAddress = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'SolanaGetAddress'
});
export const SolanaGetPublicKey = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean())
}, {
  $id: 'SolanaGetPublicKey'
});
export const SolanaPublicKey = Type.Object({
  public_key: Type.String()
}, {
  $id: 'SolanaPublicKey'
});
export const SolanaTxTokenAccountInfo = Type.Object({
  base_address: Type.String(),
  token_program: Type.String(),
  token_mint: Type.String(),
  token_account: Type.String()
}, {
  $id: 'SolanaTxTokenAccountInfo'
});
export const SolanaTxAdditionalInfo = Type.Object({
  token_accounts_infos: Type.Array(SolanaTxTokenAccountInfo),
  encoded_token: Type.Optional(Type.ArrayBuffer())
}, {
  $id: 'SolanaTxAdditionalInfo'
});
export const SolanaSignTx = Type.Object({
  address_n: Type.Array(Type.Number()),
  serialized_tx: Type.String(),
  additional_info: Type.Optional(SolanaTxAdditionalInfo),
  payment_req: Type.Optional(PaymentRequest)
}, {
  $id: 'SolanaSignTx'
});
export const SolanaTxSignature = Type.Object({
  signature: Type.String()
}, {
  $id: 'SolanaTxSignature'
});
