// @trezor/protobuf/libESM/definitions/messages-ethereum-eip712.mjs
import { Type } from "@trezor/schema-utils";
import { EthereumDefinitions } from "./messages-ethereum.mjs";
export var EthereumDataType;
(function (EthereumDataType) {
  EthereumDataType[EthereumDataType["UINT"] = 1] = "UINT";
  EthereumDataType[EthereumDataType["INT"] = 2] = "INT";
  EthereumDataType[EthereumDataType["BYTES"] = 3] = "BYTES";
  EthereumDataType[EthereumDataType["STRING"] = 4] = "STRING";
  EthereumDataType[EthereumDataType["BOOL"] = 5] = "BOOL";
  EthereumDataType[EthereumDataType["ADDRESS"] = 6] = "ADDRESS";
  EthereumDataType[EthereumDataType["ARRAY"] = 7] = "ARRAY";
  EthereumDataType[EthereumDataType["STRUCT"] = 8] = "STRUCT";
})(EthereumDataType || (EthereumDataType = {}));
export const EnumEthereumDataType = Type.Enum(EthereumDataType);
export const EthereumFieldType = Type.Recursive(This => Type.Object({
  data_type: EnumEthereumDataType,
  size: Type.Optional(Type.Number()),
  entry_type: Type.Optional(This),
  struct_name: Type.Optional(Type.String())
}), {
  $id: 'EthereumFieldType'
});
export const EthereumSignTypedData = Type.Object({
  address_n: Type.Array(Type.Number()),
  primary_type: Type.String(),
  metamask_v4_compat: Type.Optional(Type.Boolean()),
  definitions: Type.Optional(EthereumDefinitions),
  show_message_hash: Type.Optional(Type.String())
}, {
  $id: 'EthereumSignTypedData'
});
export const EthereumStructMember = Type.Object({
  type: EthereumFieldType,
  name: Type.String()
}, {
  $id: 'EthereumStructMember'
});
export const EthereumTypedDataStructAck = Type.Object({
  members: Type.Array(EthereumStructMember)
}, {
  $id: 'EthereumTypedDataStructAck'
});
export const EthereumTypedDataStructRequest = Type.Object({
  name: Type.String()
}, {
  $id: 'EthereumTypedDataStructRequest'
});
export const EthereumTypedDataValueAck = Type.Object({
  value: Type.String()
}, {
  $id: 'EthereumTypedDataValueAck'
});
export const EthereumTypedDataValueRequest = Type.Object({
  member_path: Type.Array(Type.Number())
}, {
  $id: 'EthereumTypedDataValueRequest'
});
