// @trezor/protobuf/libESM/definitions/messages-ripple.mjs
import { Type } from "@trezor/schema-utils";
import { PaymentRequest } from "./messages-common.mjs";
export const RippleAddress = Type.Object({
  address: Type.String(),
  mac: Type.Optional(Type.String())
}, {
  $id: 'RippleAddress'
});
export const RippleGetAddress = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'RippleGetAddress'
});
export const RipplePayment = Type.Object({
  amount: Type.Uint(),
  destination: Type.String(),
  destination_tag: Type.Optional(Type.Number())
}, {
  $id: 'RipplePayment'
});
export const RippleSignedTx = Type.Object({
  signature: Type.String(),
  serialized_tx: Type.String()
}, {
  $id: 'RippleSignedTx'
});
export const RippleSignTx = Type.Object({
  address_n: Type.Array(Type.Number()),
  fee: Type.Uint(),
  flags: Type.Optional(Type.Number()),
  sequence: Type.Number(),
  last_ledger_sequence: Type.Optional(Type.Number()),
  payment: RipplePayment,
  chunkify: Type.Optional(Type.Boolean()),
  payment_req: Type.Optional(PaymentRequest)
}, {
  $id: 'RippleSignTx'
});
