// @trezor/protobuf/libESM/definitions/messages-tezos.mjs
import { Type } from "@trezor/schema-utils";
export var TezosBallotType;
(function (TezosBallotType) {
  TezosBallotType[TezosBallotType["Yay"] = 0] = "Yay";
  TezosBallotType[TezosBallotType["Nay"] = 1] = "Nay";
  TezosBallotType[TezosBallotType["Pass"] = 2] = "Pass";
})(TezosBallotType || (TezosBallotType = {}));
export const EnumTezosBallotType = Type.Enum(TezosBallotType);
export var TezosContractType;
(function (TezosContractType) {
  TezosContractType[TezosContractType["Implicit"] = 0] = "Implicit";
  TezosContractType[TezosContractType["Originated"] = 1] = "Originated";
})(TezosContractType || (TezosContractType = {}));
export const EnumTezosContractType = Type.Enum(TezosContractType);
export const TezosAddress = Type.Object({
  address: Type.String(),
  mac: Type.Optional(Type.String())
}, {
  $id: 'TezosAddress'
});
export const TezosBallotOp = Type.Object({
  source: Type.String(),
  period: Type.Number(),
  proposal: Type.String(),
  ballot: EnumTezosBallotType
}, {
  $id: 'TezosBallotOp'
});
export const TezosContractID = Type.Object({
  tag: Type.Number(),
  hash: Type.Uint8Array()
}, {
  $id: 'TezosContractID'
});
export const TezosDelegationOp = Type.Object({
  source: Type.Uint8Array(),
  fee: Type.Uint(),
  counter: Type.Number(),
  gas_limit: Type.Number(),
  storage_limit: Type.Number(),
  delegate: Type.Uint8Array()
}, {
  $id: 'TezosDelegationOp'
});
export const TezosGetAddress = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'TezosGetAddress'
});
export const TezosGetPublicKey = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'TezosGetPublicKey'
});
export const TezosManagerTransfer = Type.Object({
  destination: TezosContractID,
  amount: Type.Uint()
}, {
  $id: 'TezosManagerTransfer'
});
export const TezosOriginationOp = Type.Object({
  source: Type.Uint8Array(),
  fee: Type.Uint(),
  counter: Type.Number(),
  gas_limit: Type.Number(),
  storage_limit: Type.Number(),
  manager_pubkey: Type.Optional(Type.String()),
  balance: Type.Number(),
  spendable: Type.Optional(Type.Boolean()),
  delegatable: Type.Optional(Type.Boolean()),
  delegate: Type.Optional(Type.Uint8Array()),
  script: Type.Union([Type.String(), Type.Array(Type.Number())])
}, {
  $id: 'TezosOriginationOp'
});
export const TezosParametersManager = Type.Object({
  set_delegate: Type.Optional(Type.Uint8Array()),
  cancel_delegate: Type.Optional(Type.Boolean()),
  transfer: Type.Optional(TezosManagerTransfer)
}, {
  $id: 'TezosParametersManager'
});
export const TezosProposalOp = Type.Object({
  source: Type.String(),
  period: Type.Number(),
  proposals: Type.Array(Type.String())
}, {
  $id: 'TezosProposalOp'
});
export const TezosPublicKey = Type.Object({
  public_key: Type.String()
}, {
  $id: 'TezosPublicKey'
});
export const TezosRevealOp = Type.Object({
  source: Type.Uint8Array(),
  fee: Type.Uint(),
  counter: Type.Number(),
  gas_limit: Type.Number(),
  storage_limit: Type.Number(),
  public_key: Type.Uint8Array()
}, {
  $id: 'TezosRevealOp'
});
export const TezosSignedTx = Type.Object({
  signature: Type.String(),
  sig_op_contents: Type.String(),
  operation_hash: Type.String()
}, {
  $id: 'TezosSignedTx'
});
export const TezosTransactionOp = Type.Object({
  source: Type.Uint8Array(),
  fee: Type.Uint(),
  counter: Type.Number(),
  gas_limit: Type.Number(),
  storage_limit: Type.Number(),
  amount: Type.Uint(),
  destination: TezosContractID,
  parameters: Type.Optional(Type.Array(Type.Number())),
  parameters_manager: Type.Optional(TezosParametersManager)
}, {
  $id: 'TezosTransactionOp'
});
export const TezosSignTx = Type.Object({
  address_n: Type.Array(Type.Number()),
  branch: Type.Uint8Array(),
  reveal: Type.Optional(TezosRevealOp),
  transaction: Type.Optional(TezosTransactionOp),
  origination: Type.Optional(TezosOriginationOp),
  delegation: Type.Optional(TezosDelegationOp),
  proposal: Type.Optional(TezosProposalOp),
  ballot: Type.Optional(TezosBallotOp),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'TezosSignTx'
});
