// @trezor/protobuf/libESM/definitions/messages-ethereum.mjs
import { Type } from "@trezor/schema-utils";
import { HDNodeType, PaymentRequest } from "./messages-common.mjs";
export const EthereumAccessList = Type.Object({
  address: Type.String(),
  storage_keys: Type.Array(Type.String())
}, {
  $id: 'EthereumAccessList'
});
export const EthereumAddress = Type.Object({
  _old_address: Type.Optional(Type.String()),
  address: Type.String(),
  mac: Type.Optional(Type.String())
}, {
  $id: 'EthereumAddress'
});
export const EthereumDefinitions = Type.Object({
  encoded_network: Type.Optional(Type.ArrayBuffer()),
  encoded_token: Type.Optional(Type.ArrayBuffer())
}, {
  $id: 'EthereumDefinitions'
});
export const EthereumGetAddress = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  encoded_network: Type.Optional(Type.ArrayBuffer()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'EthereumGetAddress'
});
export const EthereumGetPublicKey = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean())
}, {
  $id: 'EthereumGetPublicKey'
});
export const EthereumMessageSignature = Type.Object({
  signature: Type.String(),
  address: Type.String()
}, {
  $id: 'EthereumMessageSignature'
});
export const EthereumPublicKey = Type.Object({
  node: HDNodeType,
  xpub: Type.String()
}, {
  $id: 'EthereumPublicKey'
});
export const EthereumSignMessage = Type.Object({
  address_n: Type.Array(Type.Number()),
  message: Type.String(),
  encoded_network: Type.Optional(Type.ArrayBuffer()),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'EthereumSignMessage'
});
export const EthereumSignTx = Type.Object({
  address_n: Type.Array(Type.Number()),
  nonce: Type.Optional(Type.String()),
  gas_price: Type.String(),
  gas_limit: Type.String(),
  to: Type.Optional(Type.String()),
  value: Type.Optional(Type.String()),
  data_initial_chunk: Type.Optional(Type.String()),
  data_length: Type.Optional(Type.Number()),
  chain_id: Type.Number(),
  tx_type: Type.Optional(Type.Number()),
  definitions: Type.Optional(EthereumDefinitions),
  chunkify: Type.Optional(Type.Boolean()),
  payment_req: Type.Optional(PaymentRequest)
}, {
  $id: 'EthereumSignTx'
});
export const EthereumSignTxEIP1559 = Type.Object({
  address_n: Type.Array(Type.Number()),
  nonce: Type.String(),
  max_gas_fee: Type.String(),
  max_priority_fee: Type.String(),
  gas_limit: Type.String(),
  to: Type.Optional(Type.String()),
  value: Type.String(),
  data_initial_chunk: Type.Optional(Type.String()),
  data_length: Type.Number(),
  chain_id: Type.Number(),
  access_list: Type.Array(EthereumAccessList),
  definitions: Type.Optional(EthereumDefinitions),
  chunkify: Type.Optional(Type.Boolean()),
  payment_req: Type.Optional(PaymentRequest)
}, {
  $id: 'EthereumSignTxEIP1559'
});
export const EthereumSignTypedHash = Type.Object({
  address_n: Type.Array(Type.Number()),
  domain_separator_hash: Type.String(),
  message_hash: Type.Optional(Type.String()),
  encoded_network: Type.Optional(Type.ArrayBuffer())
}, {
  $id: 'EthereumSignTypedHash'
});
export const EthereumTxAck = Type.Object({
  data_chunk: Type.String()
}, {
  $id: 'EthereumTxAck'
});
export const EthereumTxRequest = Type.Object({
  data_length: Type.Optional(Type.Number()),
  signature_v: Type.Optional(Type.Number()),
  signature_r: Type.Optional(Type.String()),
  signature_s: Type.Optional(Type.String())
}, {
  $id: 'EthereumTxRequest'
});
export const EthereumTypedDataSignature = Type.Object({
  signature: Type.String(),
  address: Type.String()
}, {
  $id: 'EthereumTypedDataSignature'
});
export const EthereumVerifyMessage = Type.Object({
  signature: Type.String(),
  message: Type.String(),
  address: Type.String(),
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'EthereumVerifyMessage'
});
