// @trezor/protobuf/libESM/definitions/messages-cardano.mjs
import { Type } from "@trezor/schema-utils";
import { HDNodeType, PaymentRequest } from "./messages-common.mjs";
export var CardanoAddressType;
(function (CardanoAddressType) {
  CardanoAddressType[CardanoAddressType["BASE"] = 0] = "BASE";
  CardanoAddressType[CardanoAddressType["BASE_SCRIPT_KEY"] = 1] = "BASE_SCRIPT_KEY";
  CardanoAddressType[CardanoAddressType["BASE_KEY_SCRIPT"] = 2] = "BASE_KEY_SCRIPT";
  CardanoAddressType[CardanoAddressType["BASE_SCRIPT_SCRIPT"] = 3] = "BASE_SCRIPT_SCRIPT";
  CardanoAddressType[CardanoAddressType["POINTER"] = 4] = "POINTER";
  CardanoAddressType[CardanoAddressType["POINTER_SCRIPT"] = 5] = "POINTER_SCRIPT";
  CardanoAddressType[CardanoAddressType["ENTERPRISE"] = 6] = "ENTERPRISE";
  CardanoAddressType[CardanoAddressType["ENTERPRISE_SCRIPT"] = 7] = "ENTERPRISE_SCRIPT";
  CardanoAddressType[CardanoAddressType["BYRON"] = 8] = "BYRON";
  CardanoAddressType[CardanoAddressType["REWARD"] = 14] = "REWARD";
  CardanoAddressType[CardanoAddressType["REWARD_SCRIPT"] = 15] = "REWARD_SCRIPT";
})(CardanoAddressType || (CardanoAddressType = {}));
export const EnumCardanoAddressType = Type.Enum(CardanoAddressType);
export var CardanoCertificateType;
(function (CardanoCertificateType) {
  CardanoCertificateType[CardanoCertificateType["STAKE_REGISTRATION"] = 0] = "STAKE_REGISTRATION";
  CardanoCertificateType[CardanoCertificateType["STAKE_DEREGISTRATION"] = 1] = "STAKE_DEREGISTRATION";
  CardanoCertificateType[CardanoCertificateType["STAKE_DELEGATION"] = 2] = "STAKE_DELEGATION";
  CardanoCertificateType[CardanoCertificateType["STAKE_POOL_REGISTRATION"] = 3] = "STAKE_POOL_REGISTRATION";
  CardanoCertificateType[CardanoCertificateType["STAKE_REGISTRATION_CONWAY"] = 7] = "STAKE_REGISTRATION_CONWAY";
  CardanoCertificateType[CardanoCertificateType["STAKE_DEREGISTRATION_CONWAY"] = 8] = "STAKE_DEREGISTRATION_CONWAY";
  CardanoCertificateType[CardanoCertificateType["VOTE_DELEGATION"] = 9] = "VOTE_DELEGATION";
})(CardanoCertificateType || (CardanoCertificateType = {}));
export const EnumCardanoCertificateType = Type.Enum(CardanoCertificateType);
export var CardanoCVoteRegistrationFormat;
(function (CardanoCVoteRegistrationFormat) {
  CardanoCVoteRegistrationFormat[CardanoCVoteRegistrationFormat["CIP15"] = 0] = "CIP15";
  CardanoCVoteRegistrationFormat[CardanoCVoteRegistrationFormat["CIP36"] = 1] = "CIP36";
})(CardanoCVoteRegistrationFormat || (CardanoCVoteRegistrationFormat = {}));
export const EnumCardanoCVoteRegistrationFormat = Type.Enum(CardanoCVoteRegistrationFormat);
export var CardanoDerivationType;
(function (CardanoDerivationType) {
  CardanoDerivationType[CardanoDerivationType["LEDGER"] = 0] = "LEDGER";
  CardanoDerivationType[CardanoDerivationType["ICARUS"] = 1] = "ICARUS";
  CardanoDerivationType[CardanoDerivationType["ICARUS_TREZOR"] = 2] = "ICARUS_TREZOR";
})(CardanoDerivationType || (CardanoDerivationType = {}));
export const EnumCardanoDerivationType = Type.Enum(CardanoDerivationType);
export var CardanoDRepType;
(function (CardanoDRepType) {
  CardanoDRepType[CardanoDRepType["KEY_HASH"] = 0] = "KEY_HASH";
  CardanoDRepType[CardanoDRepType["SCRIPT_HASH"] = 1] = "SCRIPT_HASH";
  CardanoDRepType[CardanoDRepType["ABSTAIN"] = 2] = "ABSTAIN";
  CardanoDRepType[CardanoDRepType["NO_CONFIDENCE"] = 3] = "NO_CONFIDENCE";
})(CardanoDRepType || (CardanoDRepType = {}));
export const EnumCardanoDRepType = Type.Enum(CardanoDRepType);
export var CardanoNativeScriptHashDisplayFormat;
(function (CardanoNativeScriptHashDisplayFormat) {
  CardanoNativeScriptHashDisplayFormat[CardanoNativeScriptHashDisplayFormat["HIDE"] = 0] = "HIDE";
  CardanoNativeScriptHashDisplayFormat[CardanoNativeScriptHashDisplayFormat["BECH32"] = 1] = "BECH32";
  CardanoNativeScriptHashDisplayFormat[CardanoNativeScriptHashDisplayFormat["POLICY_ID"] = 2] = "POLICY_ID";
})(CardanoNativeScriptHashDisplayFormat || (CardanoNativeScriptHashDisplayFormat = {}));
export const EnumCardanoNativeScriptHashDisplayFormat = Type.Enum(CardanoNativeScriptHashDisplayFormat);
export var CardanoNativeScriptType;
(function (CardanoNativeScriptType) {
  CardanoNativeScriptType[CardanoNativeScriptType["PUB_KEY"] = 0] = "PUB_KEY";
  CardanoNativeScriptType[CardanoNativeScriptType["ALL"] = 1] = "ALL";
  CardanoNativeScriptType[CardanoNativeScriptType["ANY"] = 2] = "ANY";
  CardanoNativeScriptType[CardanoNativeScriptType["N_OF_K"] = 3] = "N_OF_K";
  CardanoNativeScriptType[CardanoNativeScriptType["INVALID_BEFORE"] = 4] = "INVALID_BEFORE";
  CardanoNativeScriptType[CardanoNativeScriptType["INVALID_HEREAFTER"] = 5] = "INVALID_HEREAFTER";
})(CardanoNativeScriptType || (CardanoNativeScriptType = {}));
export const EnumCardanoNativeScriptType = Type.Enum(CardanoNativeScriptType);
export var CardanoPoolRelayType;
(function (CardanoPoolRelayType) {
  CardanoPoolRelayType[CardanoPoolRelayType["SINGLE_HOST_IP"] = 0] = "SINGLE_HOST_IP";
  CardanoPoolRelayType[CardanoPoolRelayType["SINGLE_HOST_NAME"] = 1] = "SINGLE_HOST_NAME";
  CardanoPoolRelayType[CardanoPoolRelayType["MULTIPLE_HOST_NAME"] = 2] = "MULTIPLE_HOST_NAME";
})(CardanoPoolRelayType || (CardanoPoolRelayType = {}));
export const EnumCardanoPoolRelayType = Type.Enum(CardanoPoolRelayType);
export var CardanoTxAuxiliaryDataSupplementType;
(function (CardanoTxAuxiliaryDataSupplementType) {
  CardanoTxAuxiliaryDataSupplementType[CardanoTxAuxiliaryDataSupplementType["NONE"] = 0] = "NONE";
  CardanoTxAuxiliaryDataSupplementType[CardanoTxAuxiliaryDataSupplementType["CVOTE_REGISTRATION_SIGNATURE"] = 1] = "CVOTE_REGISTRATION_SIGNATURE";
})(CardanoTxAuxiliaryDataSupplementType || (CardanoTxAuxiliaryDataSupplementType = {}));
export const EnumCardanoTxAuxiliaryDataSupplementType = Type.Enum(CardanoTxAuxiliaryDataSupplementType);
export var CardanoTxOutputSerializationFormat;
(function (CardanoTxOutputSerializationFormat) {
  CardanoTxOutputSerializationFormat[CardanoTxOutputSerializationFormat["ARRAY_LEGACY"] = 0] = "ARRAY_LEGACY";
  CardanoTxOutputSerializationFormat[CardanoTxOutputSerializationFormat["MAP_BABBAGE"] = 1] = "MAP_BABBAGE";
})(CardanoTxOutputSerializationFormat || (CardanoTxOutputSerializationFormat = {}));
export const EnumCardanoTxOutputSerializationFormat = Type.Enum(CardanoTxOutputSerializationFormat);
export var CardanoTxSigningMode;
(function (CardanoTxSigningMode) {
  CardanoTxSigningMode[CardanoTxSigningMode["ORDINARY_TRANSACTION"] = 0] = "ORDINARY_TRANSACTION";
  CardanoTxSigningMode[CardanoTxSigningMode["POOL_REGISTRATION_AS_OWNER"] = 1] = "POOL_REGISTRATION_AS_OWNER";
  CardanoTxSigningMode[CardanoTxSigningMode["MULTISIG_TRANSACTION"] = 2] = "MULTISIG_TRANSACTION";
  CardanoTxSigningMode[CardanoTxSigningMode["PLUTUS_TRANSACTION"] = 3] = "PLUTUS_TRANSACTION";
})(CardanoTxSigningMode || (CardanoTxSigningMode = {}));
export const EnumCardanoTxSigningMode = Type.Enum(CardanoTxSigningMode);
export var CardanoTxWitnessType;
(function (CardanoTxWitnessType) {
  CardanoTxWitnessType[CardanoTxWitnessType["BYRON_WITNESS"] = 0] = "BYRON_WITNESS";
  CardanoTxWitnessType[CardanoTxWitnessType["SHELLEY_WITNESS"] = 1] = "SHELLEY_WITNESS";
})(CardanoTxWitnessType || (CardanoTxWitnessType = {}));
export const EnumCardanoTxWitnessType = Type.Enum(CardanoTxWitnessType);
export const CardanoAddress = Type.Object({
  address: Type.String(),
  mac: Type.Optional(Type.String())
}, {
  $id: 'CardanoAddress'
});
export const CardanoBlockchainPointerType = Type.Object({
  block_index: Type.Number(),
  tx_index: Type.Number(),
  certificate_index: Type.Number()
}, {
  $id: 'CardanoBlockchainPointerType'
});
export const CardanoAddressParametersType = Type.Object({
  address_type: EnumCardanoAddressType,
  address_n: Type.Array(Type.Number()),
  address_n_staking: Type.Array(Type.Number()),
  staking_key_hash: Type.Optional(Type.String()),
  certificate_pointer: Type.Optional(CardanoBlockchainPointerType),
  script_payment_hash: Type.Optional(Type.String()),
  script_staking_hash: Type.Optional(Type.String())
}, {
  $id: 'CardanoAddressParametersType'
});
export const CardanoAssetGroup = Type.Object({
  policy_id: Type.String(),
  tokens_count: Type.Number()
}, {
  $id: 'CardanoAssetGroup'
});
export const CardanoCVoteRegistrationDelegation = Type.Object({
  vote_public_key: Type.String(),
  weight: Type.Uint()
}, {
  $id: 'CardanoCVoteRegistrationDelegation'
});
export const CardanoCVoteRegistrationParametersType = Type.Object({
  vote_public_key: Type.Optional(Type.String()),
  staking_path: Type.Array(Type.Number()),
  payment_address_parameters: Type.Optional(CardanoAddressParametersType),
  nonce: Type.Uint(),
  format: Type.Optional(EnumCardanoCVoteRegistrationFormat),
  delegations: Type.Optional(Type.Array(CardanoCVoteRegistrationDelegation)),
  voting_purpose: Type.Optional(Type.Uint()),
  payment_address: Type.Optional(Type.String())
}, {
  $id: 'CardanoCVoteRegistrationParametersType'
});
export const CardanoDRep = Type.Object({
  type: EnumCardanoDRepType,
  key_hash: Type.Optional(Type.String()),
  script_hash: Type.Optional(Type.String())
}, {
  $id: 'CardanoDRep'
});
export const CardanoGetAddress = Type.Object({
  show_display: Type.Optional(Type.Boolean()),
  protocol_magic: Type.Number(),
  network_id: Type.Number(),
  address_parameters: CardanoAddressParametersType,
  derivation_type: EnumCardanoDerivationType,
  chunkify: Type.Optional(Type.Boolean())
}, {
  $id: 'CardanoGetAddress'
});
export const CardanoNativeScript = Type.Recursive(This => Type.Object({
  type: EnumCardanoNativeScriptType,
  scripts: Type.Optional(Type.Array(This)),
  key_hash: Type.Optional(Type.String()),
  key_path: Type.Optional(Type.Array(Type.Number())),
  required_signatures_count: Type.Optional(Type.Number()),
  invalid_before: Type.Optional(Type.Uint()),
  invalid_hereafter: Type.Optional(Type.Uint())
}), {
  $id: 'CardanoNativeScript'
});
export const CardanoGetNativeScriptHash = Type.Object({
  script: CardanoNativeScript,
  display_format: EnumCardanoNativeScriptHashDisplayFormat,
  derivation_type: EnumCardanoDerivationType
}, {
  $id: 'CardanoGetNativeScriptHash'
});
export const CardanoGetPublicKey = Type.Object({
  address_n: Type.Array(Type.Number()),
  show_display: Type.Optional(Type.Boolean()),
  derivation_type: EnumCardanoDerivationType
}, {
  $id: 'CardanoGetPublicKey'
});
export const CardanoMessageDataRequest = Type.Object({
  length: Type.Number(),
  offset: Type.Number()
}, {
  $id: 'CardanoMessageDataRequest'
});
export const CardanoMessageDataResponse = Type.Object({
  data: Type.String()
}, {
  $id: 'CardanoMessageDataResponse'
});
export const CardanoMessageSignature = Type.Object({
  signature: Type.String(),
  address: Type.String(),
  pub_key: Type.String()
}, {
  $id: 'CardanoMessageSignature'
});
export const CardanoNativeScriptHash = Type.Object({
  script_hash: Type.String()
}, {
  $id: 'CardanoNativeScriptHash'
});
export const CardanoPoolMetadataType = Type.Object({
  url: Type.String(),
  hash: Type.String()
}, {
  $id: 'CardanoPoolMetadataType'
});
export const CardanoPoolOwner = Type.Object({
  staking_key_path: Type.Optional(Type.Array(Type.Number())),
  staking_key_hash: Type.Optional(Type.String())
}, {
  $id: 'CardanoPoolOwner'
});
export const CardanoPoolParametersType = Type.Object({
  pool_id: Type.String(),
  vrf_key_hash: Type.String(),
  pledge: Type.Uint(),
  cost: Type.Uint(),
  margin_numerator: Type.Uint(),
  margin_denominator: Type.Uint(),
  reward_account: Type.String(),
  metadata: Type.Optional(CardanoPoolMetadataType),
  owners_count: Type.Number(),
  relays_count: Type.Number()
}, {
  $id: 'CardanoPoolParametersType'
});
export const CardanoPoolRelayParameters = Type.Object({
  type: EnumCardanoPoolRelayType,
  ipv4_address: Type.Optional(Type.String()),
  ipv6_address: Type.Optional(Type.String()),
  host_name: Type.Optional(Type.String()),
  port: Type.Optional(Type.Number())
}, {
  $id: 'CardanoPoolRelayParameters'
});
export const CardanoPublicKey = Type.Object({
  xpub: Type.String(),
  node: HDNodeType
}, {
  $id: 'CardanoPublicKey'
});
export const CardanoSignMessageInit = Type.Object({
  protocol_magic: Type.Optional(Type.Number()),
  network_id: Type.Optional(Type.Number()),
  signing_path: Type.Array(Type.Number()),
  payload_size: Type.Number(),
  prefer_hex_display: Type.Boolean(),
  address_parameters: Type.Optional(CardanoAddressParametersType),
  derivation_type: EnumCardanoDerivationType
}, {
  $id: 'CardanoSignMessageInit'
});
export const CardanoSignTxFinished = Type.Object({}, {
  $id: 'CardanoSignTxFinished'
});
export const CardanoSignTxInit = Type.Object({
  signing_mode: EnumCardanoTxSigningMode,
  protocol_magic: Type.Number(),
  network_id: Type.Number(),
  inputs_count: Type.Number(),
  outputs_count: Type.Number(),
  fee: Type.Uint(),
  ttl: Type.Optional(Type.Uint()),
  certificates_count: Type.Number(),
  withdrawals_count: Type.Number(),
  has_auxiliary_data: Type.Boolean(),
  validity_interval_start: Type.Optional(Type.Uint()),
  witness_requests_count: Type.Number(),
  minting_asset_groups_count: Type.Number(),
  derivation_type: EnumCardanoDerivationType,
  include_network_id: Type.Optional(Type.Boolean()),
  script_data_hash: Type.Optional(Type.String()),
  collateral_inputs_count: Type.Number(),
  required_signers_count: Type.Number(),
  has_collateral_return: Type.Optional(Type.Boolean()),
  total_collateral: Type.Optional(Type.Uint()),
  reference_inputs_count: Type.Optional(Type.Number()),
  chunkify: Type.Optional(Type.Boolean()),
  tag_cbor_sets: Type.Optional(Type.Boolean()),
  payment_req: Type.Optional(PaymentRequest)
}, {
  $id: 'CardanoSignTxInit'
});
export const CardanoToken = Type.Object({
  asset_name_bytes: Type.String(),
  amount: Type.Optional(Type.Uint()),
  mint_amount: Type.Optional(Type.Uint({
    allowNegative: true
  }))
}, {
  $id: 'CardanoToken'
});
export const CardanoTxAuxiliaryData = Type.Object({
  cvote_registration_parameters: Type.Optional(CardanoCVoteRegistrationParametersType),
  hash: Type.Optional(Type.String())
}, {
  $id: 'CardanoTxAuxiliaryData'
});
export const CardanoTxAuxiliaryDataSupplement = Type.Object({
  type: EnumCardanoTxAuxiliaryDataSupplementType,
  auxiliary_data_hash: Type.Optional(Type.String()),
  cvote_registration_signature: Type.Optional(Type.String())
}, {
  $id: 'CardanoTxAuxiliaryDataSupplement'
});
export const CardanoTxBodyHash = Type.Object({
  tx_hash: Type.String()
}, {
  $id: 'CardanoTxBodyHash'
});
export const CardanoTxCertificate = Type.Object({
  type: EnumCardanoCertificateType,
  path: Type.Optional(Type.Array(Type.Number())),
  pool: Type.Optional(Type.String()),
  pool_parameters: Type.Optional(CardanoPoolParametersType),
  script_hash: Type.Optional(Type.String()),
  key_hash: Type.Optional(Type.String()),
  deposit: Type.Optional(Type.Uint()),
  drep: Type.Optional(CardanoDRep)
}, {
  $id: 'CardanoTxCertificate'
});
export const CardanoTxCollateralInput = Type.Object({
  prev_hash: Type.String(),
  prev_index: Type.Number()
}, {
  $id: 'CardanoTxCollateralInput'
});
export const CardanoTxHostAck = Type.Object({}, {
  $id: 'CardanoTxHostAck'
});
export const CardanoTxInlineDatumChunk = Type.Object({
  data: Type.String()
}, {
  $id: 'CardanoTxInlineDatumChunk'
});
export const CardanoTxInput = Type.Object({
  prev_hash: Type.String(),
  prev_index: Type.Number()
}, {
  $id: 'CardanoTxInput'
});
export const CardanoTxItemAck = Type.Object({}, {
  $id: 'CardanoTxItemAck'
});
export const CardanoTxMint = Type.Object({
  asset_groups_count: Type.Number()
}, {
  $id: 'CardanoTxMint'
});
export const CardanoTxOutput = Type.Object({
  address: Type.Optional(Type.String()),
  address_parameters: Type.Optional(CardanoAddressParametersType),
  amount: Type.Uint(),
  asset_groups_count: Type.Number(),
  datum_hash: Type.Optional(Type.String()),
  format: Type.Optional(EnumCardanoTxOutputSerializationFormat),
  inline_datum_size: Type.Optional(Type.Number()),
  reference_script_size: Type.Optional(Type.Number())
}, {
  $id: 'CardanoTxOutput'
});
export const CardanoTxReferenceInput = Type.Object({
  prev_hash: Type.String(),
  prev_index: Type.Number()
}, {
  $id: 'CardanoTxReferenceInput'
});
export const CardanoTxReferenceScriptChunk = Type.Object({
  data: Type.String()
}, {
  $id: 'CardanoTxReferenceScriptChunk'
});
export const CardanoTxRequiredSigner = Type.Object({
  key_hash: Type.Optional(Type.String()),
  key_path: Type.Optional(Type.Array(Type.Number()))
}, {
  $id: 'CardanoTxRequiredSigner'
});
export const CardanoTxWithdrawal = Type.Object({
  path: Type.Optional(Type.Array(Type.Number())),
  amount: Type.Uint(),
  script_hash: Type.Optional(Type.String()),
  key_hash: Type.Optional(Type.String())
}, {
  $id: 'CardanoTxWithdrawal'
});
export const CardanoTxWitnessRequest = Type.Object({
  path: Type.Array(Type.Number())
}, {
  $id: 'CardanoTxWitnessRequest'
});
export const CardanoTxWitnessResponse = Type.Object({
  type: EnumCardanoTxWitnessType,
  pub_key: Type.String(),
  signature: Type.String(),
  chain_code: Type.Optional(Type.String())
}, {
  $id: 'CardanoTxWitnessResponse'
});
