// @trezor/connect-common/libESM/types/api/stellar/index.mjs
import { MessagesSchema as PROTO } from "@trezor/protobuf";
import { Type } from "@trezor/schema-utils";
import { DerivationPath } from "../../params.mjs";
export const StellarAsset = Type.Object({
  type: Type.Union([PROTO.EnumStellarAssetType, Type.KeyOfEnum(PROTO.StellarAssetType)]),
  code: Type.Optional(Type.String()),
  issuer: Type.Optional(Type.String())
});
export const StellarCreateAccountOperation = Type.Object({
  type: Type.Literal('createAccount'),
  source: Type.Optional(Type.String()),
  destination: Type.String(),
  startingBalance: Type.String()
});
export const StellarPaymentOperation = Type.Object({
  type: Type.Literal('payment'),
  source: Type.Optional(Type.String()),
  destination: Type.String(),
  asset: StellarAsset,
  amount: Type.String()
});
export const StellarPathPaymentStrictReceiveOperation = Type.Object({
  type: Type.Literal('pathPaymentStrictReceive'),
  source: Type.Optional(Type.String()),
  sendAsset: StellarAsset,
  sendMax: Type.Uint(),
  destination: Type.String(),
  destAsset: StellarAsset,
  destAmount: Type.Uint(),
  path: Type.Optional(Type.Array(StellarAsset))
});
export const StellarPathPaymentStrictSendOperation = Type.Object({
  type: Type.Literal('pathPaymentStrictSend'),
  source: Type.Optional(Type.String()),
  sendAsset: StellarAsset,
  sendAmount: Type.Uint(),
  destination: Type.String(),
  destAsset: StellarAsset,
  destMin: Type.Uint(),
  path: Type.Optional(Type.Array(StellarAsset))
});
export const StellarPassiveSellOfferOperation = Type.Object({
  type: Type.Literal('createPassiveSellOffer'),
  source: Type.Optional(Type.String()),
  buying: StellarAsset,
  selling: StellarAsset,
  amount: Type.Uint(),
  price: Type.Object({
    n: Type.Number(),
    d: Type.Number()
  })
});
export const StellarManageSellOfferOperation = Type.Object({
  type: Type.Literal('manageSellOffer'),
  source: Type.Optional(Type.String()),
  buying: StellarAsset,
  selling: StellarAsset,
  amount: Type.Uint(),
  offerId: Type.Optional(Type.Uint()),
  price: Type.Object({
    n: Type.Number(),
    d: Type.Number()
  })
});
export const StellarManageBuyOfferOperation = Type.Object({
  type: Type.Literal('manageBuyOffer'),
  source: Type.Optional(Type.String()),
  buying: StellarAsset,
  selling: StellarAsset,
  amount: Type.Uint(),
  offerId: Type.Optional(Type.Uint()),
  price: Type.Object({
    n: Type.Number(),
    d: Type.Number()
  })
});
export const StellarSetOptionsOperation = Type.Object({
  type: Type.Literal('setOptions'),
  source: Type.Optional(Type.String()),
  signer: Type.Optional(Type.Object({
    type: PROTO.EnumStellarSignerType,
    key: Type.Union([Type.String(), Type.Buffer()]),
    weight: Type.Optional(Type.Number())
  })),
  inflationDest: Type.Optional(Type.String()),
  clearFlags: Type.Optional(Type.Number()),
  setFlags: Type.Optional(Type.Number()),
  masterWeight: Type.Optional(Type.Uint()),
  lowThreshold: Type.Optional(Type.Uint()),
  medThreshold: Type.Optional(Type.Uint()),
  highThreshold: Type.Optional(Type.Uint()),
  homeDomain: Type.Optional(Type.String())
});
export const StellarChangeTrustOperation = Type.Object({
  type: Type.Literal('changeTrust'),
  source: Type.Optional(Type.String()),
  line: StellarAsset,
  limit: Type.String()
});
export const StellarAllowTrustOperation = Type.Object({
  type: Type.Literal('allowTrust'),
  source: Type.Optional(Type.String()),
  trustor: Type.String(),
  assetCode: Type.String(),
  assetType: PROTO.EnumStellarAssetType,
  authorize: Type.Optional(Type.Union([Type.Boolean(), Type.Undefined()]))
});
export const StellarAccountMergeOperation = Type.Object({
  type: Type.Literal('accountMerge'),
  source: Type.Optional(Type.String()),
  destination: Type.String()
});
export const StellarManageDataOperation = Type.Object({
  type: Type.Literal('manageData'),
  source: Type.Optional(Type.String()),
  name: Type.String(),
  value: Type.Optional(Type.Union([Type.String(), Type.Buffer()]))
});
export const StellarBumpSequenceOperation = Type.Object({
  type: Type.Literal('bumpSequence'),
  source: Type.Optional(Type.String()),
  bumpTo: Type.Uint()
});
export const StellarInflationOperation = Type.Object({
  type: Type.Literal('inflation'),
  source: Type.Optional(Type.String())
});
export const StellarClaimClaimableBalanceOperation = Type.Object({
  type: Type.Literal('claimClaimableBalance'),
  source: Type.Optional(Type.String()),
  balanceId: Type.String()
});
export const StellarOperation = Type.Union([StellarCreateAccountOperation, StellarPaymentOperation, StellarPathPaymentStrictReceiveOperation, StellarPathPaymentStrictSendOperation, StellarPassiveSellOfferOperation, StellarManageSellOfferOperation, StellarManageBuyOfferOperation, StellarSetOptionsOperation, StellarChangeTrustOperation, StellarAllowTrustOperation, StellarAccountMergeOperation, StellarInflationOperation, StellarManageDataOperation, StellarBumpSequenceOperation, StellarClaimClaimableBalanceOperation]);
export const StellarTransaction = Type.Object({
  source: Type.String(),
  fee: Type.Number(),
  sequence: Type.Uint(),
  timebounds: Type.Optional(Type.Object({
    minTime: Type.Number(),
    maxTime: Type.Number()
  })),
  memo: Type.Optional(Type.Object({
    type: PROTO.EnumStellarMemoType,
    id: Type.Optional(Type.Uint()),
    text: Type.Optional(Type.String()),
    hash: Type.Optional(Type.Union([Type.String(), Type.Buffer()]))
  })),
  operations: Type.Array(StellarOperation)
});
export const StellarSignTransaction = Type.Object({
  path: DerivationPath,
  networkPassphrase: Type.String(),
  transaction: StellarTransaction,
  payment_req: Type.Optional(PROTO.PaymentRequest)
});
export const StellarSignedTx = Type.Object({
  publicKey: Type.String(),
  signature: Type.String()
});
export const StellarOperationMessage = Type.Union([Type.Intersect([Type.Object({
  type: Type.Literal('StellarCreateAccountOp')
}), PROTO.StellarCreateAccountOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarPaymentOp')
}), PROTO.StellarPaymentOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarPathPaymentStrictReceiveOp')
}), PROTO.StellarPathPaymentStrictReceiveOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarPathPaymentStrictSendOp')
}), PROTO.StellarPathPaymentStrictSendOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarManageSellOfferOp')
}), PROTO.StellarManageSellOfferOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarManageBuyOfferOp')
}), PROTO.StellarManageBuyOfferOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarCreatePassiveSellOfferOp')
}), PROTO.StellarCreatePassiveSellOfferOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarSetOptionsOp')
}), PROTO.StellarSetOptionsOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarChangeTrustOp')
}), PROTO.StellarChangeTrustOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarAllowTrustOp')
}), PROTO.StellarAllowTrustOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarAccountMergeOp')
}), PROTO.StellarAccountMergeOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarManageDataOp')
}), PROTO.StellarManageDataOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarBumpSequenceOp')
}), PROTO.StellarBumpSequenceOp]), Type.Intersect([Type.Object({
  type: Type.Literal('StellarClaimClaimableBalanceOp')
}), PROTO.StellarClaimClaimableBalanceOp])]);
