// @trezor/protocol/libESM/protocol-thp/crypto/pairing.mjs
import { createHash, randomBytes } from "crypto";
import { aesgcm } from "./aesgcm.mjs";
import { curve25519, elligator2, getCurve25519KeyPair } from "./curve25519.mjs";
import { bigEndianBytesToBigInt, getIvFromNonce, hashOfTwo, hkdf, sha256 } from "./tools.mjs";
import { ThpPairingMethod } from "../messages/index.mjs";
const getProtocolName = () => Buffer.concat([Buffer.from('Noise_XX_25519_AESGCM_SHA256'), Buffer.alloc(4).fill(0)]);
export const getHandshakeHash = deviceProperties => hashOfTwo(getProtocolName(), deviceProperties);
export const findKnownPairingCredentials = (knownCredentials, trezorMaskedStaticPubkey, trezorEphemeralPubkey) => knownCredentials.filter(c => {
  try {
    const trezorStaticPubkey = Buffer.from(c.trezor_static_public_key, 'hex');
    const h = hashOfTwo(trezorStaticPubkey, trezorEphemeralPubkey);
    return trezorMaskedStaticPubkey.compare(curve25519(h, trezorStaticPubkey)) === 0;
  } catch {}
});
export const getTrezorState = (credentials, payload) => {
  const aes = aesgcm(credentials.trezorKey, Buffer.alloc(12));
  aes.auth(Buffer.alloc(0));
  const trezorState = aes.decrypt(payload.subarray(0, 1), payload.subarray(1, 17));
  return trezorState.readUint8();
};
export const handleHandshakeInit = ({
  handshakeInitResponse,
  thpState,
  knownCredentials,
  hostEphemeralKeys,
  tryToUnlock,
  protobufEncoder
}) => {
  if (!thpState.handshakeCredentials) {
    throw new Error('ThpStateMissing');
  }
  const {
    trezorEphemeralPubkey,
    trezorEncryptedStaticPubkey,
    tag
  } = handshakeInitResponse;
  const {
    sendNonce,
    recvNonce
  } = thpState;
  const {
    handshakeHash
  } = thpState.handshakeCredentials;
  const iv0 = getIvFromNonce(sendNonce);
  const iv1 = getIvFromNonce(recvNonce);
  let h, point, aes;
  h = handshakeHash;
  h = hashOfTwo(h, hostEphemeralKeys.publicKey);
  h = hashOfTwo(h, Buffer.from([tryToUnlock]));
  h = hashOfTwo(h, trezorEphemeralPubkey);
  point = curve25519(hostEphemeralKeys.privateKey, trezorEphemeralPubkey);
  let [ck, k] = hkdf(getProtocolName(), point);
  aes = aesgcm(k, iv0);
  aes.auth(h);
  const trezorStaticPubkey = trezorEncryptedStaticPubkey.subarray(0, 32);
  const trezorStaticPubkeyTag = trezorEncryptedStaticPubkey.subarray(32, 32 + 16);
  const trezorMaskedStaticPubkey = aes.decrypt(trezorStaticPubkey, trezorStaticPubkeyTag);
  h = hashOfTwo(h, trezorEncryptedStaticPubkey);
  point = curve25519(hostEphemeralKeys.privateKey, trezorMaskedStaticPubkey);
  [ck, k] = hkdf(ck, point);
  aes = aesgcm(k, iv0);
  aes.auth(h);
  aes.decrypt(Buffer.alloc(0), tag);
  h = hashOfTwo(h, tag);
  const allCredentials = findKnownPairingCredentials(knownCredentials, trezorMaskedStaticPubkey, trezorEphemeralPubkey);
  const credentials = allCredentials.length ? allCredentials[0] : undefined;
  const staticKey = credentials?.host_static_key ? Buffer.from(credentials.host_static_key, 'hex') : randomBytes(32);
  const hostStaticKeys = getCurve25519KeyPair(staticKey);
  aes = aesgcm(k, iv1);
  aes.auth(h);
  const hostEncryptedStaticPubkey = Buffer.concat([aes.encrypt(hostStaticKeys.publicKey), aes.finish()]);
  h = hashOfTwo(h, hostEncryptedStaticPubkey);
  point = curve25519(hostStaticKeys.privateKey, trezorEphemeralPubkey);
  [ck, k] = hkdf(ck, point);
  const {
    message
  } = protobufEncoder('ThpHandshakeCompletionReqNoisePayload', {
    host_pairing_credential: credentials?.credential
  });
  aes = aesgcm(k, iv0);
  aes.auth(h);
  const encryptedPayload = Buffer.concat([aes.encrypt(message), aes.finish()]);
  h = hashOfTwo(h, encryptedPayload);
  const [hostKey, trezorKey] = hkdf(ck, Buffer.alloc(0));
  return {
    trezorMaskedStaticPubkey,
    trezorEncryptedStaticPubkey,
    hostEncryptedStaticPubkey,
    hostKey,
    trezorKey,
    handshakeHash: h,
    credentials,
    allCredentials,
    encryptedPayload,
    staticKey,
    hostStaticKeys
  };
};
export const getCpaceHostKeys = (code, handshakeHash) => {
  const shaCtx = createHash('sha512');
  shaCtx.update(Buffer.from([0x08, 0x43, 0x50, 0x61, 0x63, 0x65, 0x32, 0x35, 0x35, 0x06]));
  shaCtx.update(code);
  shaCtx.update(Buffer.concat([Buffer.from([0x6f]), Buffer.alloc(111).fill(0), Buffer.from([0x20])]));
  shaCtx.update(handshakeHash);
  shaCtx.update(Buffer.from([0x00]));
  const sha = shaCtx.digest().subarray(0, 32);
  const generator = elligator2(sha);
  const privateKey = randomBytes(32);
  const publicKey = curve25519(privateKey, generator);
  return {
    privateKey,
    publicKey
  };
};
export const getSharedSecret = (publicKey, privateKey) => {
  const sharedSecret = curve25519(privateKey, publicKey);
  return sha256(Buffer.from(sharedSecret));
};
export const validateCodeEntryTag = (credentials, value, secret) => {
  const sha = createHash('sha256').update(Buffer.from(secret, 'hex')).digest();
  const {
    handshakeHash,
    handshakeCommitment,
    codeEntryChallenge
  } = credentials;
  if (sha.compare(handshakeCommitment) !== 0) {
    throw new Error(`HP5: commitment mismatch ${handshakeCommitment.toString('hex')} != ${sha.toString('hex')}`);
  }
  const shaCtx = createHash('sha256');
  shaCtx.update(Buffer.from([ThpPairingMethod.CodeEntry]));
  shaCtx.update(handshakeHash);
  shaCtx.update(Buffer.from(secret, 'hex'));
  shaCtx.update(codeEntryChallenge);
  const calculatedValue = bigEndianBytesToBigInt(shaCtx.digest()) % 1000000n;
  if (calculatedValue !== BigInt(value)) {
    throw new Error(`HP5: code mismatch ${value} != ${calculatedValue.toString()}`);
  }
};
const validatePairingTag = (handshakeHash, method, secret, value, errorCode) => {
  const shaCtx = createHash('sha256');
  shaCtx.update(Buffer.from([method]));
  shaCtx.update(handshakeHash);
  shaCtx.update(secret);
  const calculatedValue = shaCtx.digest().subarray(0, 16);
  const expectedValue = Buffer.from(value, 'hex').subarray(0, 16);
  if (calculatedValue.compare(expectedValue) !== 0) {
    throw new Error(`${errorCode}: code mismatch ${calculatedValue.toString('hex')} != ${expectedValue.toString('hex')}`);
  }
};
export const validateQrCodeTag = ({
  handshakeHash
}, value, secret) => validatePairingTag(handshakeHash, ThpPairingMethod.QrCode, Buffer.from(secret, 'hex'), value, 'HP6');
export const validateNfcTag = ({
  handshakeHash
}, value, secret) => validatePairingTag(handshakeHash, ThpPairingMethod.NFC, secret, value, 'HP7');
