// @trezor/connect-web/libESM/impl/core-in-suite-web.mjs
import * as ERRORS from "@trezor/connect-common/libESM/constants/errors.mjs";
import { CORE_CALL, CORE_CALL_CANCEL, createErrorMessage } from "@trezor/connect-common/libESM/events/index.mjs";
import { initLog } from "@trezor/connect-common/libESM/utils/debug.mjs";
import { getEnv } from "../connectSettings.mjs";
import { getSuiteWebUrl } from "./getSuiteWebUrl.mjs";
import { WebPopup } from "../popup/web.mjs";
import { WebExtensionPopup } from "../popup/webextension.mjs";
export class CoreInSuiteWeb {
  _popupManager;
  logger;
  constructor() {
    this.logger = initLog('@trezor/connect-web');
  }
  dispose() {
    return Promise.resolve(undefined);
  }
  cancel(reason) {
    this.logger.debug('cancel', reason);
    this._popupManager?.channel?.clearPendingSends();
    this._popupManager?.channel?.postMessage({
      type: CORE_CALL_CANCEL,
      payload: reason ? {
        reason
      } : null
    }, {
      usePromise: false
    });
  }
  init({
    env,
    manifest,
    version,
    debug
  }) {
    this.logger.enabled = !!debug;
    if (!this._popupManager) {
      const resolvedEnv = env ?? getEnv();
      const params = {
        manifest,
        version,
        popupSrc: this.getSuiteUrl(),
        logger: this.logger
      };
      switch (resolvedEnv) {
        case 'webextension':
          this._popupManager = new WebExtensionPopup(params);
          break;
        case 'web':
          this._popupManager = new WebPopup(params);
          break;
        default:
          this.logger.error('Unsupported environment for connect-web:', resolvedEnv);
          break;
      }
    }
    this.logger.debug('initiated');
    return Promise.resolve();
  }
  getSuiteUrl() {
    if (typeof __SUITE_WEB_URL__ !== 'undefined' && __SUITE_WEB_URL__) {
      return `${__SUITE_WEB_URL__}/connect-popup`;
    }
    if (typeof window !== 'undefined') {
      return getSuiteWebUrl(window.location.href, window.location.origin);
    }
    return 'https://suite.trezor.io/web/connect-popup';
  }
  async call(params) {
    this.logger.debug('call', params);
    if (!this._popupManager) {
      return createErrorMessage(ERRORS.TypedError('Init_NotInitialized'));
    }
    try {
      this.logger.debug('call - awaiting popupManger.focusOrCreate()');
      await this._popupManager.focusOrCreate();
      this.logger.debug('call - popupManager.request() resolved');
      await this._popupManager.channel.init();
      this.logger.debug('call - popupManager.channel.init() resolved');
      await this._popupManager.handshakePromise?.promise;
      this.logger.debug('call - sending iframe.call');
      const response = await this._popupManager.channel.postMessage({
        type: CORE_CALL,
        payload: params
      });
      this.logger.debug('call: response: ', response);
      if (!response?.payload) {
        throw ERRORS.TypedError('Method_NoResponse');
      }
      if (response.payload.error && response.payload.error.code) {
        throw response.payload.error;
      }
      return {
        success: response.payload.success,
        payload: response.payload.payload,
        device: response.payload.device
      };
    } catch (error) {
      this.logger.error('call: error', error);
      return createErrorMessage(error);
    }
  }
}
