// @trezor/connect-web/libESM/impl/core-in-suite-desktop.mjs
import * as ERRORS from "@trezor/connect-common/libESM/constants/errors.mjs";
import { CORE_CALL, CORE_CALL_CANCEL, POPUP } from "@trezor/connect-common/libESM/events/index.mjs";
import { WebsocketClient, WebsocketError } from "@trezor/websocket-client";
export class CoreInSuiteDesktop {
  manifest;
  version;
  ws;
  localNetworkPermissionState = 'unknown';
  constructor() {
    this.ws = new WebsocketClient({
      url: 'ws://127.0.0.1:21335/connect-ws'
    });
  }
  dispose() {
    this.manifest = undefined;
    this.version = undefined;
    this.ws.dispose();
    return Promise.resolve(undefined);
  }
  cancel(reason) {
    this.ws.sendMessage({
      type: CORE_CALL_CANCEL,
      payload: reason ? {
        reason
      } : null
    });
  }
  async handshake() {
    if (!this.ws) {
      throw ERRORS.TypedError('Desktop_ConnectionMissing', 'No websocket connection');
    }
    try {
      const response = await this.ws.sendMessage({
        type: POPUP.HANDSHAKE,
        payload: {
          settings: {
            manifest: this.manifest,
            version: this.version
          }
        }
      }, {
        timeout: 3000
      });
      if (!response) {
        throw ERRORS.TypedError('Desktop_ConnectionMissing', 'No response');
      }
      return response;
    } catch (err) {
      throw ERRORS.TypedError('Desktop_ConnectionMissing', err.message);
    }
  }
  async init({
    manifest,
    version
  }) {
    if (typeof navigator !== 'undefined' && navigator?.permissions?.query) {
      const permission = await navigator.permissions.query({
        name: 'local-network-access'
      }).catch(() => undefined);
      if (permission) {
        this.localNetworkPermissionState = permission.state;
        permission.onchange = () => {
          this.localNetworkPermissionState = permission.state;
        };
      }
    }
    if (!manifest.appName) {
      throw ERRORS.TypedError('Init_ManifestMissing', 'Manifest is missing or manifest.appName is not set');
    }
    this.manifest = manifest;
    this.version = version;
    return await this.connect();
  }
  error(err) {
    if (err instanceof WebsocketError) {
      if (this.localNetworkPermissionState === 'denied') {
        return ERRORS.TypedError('Browser_LocalNetworkPermissionMissing');
      } else {
        return ERRORS.TypedError('Desktop_ConnectionMissing', err.message);
      }
    }
    return err;
  }
  async connect() {
    try {
      await this.ws.connect();
    } catch (err) {
      throw this.error(err);
    }
  }
  async call(params) {
    try {
      if (!this.ws.isConnected()) {
        await this.connect();
      }
      await this.handshake();
      const response = await this.ws.sendMessage({
        type: CORE_CALL,
        payload: params
      }, {
        timeout: 0
      });
      if (!response) {
        throw ERRORS.TypedError('Desktop_ConnectionMissing', 'No response');
      }
      if (response.success === false) {
        return {
          success: false,
          error: response.error
        };
      }
      return {
        success: true,
        payload: response.payload,
        device: response.device
      };
    } catch (err) {
      return {
        success: false,
        error: ERRORS.serializeError(this.error(err))
      };
    }
  }
}
