// @trezor/device-utils/libESM/firmwareUtils.mjs
import { isDeviceInBootloaderMode } from "./modeUtils.mjs";
import { FirmwareType } from "./types.mjs";
export const getFirmwareSource = device => {
  if (device?.mode === 'bootloader') {
    return 'NA - bootloader';
  }
  if (device?.authenticityChecks?.firmwareRevision?.success && device?.authenticityChecks?.firmwareHash?.success) {
    return 'official';
  }
  return 'unknown';
};
export const getFirmwareRevision = device => device?.features?.revision || '';
export const getFirmwareOrBootloaderVersionArray = features => [features.major_version, features.minor_version, features.patch_version];
export const getFirmwareVersionArray = device => {
  if (!device?.features) {
    return null;
  }
  const {
    features
  } = device;
  if (isDeviceInBootloaderMode(device)) {
    return features.fw_major ? [features.fw_major, features.fw_minor, features.fw_patch] : null;
  }
  return getFirmwareOrBootloaderVersionArray(features);
};
export const getFirmwareVersion = device => {
  const versionArray = getFirmwareVersionArray(device);
  return versionArray === null ? '' : versionArray.join('.');
};
export const hasBitcoinOnlyFirmware = device => device?.firmwareType === FirmwareType.BitcoinOnly;
export const isBitcoinOnlyDevice = device => !!device?.features?.unit_btconly && device?.firmwareType !== FirmwareType.Universal;
