// @trezor/protobuf/libESM/manager.mjs
import { ScalarType, fromBinary, fromJson, toBinary, toJson } from "@bufbuild/protobuf";
const patchLongValue = value => {
  if (typeof value === 'number') {
    return value;
  }
  const numericValue = Number(value);
  if (typeof value === 'bigint') {
    return Number.isSafeInteger(numericValue) ? numericValue : value.toString();
  }
  if (typeof value === 'string') {
    if (Number.isSafeInteger(numericValue)) {
      try {
        if (BigInt(value) === BigInt(numericValue)) {
          return numericValue;
        }
      } catch {
        return value;
      }
    }
    return value;
  }
  return value;
};
const patchFieldForEncode = (field, value) => {
  if (value === undefined) {
    return value;
  }
  if (field.scalar === ScalarType.STRING) {
    if (typeof value === 'number') {
      return value.toString();
    }
    return value;
  }
  if (field.scalar === ScalarType.BYTES) {
    if (typeof value === 'string') {
      return value.length === 0 ? Buffer.alloc(0) : Buffer.from(value, 'hex');
    }
    return Buffer.from(value);
  }
  return value;
};
const patchFieldForDecode = (field, value) => {
  if (value === undefined) {
    return null;
  }
  if (field.scalar === ScalarType.BYTES) {
    if (typeof value === 'string') {
      return Buffer.from(value, 'base64').toString('hex');
    }
    return Buffer.from(value).toString('hex');
  }
  if (field.scalar === ScalarType.UINT32 || field.scalar === ScalarType.SINT32) {
    return Number(value);
  }
  if (field.scalar === ScalarType.INT64 || field.scalar === ScalarType.UINT64 || field.scalar === ScalarType.SINT64 || field.scalar === ScalarType.FIXED64 || field.scalar === ScalarType.SFIXED64) {
    return patchLongValue(value);
  }
  return value;
};
const transformSchemaFields = (operation, schema, sourceData) => {
  const transformedData = {};
  const transformField = operation === 'encode' ? patchFieldForEncode : patchFieldForDecode;
  schema.fields.forEach(schemaField => {
    const fieldValue = sourceData?.[schemaField.name];
    if (schemaField.fieldKind === 'list') {
      if (!Array.isArray(fieldValue)) {
        transformedData[schemaField.name] = [];
      } else if (schemaField.listKind === 'message') {
        transformedData[schemaField.name] = fieldValue.map(item => transformSchemaFields(operation, schemaField.message, item));
      } else if (schemaField.listKind === 'scalar') {
        transformedData[schemaField.name] = fieldValue.map(item => transformField(schemaField, item));
      } else {
        transformedData[schemaField.name] = fieldValue;
      }
    } else if (schemaField.fieldKind === 'message') {
      if (!fieldValue) {
        transformedData[schemaField.name] = operation === 'decode' ? {} : undefined;
      } else {
        transformedData[schemaField.name] = transformSchemaFields(operation, schemaField.message, fieldValue);
      }
    } else if (schemaField.fieldKind === 'scalar') {
      transformedData[schemaField.name] = transformField(schemaField, fieldValue);
    } else if (schemaField.fieldKind === 'enum') {
      transformedData[schemaField.name] = fieldValue;
    } else {
      transformedData[schemaField.name] = fieldValue;
    }
    if (transformedData[schemaField.name] === undefined) {
      delete transformedData[schemaField.name];
    }
  });
  return transformedData;
};
export const ProtobufManager = () => {
  const messages = {};
  const extensions = {};
  const enums = {};
  const load = modules => {
    if (Array.isArray(modules)) {
      modules.forEach(m => load(m));
    } else {
      Object.keys(modules).forEach(key => {
        const def = modules[key];
        if (def.kind && def.kind !== 'file') {
          if (def.kind === 'message') {
            messages[key] = def;
          } else if (def.kind === 'extension') {
            extensions[key] = def;
          } else if (def.kind === 'enum') {
            enums[def.name] = def;
          }
        }
      });
    }
  };
  const getWireTypeOverride = schema => {
    const unknownOptions = schema.proto.options?.['$unknown'] ?? [];
    if (unknownOptions.length === 0) {
      return;
    }
    const wireTypeExtension = extensions['wire_type'];
    const wireTypeOption = unknownOptions.find(o => o.no === wireTypeExtension?.number);
    return wireTypeOption ? wireTypeOption.data[0] : undefined;
  };
  const findEnum = enumName => enums[enumName];
  const findSchema = nameOrId => {
    const messageTypeEnum = findEnum('MessageType');
    const thpMessageTypeEnum = findEnum('ThpMessageType');
    if (typeof nameOrId === 'string') {
      const schemaKey = `${nameOrId}Schema`;
      const schema = messages[schemaKey];
      if (!schema) {
        throw new Error(`Schema ${nameOrId} not found`);
      }
      let messageType = -1;
      if (thpMessageTypeEnum) {
        const thpEnum = thpMessageTypeEnum.values.find(k => k.name === 'ThpMessageType_' + schema.name);
        if (thpEnum) {
          messageType = thpEnum.number;
        }
      }
      if (messageType === -1 && messageTypeEnum) {
        const mtEnum = messageTypeEnum.values.find(k => k.name === 'MessageType_' + schema.name);
        if (mtEnum) {
          messageType = mtEnum.number;
        }
      }
      return {
        schema,
        messageType: getWireTypeOverride(schema) || messageType,
        messageName: nameOrId
      };
    }
    let matchedMessageType = messageTypeEnum?.value[nameOrId];
    if (!matchedMessageType && thpMessageTypeEnum) {
      matchedMessageType = thpMessageTypeEnum.value[nameOrId];
    }
    if (!matchedMessageType) {
      throw new Error(`MessageType with ID ${nameOrId} not found. Check MessageType and ThpMessageType enums.`);
    }
    const messageName = matchedMessageType.name.replace('ThpMessageType_', '').replace('MessageType_', '');
    const schemaKey = messageName + 'Schema';
    const schema = messages[schemaKey];
    if (!schema) {
      throw new Error(`Schema "${messageName}" (from ID ${nameOrId}) not found. Available: ${Object.keys(messages).length}`);
    }
    return {
      schema,
      messageType: getWireTypeOverride(schema) || nameOrId,
      messageName
    };
  };
  const encode = (messageName, data) => {
    const {
      schema,
      messageType
    } = findSchema(messageName);
    const normalizedPayload = transformSchemaFields('encode', schema, data);
    const messageJson = fromJson(schema, normalizedPayload);
    const encodedBytes = toBinary(schema, messageJson);
    return {
      messageType,
      message: Buffer.from(encodedBytes)
    };
  };
  const decode = (type, data) => {
    const {
      schema,
      messageName
    } = findSchema(type);
    const binaryPayload = Buffer.isBuffer(data) ? data : Buffer.from(data);
    const patchedPayload = messageName === 'Failure' && binaryPayload.subarray(0, 2).compare(Buffer.from('0811', 'hex')) === 0 ? binaryPayload.subarray(0, 2) : binaryPayload;
    const decoded = fromBinary(schema, patchedPayload, {
      readUnknownFields: false
    });
    const json = toJson(schema, decoded, {
      useProtoFieldName: true
    });
    const message = transformSchemaFields('decode', schema, json);
    return {
      type: messageName,
      message
    };
  };
  return {
    load,
    findSchema,
    findEnum,
    encode,
    decode
  };
};
