// @trezor/connect-common/libESM/types/api/ethereum/index.mjs
import { MessagesSchema as PROTO } from "@trezor/protobuf";
import { Type } from "@trezor/schema-utils";
import { DerivationPath } from "../../params.mjs";
export const EthereumSignMessage = Type.Object({
  path: DerivationPath,
  message: Type.String(),
  hex: Type.Optional(Type.Boolean())
});
export const EthereumTransaction = Type.Object({
  to: Type.Union([Type.String(), Type.Null()]),
  value: Type.String(),
  gasPrice: Type.String(),
  gasLimit: Type.String(),
  maxFeePerGas: Type.Optional(Type.Undefined()),
  maxPriorityFeePerGas: Type.Optional(Type.Undefined()),
  nonce: Type.String(),
  data: Type.Optional(Type.String()),
  chainId: Type.Number(),
  txType: Type.Optional(Type.Number()),
  payment_req: Type.Optional(PROTO.PaymentRequest)
});
export const EthereumAccessList = Type.Object({
  address: Type.String(),
  storageKeys: Type.Array(Type.String())
});
export const EthereumTransactionEIP1559 = Type.Object({
  to: Type.Union([Type.String(), Type.Null()]),
  value: Type.String(),
  gasLimit: Type.String(),
  gasPrice: Type.Optional(Type.Undefined()),
  nonce: Type.String(),
  data: Type.Optional(Type.String()),
  chainId: Type.Number(),
  maxFeePerGas: Type.String(),
  maxPriorityFeePerGas: Type.String(),
  accessList: Type.Optional(Type.Array(EthereumAccessList)),
  payment_req: Type.Optional(PROTO.PaymentRequest)
});
export const EthereumSignTransaction = Type.Object({
  path: DerivationPath,
  transaction: Type.Union([EthereumTransaction, EthereumTransactionEIP1559]),
  chunkify: Type.Optional(Type.Boolean())
});
export const EthereumSignedTx = Type.Object({
  v: Type.String(),
  r: Type.String(),
  s: Type.String(),
  serializedTx: Type.String()
});
const EthereumSignTypedDataTypeProperty = Type.Object({
  name: Type.String(),
  type: Type.String()
});
export const EthereumSignTypedDataTypes = Type.Object({
  EIP712Domain: Type.Array(EthereumSignTypedDataTypeProperty)
}, {
  additionalProperties: Type.Array(EthereumSignTypedDataTypeProperty)
});
export const EthereumSignTypedDataMessage = Type.Object({
  types: EthereumSignTypedDataTypes,
  primaryType: Type.String(),
  domain: Type.Object({
    name: Type.Optional(Type.String()),
    version: Type.Optional(Type.String()),
    chainId: Type.Optional(Type.Union([Type.Number(), Type.BigInt(), Type.String()])),
    verifyingContract: Type.Optional(Type.String()),
    salt: Type.Optional(Type.Union([Type.ArrayBuffer(), Type.String()]))
  }),
  message: Type.Object({}, {
    additionalProperties: Type.Any()
  })
});
export const EthereumSignTypedData = Type.Object({
  path: DerivationPath,
  data: EthereumSignTypedDataMessage,
  metamask_v4_compat: Type.Boolean(),
  domain_separator_hash: Type.Optional(Type.Undefined()),
  message_hash: Type.Optional(Type.Undefined()),
  show_message_hash: Type.Optional(Type.Boolean())
});
export const EthereumSignTypedHash = Type.Object({
  path: DerivationPath,
  data: EthereumSignTypedDataMessage,
  metamask_v4_compat: Type.Boolean(),
  domain_separator_hash: Type.String(),
  message_hash: Type.Optional(Type.String()),
  show_message_hash: Type.Optional(Type.Boolean())
});
export const EthereumVerifyMessage = Type.Object({
  address: Type.String(),
  message: Type.String(),
  hex: Type.Optional(Type.Boolean()),
  signature: Type.String()
});
