// @trezor/connect-common/libESM/types/api/cardanoComposeTransaction.mjs
import { MessagesSchema as PROTO } from "@trezor/protobuf";
import { Type } from "@trezor/schema-utils";
import { DerivationPath } from "../params.mjs";
import { CardanoCertificate, CardanoCertificatePointer } from "./cardano/index.mjs";
export const AccountAddress = Type.Object({
  address: Type.String(),
  path: Type.String(),
  transfers: Type.Number(),
  balance: Type.Optional(Type.String()),
  sent: Type.Optional(Type.String()),
  received: Type.Optional(Type.String())
}, {
  $id: 'AccountAddress'
});
export const CardanoComposeTransactionParamsSchema = Type.Object({
  account: Type.Object({
    descriptor: Type.String(),
    utxo: Type.Array(Type.Object({
      txid: Type.String(),
      vout: Type.Number(),
      amount: Type.String(),
      blockHeight: Type.Number(),
      address: Type.String(),
      path: Type.String(),
      confirmations: Type.Number(),
      coinbase: Type.Optional(Type.Boolean()),
      cardanoSpecific: Type.Optional(Type.Object({
        unit: Type.String()
      }))
    }, {
      $id: 'AccountUtxo'
    }))
  }),
  feeLevels: Type.Optional(Type.Array(Type.Object({
    feePerUnit: Type.Optional(Type.String())
  }))),
  outputs: Type.Optional(Type.Array(Type.Intersect([Type.Object({
    isChange: Type.Optional(Type.Boolean()),
    assets: Type.Array(Type.Object({
      unit: Type.String(),
      quantity: Type.String()
    }))
  }, {
    $id: 'BaseOutput'
  }), Type.Union([Type.Object({
    address: Type.String(),
    amount: Type.String(),
    setMax: Type.Optional(Type.Literal(false))
  }, {
    $id: 'ExternalOutput'
  }), Type.Object({
    address: Type.Optional(Type.String()),
    amount: Type.Optional(Type.String()),
    setMax: Type.Boolean()
  }, {
    $id: 'ExternalOutputIncomplete'
  })])], {
    $id: 'UserOutput'
  }))),
  certificates: Type.Optional(Type.Array(CardanoCertificate)),
  withdrawals: Type.Optional(Type.Array(Type.Object({
    stakeAddress: Type.String(),
    amount: Type.String()
  }, {
    $id: 'Withdrawal'
  }))),
  changeAddress: Type.Object({
    address: Type.String(),
    path: Type.String()
  }),
  addressParameters: Type.Object({
    addressType: PROTO.EnumCardanoAddressType,
    path: DerivationPath,
    stakingPath: Type.Optional(DerivationPath),
    stakingKeyHash: Type.Optional(Type.String()),
    certificatePointer: Type.Optional(CardanoCertificatePointer)
  }, {
    $id: 'CardanoAddressParameters'
  }),
  testnet: Type.Optional(Type.Boolean())
});
const _params = {};
const _paramsOld = {};
[_params, _paramsOld];
