// @trezor/connect-common/libESM/types/api/getPublicKey.mjs
import { MessagesSchema as PROTO } from "@trezor/protobuf";
import { Type } from "@trezor/schema-utils";
import { GetPublicKey as GetPublicKeyShared, PublicKey } from "../params.mjs";
export const GetPublicKey = Type.Intersect([GetPublicKeyShared, Type.Object({
  coin: Type.Optional(Type.String({
    description: 'determines network definition specified in coins.json file. Coin shortcut, name or label can be used. If coin is not set API will try to get network definition from path.',
    default: 'btc'
  })),
  crossChain: Type.Optional(Type.Boolean()),
  scriptType: Type.Optional(PROTO.InternalInputScriptType),
  ignoreXpubMagic: Type.Optional(Type.Boolean()),
  ecdsaCurveName: Type.Optional(Type.String()),
  unlockPath: Type.Optional(PROTO.UnlockPath)
})]);
export const HDNodeResponse = Type.Intersect([PublicKey, Type.Object({
  childNum: Type.Number(),
  xpub: Type.String(),
  xpubSegwit: Type.Optional(Type.String()),
  descriptorChecksum: Type.Optional(Type.String()),
  chainCode: Type.String(),
  fingerprint: Type.Number(),
  rootFingerprint: Type.Optional(Type.Number()),
  depth: Type.Number(),
  descriptor: Type.Optional(Type.String())
})]);
