// @trezor/connect-common/libESM/types/api/index.mjs
import { Type } from "@trezor/schema-utils";
export const TrezorConnectManagement = Type.Object({
  dispose: Type.Unsafe(),
  cancel: Type.Unsafe(),
  on: Type.Unsafe(),
  off: Type.Unsafe(),
  removeAllListeners: Type.Unsafe(),
  uiResponse: Type.Unsafe(),
  updateConnectSettings: Type.Unsafe()
});
export const TrezorConnectDevice = Type.Object({
  getFeatures: Type.Unsafe(),
  getDeviceState: Type.Unsafe(),
  getFirmwareHash: Type.Unsafe(),
  firmwareUpdate: Type.Unsafe(),
  resetDevice: Type.Unsafe(),
  loadDevice: Type.Unsafe(),
  recoveryDevice: Type.Unsafe(),
  wipeDevice: Type.Unsafe(),
  backupDevice: Type.Unsafe(),
  changePin: Type.Unsafe(),
  changeWipeCode: Type.Unsafe(),
  changeLanguage: Type.Unsafe(),
  applySettings: Type.Unsafe(),
  applyFlags: Type.Unsafe(),
  authenticateDevice: Type.Unsafe(),
  setBusy: Type.Unsafe(),
  setBrightness: Type.Unsafe(),
  showDeviceTutorial: Type.Unsafe(),
  bleUnpair: Type.Unsafe(),
  requestLogin: Type.Unsafe(),
  cipherKeyValue: Type.Unsafe(),
  unlockPath: Type.Unsafe(),
  getOwnershipId: Type.Unsafe(),
  getOwnershipProof: Type.Unsafe(),
  thpGetCredentials: Type.Unsafe(),
  thpRemoveCredentials: Type.Unsafe(),
  telemetryGet: Type.Unsafe(),
  pingDevice: Type.Unsafe()
});
export const TrezorConnectBlockchain = Type.Object({
  blockchainSubscribe: Type.Unsafe(),
  blockchainUnsubscribe: Type.Unsafe(),
  blockchainDisconnect: Type.Unsafe(),
  blockchainSetCustomBackend: Type.Unsafe(),
  blockchainGetInfo: Type.Unsafe(),
  blockchainValidateEvmRpcUrl: Type.Unsafe(),
  blockchainEstimateFee: Type.Unsafe(),
  blockchainGetAccountBalanceHistory: Type.Unsafe(),
  blockchainGetTransactions: Type.Unsafe(),
  blockchainEvmRpcCall: Type.Unsafe(),
  blockchainGetCurrentFiatRates: Type.Unsafe(),
  blockchainGetContractInfo: Type.Unsafe(),
  blockchainGetFiatRatesForTimestamps: Type.Unsafe(),
  blockchainSubscribeFiatRates: Type.Unsafe(),
  blockchainUnsubscribeFiatRates: Type.Unsafe(),
  pushTransaction: Type.Unsafe(),
  getNonce: Type.Unsafe()
});
export const TrezorConnectAccount = Type.Object({
  getAddress: Type.Unsafe(),
  getPublicKey: Type.Unsafe(),
  getAccountInfo: Type.Unsafe(),
  getAccountDescriptor: Type.Unsafe(),
  discoverAccounts: Type.Unsafe(),
  signMessage: Type.Unsafe(),
  verifyMessage: Type.Unsafe(),
  getSettings: Type.Unsafe(),
  getCoinInfo: Type.Unsafe()
});
export const TrezorConnectBitcoin = Type.Object({
  signTransaction: Type.Unsafe(),
  composeTransaction: Type.Unsafe(),
  authorizeCoinjoin: Type.Unsafe(),
  cancelCoinjoinAuthorization: Type.Unsafe()
});
export const TrezorConnectEthereum = Type.Object({
  ethereumGetAddress: Type.Unsafe(),
  ethereumGetPublicKey: Type.Unsafe(),
  ethereumSignTransaction: Type.Unsafe(),
  ethereumSignMessage: Type.Unsafe(),
  ethereumSignTypedData: Type.Unsafe(),
  ethereumVerifyMessage: Type.Unsafe()
});
export const TrezorConnectCardano = Type.Object({
  cardanoGetAddress: Type.Unsafe(),
  cardanoGetPublicKey: Type.Unsafe(),
  cardanoGetNativeScriptHash: Type.Unsafe(),
  cardanoSignTransaction: Type.Unsafe(),
  cardanoSignMessage: Type.Unsafe(),
  cardanoComposeTransaction: Type.Unsafe()
});
export const TrezorConnectMonero = Type.Object({
  moneroGetAddress: Type.Unsafe(),
  moneroGetWatchKey: Type.Unsafe(),
  moneroKeyImageSync: Type.Unsafe(),
  moneroSignTransaction: Type.Unsafe()
});
export const TrezorConnectRipple = Type.Object({
  rippleGetAddress: Type.Unsafe(),
  rippleSignTransaction: Type.Unsafe()
});
export const TrezorConnectSolana = Type.Object({
  solanaGetAddress: Type.Unsafe(),
  solanaGetPublicKey: Type.Unsafe(),
  solanaSignTransaction: Type.Unsafe(),
  solanaComposeTransaction: Type.Unsafe()
});
export const TrezorConnectStellar = Type.Object({
  stellarGetAddress: Type.Unsafe(),
  stellarSignTransaction: Type.Unsafe()
});
export const TrezorConnectTezos = Type.Object({
  tezosGetAddress: Type.Unsafe(),
  tezosGetPublicKey: Type.Unsafe(),
  tezosSignTransaction: Type.Unsafe()
});
export const TrezorConnectTron = Type.Object({
  tronGetAddress: Type.Unsafe(),
  tronSignTransaction: Type.Unsafe(),
  tronComposeTransaction: Type.Unsafe()
});
export const TrezorConnectEvolu = Type.Object({
  evoluGetNode: Type.Unsafe(),
  evoluSignRegistrationRequest: Type.Unsafe(),
  evoluGetDelegatedIdentityKey: Type.Unsafe()
});
export const TrezorConnectSchema = Type.Composite([TrezorConnectManagement, TrezorConnectDevice, TrezorConnectBlockchain, TrezorConnectAccount, TrezorConnectBitcoin, TrezorConnectEthereum, TrezorConnectCardano, TrezorConnectMonero, TrezorConnectRipple, TrezorConnectSolana, TrezorConnectStellar, TrezorConnectTezos, TrezorConnectTron, TrezorConnectEvolu]);
