// @trezor/connect-common/libESM/types/api/tezos/index.mjs
import { Type } from "@trezor/schema-utils";
import { DerivationPath } from "../../params.mjs";
export const TezosRevealOperation = Type.Object({
  source: Type.String(),
  fee: Type.Number(),
  counter: Type.Number(),
  gas_limit: Type.Number(),
  storage_limit: Type.Number(),
  public_key: Type.String()
});
export const TezosManagerTransfer = Type.Object({
  destination: Type.String(),
  amount: Type.Number()
});
export const TezosParametersManager = Type.Object({
  set_delegate: Type.Optional(Type.String()),
  cancel_delegate: Type.Optional(Type.Boolean()),
  transfer: Type.Optional(TezosManagerTransfer)
});
export const TezosTransactionOperation = Type.Object({
  source: Type.String(),
  destination: Type.String(),
  amount: Type.Number(),
  counter: Type.Number(),
  fee: Type.Number(),
  gas_limit: Type.Number(),
  storage_limit: Type.Number(),
  parameters: Type.Optional(Type.Array(Type.Number())),
  parameters_manager: Type.Optional(TezosParametersManager)
});
export const TezosOriginationOperation = Type.Object({
  source: Type.String(),
  balance: Type.Number(),
  delegate: Type.Optional(Type.String()),
  script: DerivationPath,
  fee: Type.Number(),
  counter: Type.Number(),
  gas_limit: Type.Number(),
  storage_limit: Type.Number()
});
export const TezosDelegationOperation = Type.Object({
  source: Type.String(),
  delegate: Type.String(),
  fee: Type.Number(),
  counter: Type.Number(),
  gas_limit: Type.Number(),
  storage_limit: Type.Number()
});
export const TezosOperation = Type.Object({
  reveal: Type.Optional(TezosRevealOperation),
  transaction: Type.Optional(TezosTransactionOperation),
  origination: Type.Optional(TezosOriginationOperation),
  delegation: Type.Optional(TezosDelegationOperation)
});
export const TezosSignTransaction = Type.Object({
  path: DerivationPath,
  branch: Type.String(),
  operation: TezosOperation,
  chunkify: Type.Optional(Type.Boolean())
});
