// @trezor/connect-common/libESM/messageChannel/abstract.mjs
import { TypedEmitter, createDeferred, createDeferredManager, scheduleAction } from "@trezor/utils";
export class AbstractMessageChannel extends TypedEmitter {
  messages = createDeferredManager({
    generateId: () => crypto.randomUUID()
  });
  messagesQueue = [];
  isConnected = false;
  handshakeMaxRetries = 5;
  handshakeRetryInterval = 2000;
  handshakeFinished;
  lazyHandshake;
  logger;
  sendFn;
  channel;
  constructor({
    sendFn,
    channel,
    logger,
    lazyHandshake = false
  }) {
    super();
    this.channel = channel;
    this.sendFn = sendFn;
    this.lazyHandshake = lazyHandshake;
    this.logger = logger;
  }
  init() {
    if (!this.handshakeFinished) {
      this.handshakeFinished = createDeferred();
      if (!this.lazyHandshake) {
        this.handshakeWithPeer();
      }
    }
    return this.handshakeFinished.promise;
  }
  handshakeWithPeer() {
    this.logger?.log(this.channel.here, 'handshake');
    return scheduleAction(async () => {
      this.postMessage({
        type: 'channel-handshake-request',
        data: {
          success: true,
          payload: undefined
        }
      }, {
        usePromise: false,
        useQueue: false
      });
      await this.handshakeFinished?.promise;
    }, {
      attempts: this.handshakeMaxRetries,
      timeout: this.handshakeRetryInterval
    }).then(() => {
      this.logger?.log(this.channel.here, 'handshake confirmed');
      this.messagesQueue.forEach(message => {
        message.channel = this.channel;
        this.sendFn(message);
      });
      this.messagesQueue = [];
    }).catch(() => {
      this.handshakeFinished?.reject(new Error('handshake failed'));
      this.handshakeFinished = undefined;
    });
  }
  onMessage(_message) {
    const message = _message;
    const {
      channel,
      id,
      type,
      ...data
    } = message;
    if (!channel?.peer || channel.peer !== this.channel.here) {
      return;
    }
    if (!channel?.here || this.channel.peer !== channel.here) {
      return;
    }
    if (type === 'channel-handshake-request') {
      this.postMessage({
        type: 'channel-handshake-confirm',
        data: {
          success: true,
          payload: undefined
        }
      }, {
        usePromise: false,
        useQueue: false
      });
      if (this.lazyHandshake) {
        this.handshakeWithPeer();
      }
      return;
    }
    if (type === 'channel-handshake-confirm') {
      this.handshakeFinished?.resolve(undefined);
      return;
    }
    this.messages.resolve(id, {
      id,
      ...data
    });
    const messagePromisesLength = this.messages.length();
    if (messagePromisesLength > 5) {
      this.logger?.warn(`too many message promises (${messagePromisesLength}). this feels unexpected!`);
    }
    this.emit('message', message);
  }
  postMessage(message, {
    usePromise = true,
    useQueue = true
  } = {}) {
    message.channel = this.channel;
    if (!usePromise) {
      try {
        this.sendFn(message);
      } catch {
        if (useQueue) {
          this.messagesQueue.push(message);
        }
      }
      return;
    }
    const {
      promise,
      promiseId
    } = this.messages.create();
    message.id = promiseId;
    try {
      this.sendFn(message);
    } catch {
      if (useQueue) {
        this.messagesQueue.push(message);
      }
    }
    return promise;
  }
  resolveMessagePromises(payload) {
    this.messages.resolveAll(id => ({
      id,
      payload
    }));
  }
  clearPendingSends() {}
  abortHandshake(reason) {
    this.handshakeFinished?.reject(new Error(reason ?? 'Handshake aborted'));
    this.handshakeFinished = undefined;
  }
  clear() {
    this.handshakeFinished = undefined;
  }
}
