// @trezor/connect-common/libESM/messageChannel/serviceworker-window-ext-connectable.mjs
import { AbstractMessageChannel } from "./abstract.mjs";
export class ServiceWorkerWindowExtConnectableChannel extends AbstractMessageChannel {
  currentId;
  allowedOrigin;
  messageListener;
  sendChain = Promise.resolve();
  static SEND_SETTLE_MS = 100;
  constructor({
    channel,
    logger,
    currentId,
    allowedOrigin
  }) {
    super({
      channel,
      sendFn: message => {
        this.sendChain = this.sendChain.catch(() => {}).then(() => this.doSend(message)).then(() => new Promise(resolve => setTimeout(resolve, ServiceWorkerWindowExtConnectableChannel.SEND_SETTLE_MS)));
      },
      logger
    });
    this.currentId = currentId;
    this.allowedOrigin = allowedOrigin;
  }
  async doSend(message) {
    try {
      const tabId = await this.currentId?.();
      this.logger?.debug('CHANNEL sending to tab:', tabId, 'message:', message);
      if (!tabId) {
        this.logger?.error('CHANNEL: No tab ID available, cannot send message');
        throw new Error('No tab ID available');
      }
      if (!message.id) {
        this.logger?.warn('CHANNEL: Message without ID, this might cause issues with response handling', message);
      }
      const tab = await Promise.race([new Promise((resolve, reject) => {
        chrome.tabs.get(tabId, t => {
          if (chrome.runtime.lastError) {
            this.logger?.error('CHANNEL: Error getting tab:', chrome.runtime.lastError);
            reject(chrome.runtime.lastError);
          } else {
            resolve(t);
          }
        });
      }), new Promise((_, reject) => setTimeout(() => reject(new Error('chrome.tabs.get timeout after 5s')), 5000))]);
      if (!tab?.url) {
        throw new Error('Trezor suite tab URL not available, the manifest must declare host_permissions matching the popup origin');
      }
      const currentUrl = new URL(tab.url);
      const encodedMessage = encodeURIComponent(JSON.stringify(message));
      currentUrl.hash = `#extension-id=${chrome.runtime.id}&message=${encodedMessage}`;
      chrome.tabs.update(tabId, {
        url: currentUrl.toString()
      });
    } catch (error) {
      this.logger?.error('CHANNEL: Error sending message via tab update:', error);
      throw error;
    }
  }
  clearPendingSends() {
    this.sendChain = Promise.resolve();
  }
  connect() {
    if (this.messageListener) {
      this.logger?.debug('CHANNEL: Listener already registered, skipping');
      return;
    }
    this.logger?.debug('CHANNEL: connect called, setting up onMessageExternal listener');
    this.messageListener = (message, sender, sendResponse) => {
      this.logger?.debug('CHANNEL: onMessageExternal fired:', {
        message,
        sender
      });
      if (!sender.tab) {
        this.logger?.warn('CHANNEL: Received message not from a tab, ignoring', sender);
        return false;
      }
      if (this.allowedOrigin) {
        if (!sender.tab.url) {
          this.logger?.warn(`CHANNEL: Missing sender tab URL while allowed origin is configured (allowed=${this.allowedOrigin}). Ignoring.`);
          return false;
        }
        let senderOrigin;
        let expectedOrigin;
        try {
          senderOrigin = new URL(sender.tab.url).origin;
          expectedOrigin = new URL(this.allowedOrigin).origin;
        } catch {
          this.logger?.warn(`CHANNEL: Could not parse sender or allowed origin (sender=${sender.tab.url}, allowed=${this.allowedOrigin}). Ignoring.`);
          return false;
        }
        if (senderOrigin !== expectedOrigin) {
          this.logger?.warn(`CHANNEL: Sender origin mismatch — expected ${expectedOrigin}, got ${senderOrigin}. Ignoring.`);
          return false;
        }
      }
      if (!message.channel) {
        this.logger?.warn('CHANNEL: Received message without channel info, ignoring');
        sendResponse({
          error: 'Missing channel info'
        });
        return false;
      }
      this.logger?.debug('CHANNEL: Processing message through onMessage');
      this.onMessage(message);
    };
    chrome.runtime.onMessageExternal.addListener(this.messageListener);
    this.isConnected = true;
    this.logger?.debug('CHANNEL: Listener registered successfully');
  }
  disconnect() {
    if (!this.isConnected) return;
    if (this.messageListener) {
      chrome.runtime.onMessageExternal.removeListener(this.messageListener);
      this.messageListener = undefined;
    }
    this.clear();
    this.isConnected = false;
  }
}
