// @trezor/connect-common/libESM/messageChannel/serviceworker-window.mjs
import { AbstractMessageChannel } from "./abstract.mjs";
export class ServiceWorkerWindowChannel extends AbstractMessageChannel {
  port;
  name;
  allowSelfOrigin;
  currentId;
  constructor({
    name,
    channel,
    logger,
    lazyHandshake,
    allowSelfOrigin = false,
    currentId
  }) {
    super({
      channel,
      sendFn: message => {
        if (!this.port) throw new Error('port not assigned');
        this.port.postMessage(message);
      },
      logger,
      lazyHandshake
    });
    this.name = name;
    this.allowSelfOrigin = allowSelfOrigin;
    this.currentId = currentId;
    this.connect();
  }
  connect() {
    chrome.runtime.onConnect.addListener(port => {
      if (port.name !== this.name) return;
      if (this.currentId?.() && this.currentId?.() !== port.sender?.tab?.id) return;
      this.port = port;
      this.port.onMessage.addListener((message, {
        sender
      }) => {
        if (!sender) {
          this.logger?.error('service-worker-window', 'no sender');
          return;
        }
        const {
          origin
        } = sender;
        const whitelist = ['https://connect.trezor.io', 'https://staging-connect.trezor.io', 'https://suite.corp.sldev.cz', 'https://dev.suite.sldev.cz', 'http://localhost:8088', 'https://suite.trezor.io', 'https://staging-suite.trezor.io', 'http://localhost:8000'];
        const webextensionId = chrome?.runtime?.id;
        if (webextensionId) {
          whitelist.push(`chrome-extension://${webextensionId}`);
        }
        const webextensionUrl = chrome?.runtime?.getURL('');
        if (webextensionUrl) {
          whitelist.push(webextensionUrl.slice(0, -1));
        }
        if (!origin) {
          this.logger?.error('connect-webextension/messageChannel/extensionPort/onMessage', 'no origin');
          return;
        }
        if (!whitelist.includes(origin)) {
          this.logger?.error('connect-webextension/messageChannel/extensionPort/onMessage', 'origin not whitelisted', origin);
          return;
        }
        if (origin === self.origin && !this.allowSelfOrigin) {
          return;
        }
        this.onMessage(message);
      });
    });
    this.isConnected = true;
  }
  disconnect() {
    if (!this.isConnected) return;
    this.port?.disconnect();
    this.clear();
    this.isConnected = false;
  }
}
