// @trezor/connect-common/libESM/messageChannel/window-window.mjs
import { AbstractMessageChannel } from "./abstract.mjs";
export class WindowWindowChannel extends AbstractMessageChannel {
  _windowHere;
  _listener;
  _origin;
  constructor({
    windowHere,
    windowPeer,
    channel,
    logger,
    origin
  }) {
    super({
      channel,
      sendFn: message => {
        windowPeer()?.postMessage(message, origin);
      },
      logger
    });
    this._listener = this.listener.bind(this);
    this._windowHere = windowHere;
    this._origin = origin;
    this.connect();
  }
  listener(event) {
    if (event.origin !== this._origin || event.origin === 'null') {
      this.logger?.warn(`WindowWindowChannel: ignoring message from unexpected origin "${event.origin}", expected "${this._origin}"`);
      return;
    }
    const message = {
      ...event.data,
      success: true,
      origin: event.origin,
      payload: event.data.payload || {},
      channel: event.data.channel || {
        peer: this.channel.here,
        here: this.channel.peer
      }
    };
    this.onMessage(message);
  }
  connect() {
    this._windowHere.addEventListener('message', this._listener);
    this.isConnected = true;
  }
  disconnect() {
    if (!this.isConnected) return;
    this._windowHere.removeEventListener('message', this._listener);
    this.isConnected = false;
  }
}
