// @trezor/connect-common/libESM/types/coinInfo.mjs
import { DeviceModelInternal } from "@trezor/device-utils";
import { Type } from "@trezor/schema-utils";
import { FeeLevel } from "./fees.mjs";
export const Bip32 = Type.Object({
  public: Type.Number(),
  private: Type.Number()
});
export const Network = Type.Object({
  messagePrefix: Type.String(),
  bech32: Type.String(),
  bip32: Bip32,
  pubKeyHash: Type.Number(),
  scriptHash: Type.Number(),
  wif: Type.Number(),
  forkId: Type.Optional(Type.Number())
});
export const CoinObj = Type.Object({
  coin: Type.String(),
  identity: Type.Optional(Type.String())
});
export const CoinSupport = Type.Record(Type.KeyOfEnum(DeviceModelInternal), Type.Union([Type.String(), Type.Literal(false)]));
export const BlockchainLink = Type.Object({
  type: Type.String(),
  url: Type.Array(Type.String())
});
const Common = Type.Object({
  label: Type.String(),
  name: Type.String(),
  shortcut: Type.String(),
  slip44: Type.Number(),
  support: CoinSupport,
  decimals: Type.Number(),
  blockchainLink: Type.Optional(BlockchainLink),
  blockTime: Type.Number(),
  minFee: Type.Number(),
  maxFee: Type.Number(),
  minPriorityFee: Type.Number(),
  defaultFees: Type.Array(FeeLevel)
});
export const BitcoinNetworkInfo = Type.Intersect([Common, Type.Object({
  type: Type.Literal('bitcoin'),
  cashAddrPrefix: Type.Optional(Type.String()),
  curveName: Type.String(),
  dustLimit: Type.Number(),
  forceBip143: Type.Boolean(),
  hashGenesisBlock: Type.String(),
  maxAddressLength: Type.Number(),
  maxFeeSatoshiKb: Type.Number(),
  minAddressLength: Type.Number(),
  minFeeSatoshiKb: Type.Number(),
  segwit: Type.Boolean(),
  xPubMagic: Type.Number(),
  xPubMagicSegwitNative: Type.Optional(Type.Number()),
  xPubMagicSegwit: Type.Optional(Type.Number()),
  taproot: Type.Optional(Type.Boolean()),
  network: Network,
  isBitcoin: Type.Boolean()
})]);
export const EthereumNetworkInfo = Type.Intersect([Common, Type.Object({
  type: Type.Literal('ethereum'),
  chainId: Type.Number(),
  network: Type.Optional(Type.Undefined())
})]);
export const EthereumNetworkInfoDefinitionValues = Type.Omit(EthereumNetworkInfo, ['minFee', 'maxFee', 'defaultFees', 'minPriorityFee', 'blockTime']);
export const MiscNetworkInfo = Type.Intersect([Common, Type.Object({
  type: Type.Literal('misc'),
  curve: Type.String(),
  network: Type.Optional(Type.Undefined())
})]);
export const CoinInfo = Type.Union([BitcoinNetworkInfo, EthereumNetworkInfo, MiscNetworkInfo]);
