// @trezor/connect-common/libESM/types/fees.mjs
import { Type } from "@trezor/schema-utils";
export const FeeInfo = Type.Object({
  blockTime: Type.Number(),
  minFee: Type.Number(),
  maxFee: Type.Number(),
  minPriorityFee: Type.Number(),
  dustLimit: Type.Number()
});
export const FeeLevel = Type.Object({
  label: Type.Union([Type.Literal('high'), Type.Literal('normal'), Type.Literal('economy'), Type.Literal('low'), Type.Literal('custom')]),
  blocks: Type.Number(),
  feePerUnit: Type.String(),
  feePerTx: Type.Optional(Type.String()),
  feeLimit: Type.Optional(Type.String()),
  baseFeePerGas: Type.Optional(Type.String()),
  maxFeePerGas: Type.Optional(Type.String()),
  maxPriorityFeePerGas: Type.Optional(Type.String())
});
export const SelectFeeLevel = Type.Union([Type.Object({
  name: Type.String(),
  fee: Type.Literal('0'),
  feePerByte: Type.Optional(Type.Undefined()),
  blocks: Type.Optional(Type.Undefined()),
  disabled: Type.Literal(true)
}), Type.Object({
  name: Type.String(),
  fee: Type.String(),
  feePerByte: Type.String(),
  blocks: Type.Number(),
  minutes: Type.Number(),
  total: Type.String()
})]);
