// @trezor/connect-common/libESM/impl/dynamic.mjs
import { getSynchronize } from "@trezor/utils";
import { ERRORS } from "../constants/index.mjs";
import { parseManifest, parseVersion } from "../data/connectSettings.mjs";
import { createErrorMessage } from "../events/index.mjs";
import { ConnectEmitter } from "../types/emitter.mjs";
export class TrezorConnectDynamic {
  eventEmitter = new ConnectEmitter();
  currentTarget;
  implementations;
  coreMode;
  implSettings;
  callPending = 0;
  beforeCallSynchronize = getSynchronize();
  constructor({
    implementations
  }) {
    this.implementations = implementations;
    this.currentTarget = 'core-in-suite-desktop';
  }
  getTarget() {
    return this.implementations[this.currentTarget];
  }
  getTargetType() {
    return this.currentTarget;
  }
  async switchTarget(target) {
    if (this.currentTarget === target) {
      return;
    }
    if (!this.implSettings) {
      throw ERRORS.TypedError('Init_ManifestMissing');
    }
    const oldTargetType = this.getTargetType();
    const oldTarget = this.getTarget();
    try {
      this.currentTarget = target;
      await this.getTarget().init(this.implSettings);
      await oldTarget.dispose();
    } catch {
      this.currentTarget = oldTargetType;
    }
  }
  async init(settings) {
    this.coreMode = settings.coreMode;
    const manifest = parseManifest(settings.manifest);
    if (!manifest) {
      throw ERRORS.TypedError('Init_ManifestMissing');
    }
    this.implSettings = {
      manifest,
      env: settings.env,
      debug: settings.debug,
      version: parseVersion(settings.version)
    };
    this.currentTarget = this.getInitTarget();
    this.callPending = 0;
    try {
      return await this.getTarget().init(this.implSettings);
    } catch (error) {
      if (await this.handleErrorFallback(error.code)) {
        return;
      }
      throw error;
    }
  }
  updateConnectSettings(_params) {
    return Promise.resolve(createErrorMessage(ERRORS.TypedError('Method_InvalidPackage', 'updateConnectSettings is not supported in this implementation')));
  }
  async call(params) {
    try {
      if (this.callPending === 0) {
        await this.beforeCallSynchronize(async () => {
          this.callPending++;
          await this.handleBeforeCall();
        });
      }
      const response = await this.getTarget().call(params);
      if (!response.success) {
        if (await this.handleErrorFallback(response.error.code)) {
          return await this.getTarget().call(params);
        }
      }
      return response;
    } catch (error) {
      return createErrorMessage(error);
    } finally {
      this.callPending--;
    }
  }
  cancel(error) {
    return this.getTarget().cancel(error);
  }
  dispose() {
    this.eventEmitter.removeAllListeners();
    this.callPending = 0;
    return this.getTarget().dispose();
  }
  getInitTarget() {
    const {
      coreMode
    } = this;
    if (coreMode === 'suite-desktop') {
      return 'core-in-suite-desktop';
    } else if (coreMode === 'suite-web') {
      return 'core-in-suite-web';
    } else {
      if (coreMode && coreMode !== 'auto') {
        console.warn(`Invalid coreMode: ${coreMode}`);
      }
      return 'core-in-suite-desktop';
    }
  }
  async handleBeforeCall() {
    const {
      coreMode
    } = this;
    if (coreMode === 'suite-desktop' || coreMode === 'auto' || coreMode === undefined) {
      await this.switchTarget('core-in-suite-desktop');
    }
  }
  async handleErrorFallback(errorCode) {
    if (this.getTargetType() === 'core-in-suite-desktop' && (errorCode === 'Desktop_ConnectionMissing' || errorCode === 'Method_Unsupported')) {
      await this.switchTarget('core-in-suite-web');
      return true;
    }
    return false;
  }
  uiResponse() {
    throw ERRORS.TypedError('Method_InvalidPackage');
  }
}
