// @trezor/connect-web/libESM/bootstrap/bootstrap.mjs
import { BootstrapError } from "./bootstrap-errors.mjs";
const logger = (() => {
  let enabled = false;
  return {
    enable: () => {
      enabled = true;
    },
    log: (...args) => {
      if (enabled) console.log(...args);
    },
    error: (...args) => {
      console.error(...args);
    }
  };
})();
const IFRAME_MODE = window.parent !== window;
const BOOTSTRAP_IFRAME = '@trezor/connect-bootstrap-iframe';
const BOOTSTRAP_POPUP = '@trezor/connect-bootstrap-popup';
const PEERS = {
  HERE: IFRAME_MODE ? BOOTSTRAP_IFRAME : BOOTSTRAP_POPUP,
  BOOTSTRAP: IFRAME_MODE ? BOOTSTRAP_POPUP : BOOTSTRAP_IFRAME,
  WEB: '@trezor/connect-web',
  POPUP: '@trezor/connect-popup'
};
const HANDSHAKE_TIMEOUT = 500;
const HANDSHAKE_REQ = 'channel-handshake-request';
const HANDSHAKE_ERR = 'channel-handshake-error';
const HANDSHAKE_CONF = 'channel-handshake-confirm';
const getParams = () => {
  if (typeof URLSearchParams === 'undefined') {
    throw BootstrapError.ENV_NOT_SUPPORTED;
  }
  try {
    const urlParams = new URLSearchParams(window.location.href.split('?')[1]);
    const debug = urlParams.get('debug');
    if (debug === '1' || debug === 'true') logger.enable();
    const channelId = urlParams.get('connect-popup-req');
    const channelName = channelId ? '@trezor/connect-popup/' + channelId : null;
    return {
      channelId,
      channelName,
      debug
    };
  } catch {
    throw BootstrapError.ENV_NOT_SUPPORTED;
  }
};
const getUnpartitionedBroadcastChannel = async channelName => {
  let handle = null;
  try {
    handle = await document.requestStorageAccess({
      BroadcastChannel: true
    });
  } catch {
    throw BootstrapError.STORAGE_ACCESS_DENIED;
  }
  if (!handle || typeof handle.BroadcastChannel !== 'function') {
    throw BootstrapError.ENV_NOT_SUPPORTED;
  }
  try {
    return handle.BroadcastChannel(channelName);
  } catch {
    throw BootstrapError.ENV_NOT_SUPPORTED;
  }
};
const getBroadcastChannel = channelName => {
  if (typeof BroadcastChannel === 'undefined') {
    throw BootstrapError.ENV_NOT_SUPPORTED;
  }
  try {
    return new BroadcastChannel(channelName);
  } catch {
    throw BootstrapError.ENV_NOT_SUPPORTED;
  }
};
const getConfirmHandshake = eventData => ({
  type: HANDSHAKE_CONF,
  channel: {
    here: PEERS.HERE,
    peer: eventData.channel.here
  }
});
const redirectToSuite = errorCode => {
  const currentUrl = new URL(window.location.href);
  const nextUrl = new URL(currentUrl.toString());
  nextUrl.pathname = nextUrl.pathname.replace(/\/bootstrap\.html$/, '/');
  if (errorCode) {
    nextUrl.search = '';
    nextUrl.searchParams.set('connect-popup-err', errorCode);
  } else {
    nextUrl.search = currentUrl.search;
  }
  window.location.href = nextUrl.toString();
};
const handleBootstrapHandshake = (broadcast, timeout = 5000) => new Promise((resolve, reject) => {
  const onHandshakeRequest = event => {
    const channel = event.data?.channel;
    if (!channel) return;
    if (event.data.type === HANDSHAKE_REQ && channel.here === PEERS.BOOTSTRAP && channel.peer === PEERS.HERE) {
      clearTimeout(timer);
      broadcast.removeEventListener('message', onHandshakeRequest);
      broadcast.postMessage(getConfirmHandshake(event.data));
      resolve();
    }
  };
  broadcast.addEventListener('message', onHandshakeRequest);
  const timer = setTimeout(() => {
    broadcast.removeEventListener('message', onHandshakeRequest);
    reject(BootstrapError.HANDSHAKE_TIMEOUT);
  }, timeout);
});
const bootstrapPopup = async () => {
  let channelName;
  try {
    const params = getParams();
    channelName = params.channelName;
  } catch (error) {
    redirectToSuite(error);
    return;
  }
  logger.log(PEERS.HERE, channelName);
  if (!channelName) {
    redirectToSuite(BootstrapError.CHANNEL_ID_MISSING);
    return;
  }
  window.addEventListener('message', event => {
    if (event.data?.type === HANDSHAKE_ERR) {
      redirectToSuite(event.data.error);
    }
  });
  try {
    const broadcast = getBroadcastChannel(channelName);
    await handleBootstrapHandshake(broadcast);
    broadcast.close();
    redirectToSuite();
  } catch (error) {
    redirectToSuite(typeof error === 'string' ? error : BootstrapError.UNKNOWN);
  }
};
const startForwarding = (broadcast, owner, ownerOrigin) => {
  logger.log(PEERS.HERE, 'start forwarding');
  const onBroadcastMessage = event => {
    const channel = event.data?.channel;
    if (!channel || channel.here !== PEERS.POPUP || channel.peer !== PEERS.WEB) {
      return;
    }
    logger.log(PEERS.POPUP, '→', PEERS.WEB, event.data.type);
    owner.postMessage(event.data, ownerOrigin);
  };
  const onOwnerMessage = event => {
    if (event.origin !== ownerOrigin) {
      return;
    }
    const channel = event.data?.channel;
    if (!channel || channel.here !== PEERS.WEB || channel.peer !== PEERS.POPUP) {
      return;
    }
    logger.log(PEERS.WEB, '→', PEERS.POPUP, event.data.type);
    broadcast.postMessage({
      ...event.data,
      origin: event.origin
    });
  };
  broadcast.addEventListener('message', onBroadcastMessage);
  window.addEventListener('message', onOwnerMessage);
  return () => {
    logger.log(PEERS.HERE, 'stop forwarding');
    broadcast.removeEventListener('message', onBroadcastMessage);
    window.removeEventListener('message', onOwnerMessage);
  };
};
const startBootstrapHandshake = (broadcast, maxAttempts = 5) => {
  const abortController = new AbortController();
  const promise = new Promise((resolve, reject) => {
    let attempt = 0;
    let timer;
    const tryOnce = () => {
      attempt++;
      logger.log(PEERS.HERE, 'handshake attempt', attempt, 'of', maxAttempts);
      const onHandshakeConfirm = event => {
        const channel = event.data?.channel;
        if (!channel) return;
        if (event.data.type === HANDSHAKE_CONF && channel.here === PEERS.BOOTSTRAP && channel.peer === PEERS.HERE) {
          cleanup();
          resolve();
        }
      };
      const cleanup = () => {
        clearTimeout(timer);
        broadcast.removeEventListener('message', onHandshakeConfirm);
      };
      abortController.signal.addEventListener('abort', () => {
        cleanup();
        reject(BootstrapError.HANDSHAKE_TIMEOUT);
      });
      broadcast.addEventListener('message', onHandshakeConfirm);
      broadcast.postMessage({
        type: HANDSHAKE_REQ,
        channel: {
          here: PEERS.HERE,
          peer: PEERS.BOOTSTRAP
        }
      });
      timer = setTimeout(() => {
        broadcast.removeEventListener('message', onHandshakeConfirm);
        if (attempt < maxAttempts) {
          tryOnce();
        } else {
          reject(BootstrapError.HANDSHAKE_TIMEOUT);
        }
      }, HANDSHAKE_TIMEOUT);
    };
    tryOnce();
  });
  return {
    abort: () => abortController.abort(),
    promise
  };
};
const bootstrapIframe = () => {
  let stopForwarding;
  let handshakeInProgress = false;
  const sendHandshakeError = (errorCode, ownerOrigin) => {
    handshakeInProgress = false;
    window.parent.postMessage({
      type: HANDSHAKE_ERR,
      error: typeof errorCode === 'string' ? errorCode : BootstrapError.UNKNOWN
    }, ownerOrigin);
  };
  return new Promise(resolve => {
    window.addEventListener('message', async event => {
      if (event.data.channel?.here !== PEERS.WEB || event.data.channel?.peer !== PEERS.HERE) {
        return;
      }
      if (event.data.type === HANDSHAKE_REQ) {
        if (handshakeInProgress) {
          logger.log(PEERS.HERE, 'handshake already in progress');
          return;
        }
        handshakeInProgress = true;
        const ownerOrigin = event.origin && event.origin !== 'null' ? event.origin : null;
        if (!ownerOrigin) {
          sendHandshakeError(BootstrapError.ORIGIN_MISSING, '*');
          return;
        }
        let channelName;
        try {
          const params = getParams();
          channelName = params.channelName;
        } catch (error) {
          sendHandshakeError(error, ownerOrigin);
          return;
        }
        logger.log(PEERS.HERE, channelName);
        if (!channelName) {
          sendHandshakeError(BootstrapError.CHANNEL_ID_MISSING, ownerOrigin);
          return;
        }
        let broadcast;
        try {
          broadcast = await getUnpartitionedBroadcastChannel(channelName);
        } catch (error) {
          sendHandshakeError(error, ownerOrigin);
          return;
        }
        const bootstrapHandshake = startBootstrapHandshake(broadcast);
        try {
          await bootstrapHandshake.promise;
          stopForwarding?.();
          stopForwarding = startForwarding(broadcast, window.parent, ownerOrigin);
          window.parent.postMessage(getConfirmHandshake(event.data), ownerOrigin);
        } catch (error) {
          logger.error(PEERS.BOOTSTRAP, error);
          bootstrapHandshake.abort();
          broadcast.close();
          sendHandshakeError(error, ownerOrigin);
        }
        handshakeInProgress = false;
        resolve();
      }
    });
  });
};
export const bootstrap = async () => {
  if (IFRAME_MODE) {
    await bootstrapIframe();
  } else {
    await bootstrapPopup();
  }
};
