// @trezor/protocol/libESM/protocol-thp/decode.mjs
import { CRC_LENGTH, TAG_LENGTH } from "./constants.mjs";
import { aesgcm } from "./crypto/index.mjs";
import { THP_CONTROL_BYTE } from "../protocol-v2/constants.mjs";
import { crc32 } from "./crypto/crc32.mjs";
import { getHandshakeHash, getTrezorState } from "./crypto/pairing.mjs";
import { getIvFromNonce } from "./crypto/tools.mjs";
const decipherMessage = (key, recvNonce, payload, tag) => {
  const aes = aesgcm(key, getIvFromNonce(recvNonce));
  aes.auth(Buffer.alloc(0));
  const trezorMaskedStaticPubkey = aes.decrypt(payload, tag);
  return trezorMaskedStaticPubkey.subarray(1);
};
const createChannelResponse = ({
  payload
}, protobufDecoder) => {
  const nonce = payload.subarray(0, 8);
  const channel = payload.subarray(8, 10);
  const props = payload.subarray(10, payload.length - CRC_LENGTH);
  const properties = protobufDecoder('ThpDeviceProperties', props).message;
  const handshakeHash = getHandshakeHash(props);
  return {
    type: 'ThpCreateChannelResponse',
    message: {
      nonce,
      channel,
      properties,
      handshakeHash
    }
  };
};
const readHandshakeInitResponse = ({
  payload
}) => {
  const trezorEphemeralPubkey = payload.subarray(0, 32);
  const trezorEncryptedStaticPubkey = payload.subarray(32, 32 + 48);
  const tag = payload.subarray(32 + 48, 32 + 48 + TAG_LENGTH);
  return {
    type: 'ThpHandshakeInitResponse',
    message: {
      trezorEphemeralPubkey,
      trezorEncryptedStaticPubkey,
      tag
    }
  };
};
const readHandshakeCompletionResponse = ({
  payload,
  thpState
}) => {
  const state = getTrezorState(thpState.handshakeCredentials, payload);
  return {
    type: 'ThpHandshakeCompletionResponse',
    message: {
      state
    }
  };
};
const readProtobufMessage = ({
  payload,
  thpState
}, protobufDecoder) => {
  const tagPos = payload.length - TAG_LENGTH - CRC_LENGTH;
  const cipheredMessage = payload.subarray(0, tagPos);
  const tag = payload.subarray(tagPos, payload.length - CRC_LENGTH);
  const decipheredMessage = decipherMessage(thpState.handshakeCredentials.trezorKey, thpState.recvNonce, cipheredMessage, tag);
  const messageType = decipheredMessage.readUInt16BE(0);
  const messagePayload = decipheredMessage.subarray(2);
  return protobufDecoder(messageType, messagePayload);
};
const decodeReadAck = () => ({
  type: 'ThpAck',
  message: {}
});
const decodeThpError = payload => {
  const [errorType] = payload;
  const error = (() => {
    switch (errorType) {
      case 0x01:
        return 'ThpTransportBusy';
      case 0x02:
        return 'ThpUnallocatedChannel';
      case 0x03:
        return 'ThpDecryptionFailed';
      case 0x05:
        return 'ThpDeviceLocked';
      default:
        return 'ThpUnknownError';
    }
  })();
  const message = {
    code: error,
    message: error ?? `Unknown ThpError ${errorType}`
  };
  return {
    type: 'ThpError',
    message
  };
};
export const getCRC = decodedMessage => decodedMessage.payload.subarray(decodedMessage.length - CRC_LENGTH, decodedMessage.length);
export const validateCrc = decodedMessage => {
  const payloadLen = decodedMessage.length - CRC_LENGTH;
  const length = Buffer.alloc(2);
  length.writeUInt16BE(decodedMessage.length);
  const expectedCrc = crc32(Buffer.concat([decodedMessage.header, length, decodedMessage.payload.subarray(0, payloadLen)]));
  const crc = getCRC(decodedMessage);
  if (expectedCrc.compare(crc) !== 0) {
    throw new Error(`Invalid CRC. expected: ${expectedCrc.toString('hex')} received: ${crc.toString('hex')}`);
  }
};
export const decode = (decodedMessage, protobufDecoder, thpState) => {
  if (!thpState) {
    throw new Error('ThpStateMissing');
  }
  const message = {
    ...decodedMessage,
    thpState
  };
  const {
    messageType
  } = decodedMessage;
  if (messageType === THP_CONTROL_BYTE.ERROR) {
    return decodeThpError(message.payload);
  }
  if (messageType === THP_CONTROL_BYTE.ACK_MESSAGE) {
    return decodeReadAck();
  }
  if (messageType === THP_CONTROL_BYTE.CHANNEL_ALLOCATION_RES) {
    return createChannelResponse(message, protobufDecoder);
  }
  if (messageType === THP_CONTROL_BYTE.HANDSHAKE_INIT_RES) {
    return readHandshakeInitResponse(message);
  }
  if (messageType === THP_CONTROL_BYTE.HANDSHAKE_COMP_RES) {
    return readHandshakeCompletionResponse(message);
  }
  if (messageType === THP_CONTROL_BYTE.ENCRYPTED) {
    return readProtobufMessage(message, protobufDecoder);
  }
  throw new Error('Unknown message type: ' + messageType);
};
