// @trezor/protocol/libESM/protocol-thp/encode.mjs
import { ThpState } from "./ThpState.mjs";
import { CRC_LENGTH, TAG_LENGTH, THP_DEFAULT_CHANNEL } from "./constants.mjs";
import { aesgcm, crc32 } from "./crypto/index.mjs";
import { THP_CONTROL_BYTE } from "../protocol-v2/constants.mjs";
import { getIvFromNonce } from "./crypto/tools.mjs";
import { addAckBit, addSequenceBit, isThpMessageName } from "./utils.mjs";
const cipherMessage = (key, sendNonce, handshakeHash, payload) => {
  const aes = aesgcm(key, getIvFromNonce(sendNonce));
  aes.auth(handshakeHash);
  const encryptedPayload = aes.encrypt(payload);
  const encryptedPayloadTag = aes.finish();
  return Buffer.concat([encryptedPayload, encryptedPayloadTag]);
};
const getBytesFromField = (data, fieldName) => {
  const value = data[fieldName];
  if (typeof value === 'string') {
    return Buffer.from(value, 'hex');
  }
  if (typeof value === 'number') {
    return Buffer.from([value]);
  }
  if (Buffer.isBuffer(value)) {
    return value;
  }
};
const createChannelRequestPayload = data => {
  const nonce = getBytesFromField(data, 'nonce');
  if (!nonce) {
    throw new Error('Missing nonce field');
  }
  return nonce;
};
const handshakeInitRequestPayload = (data, _thpState) => {
  const key = getBytesFromField(data, 'key');
  if (!key) {
    throw new Error('ThpHandshakeInitRequest missing key field');
  }
  const tryToUnlock = getBytesFromField(data, 'tryToUnlock') ?? Buffer.from([0]);
  return Buffer.concat([key, tryToUnlock]);
};
const handshakeCompletionRequestPayload = data => {
  const hostPubkey = getBytesFromField(data, 'hostPubkey');
  if (!hostPubkey) {
    throw new Error('ThpHandshakeCompletionRequest missing hostPubkey field');
  }
  const encryptedPayload = getBytesFromField(data, 'encryptedPayload');
  if (!encryptedPayload) {
    throw new Error('ThpHandshakeCompletionRequest missing encryptedPayload field');
  }
  return Buffer.concat([hostPubkey, encryptedPayload]);
};
export const encodePayload = (name, data, thpState) => {
  if (name === 'ThpCreateChannelRequest') {
    return createChannelRequestPayload(data);
  }
  if (name === 'ThpHandshakeInitRequest') {
    return handshakeInitRequestPayload(data, thpState);
  }
  if (name === 'ThpHandshakeCompletionRequest') {
    return handshakeCompletionRequestPayload(data);
  }
  return Buffer.alloc(0);
};
const createChannelRequest = (data, channel) => {
  const length = Buffer.alloc(2);
  length.writeUInt16BE(data.length + CRC_LENGTH);
  const magic = Buffer.from([THP_CONTROL_BYTE.CHANNEL_ALLOCATION_REQ]);
  const message = Buffer.concat([magic, channel, length, data]);
  const crc = crc32(message);
  return Buffer.concat([message, crc]);
};
const handshakeInitRequest = (data, channel, enablePiggybackAck) => {
  const length = Buffer.alloc(2);
  length.writeUInt16BE(data.length + CRC_LENGTH);
  const magic = addAckBit(THP_CONTROL_BYTE.HANDSHAKE_INIT_REQ, enablePiggybackAck ? 1 : 0);
  const message = Buffer.concat([magic, channel, length, data]);
  const crc = crc32(message);
  return Buffer.concat([message, crc]);
};
const handshakeCompletionRequest = (data, channel, sendBit) => {
  const length = Buffer.alloc(2);
  length.writeUInt16BE(data.length + CRC_LENGTH);
  const magic = addSequenceBit(THP_CONTROL_BYTE.HANDSHAKE_COMP_REQ, sendBit);
  const message = Buffer.concat([magic, channel, length, data]);
  const crc = crc32(message);
  return Buffer.concat([message, crc]);
};
const getPreviousAckBit = thpState => thpState.recvAckBit ? 0 : 1;
export const encodeAck = state => {
  const length = Buffer.alloc(2);
  length.writeUInt16BE(CRC_LENGTH);
  const magic = addAckBit(THP_CONTROL_BYTE.ACK_MESSAGE, state.recvAckBit);
  const message = Buffer.concat([magic, state.channel, length]);
  const crc = crc32(message);
  return Buffer.concat([message, crc]);
};
export const encodePreviousAck = thpState => {
  const prevState = new ThpState();
  prevState.deserialize({
    ...thpState.serialize(),
    recvAckBit: getPreviousAckBit(thpState)
  });
  return encodeAck(prevState);
};
const encodeThpMessage = (messageType, data, channel, thpState) => {
  if (messageType === 'ThpCreateChannelRequest') {
    return createChannelRequest(data, channel);
  }
  if (messageType === 'ThpHandshakeInitRequest') {
    return handshakeInitRequest(data, channel, thpState.isPiggybackAckAvailable);
  }
  if (messageType === 'ThpHandshakeCompletionRequest') {
    return handshakeCompletionRequest(data, channel, thpState.sendBit || 0);
  }
  if (messageType === 'ThpAck') {
    return thpState.isPiggybackAckEnabled ? encodePreviousAck(thpState) : encodeAck(thpState);
  }
  throw new Error(`Unknown Thp message type ${messageType}`);
};
export const encodeProtobufMessage = (messageType, data, channel, thpState) => {
  if (!thpState) {
    throw new Error('ThpStateMissing');
  }
  const length = Buffer.alloc(2);
  length.writeUInt16BE(1 + 2 + data.length + TAG_LENGTH + CRC_LENGTH);
  let magic = addSequenceBit(THP_CONTROL_BYTE.ENCRYPTED, thpState.sendBit);
  if (thpState.isPiggybackAckEnabled) {
    magic = addAckBit(magic.readUint8(), getPreviousAckBit(thpState));
  }
  const header = Buffer.concat([magic, channel]);
  const messageTypeBytes = Buffer.alloc(2);
  messageTypeBytes.writeUInt16BE(messageType);
  const cipheredMessage = cipherMessage(thpState.handshakeCredentials.hostKey, thpState.sendNonce, Buffer.alloc(0), Buffer.concat([thpState.sessionId, messageTypeBytes, data]));
  const message = Buffer.concat([header, length, cipheredMessage]);
  const crc = crc32(message);
  return Buffer.concat([message, crc]);
};
export const encode = options => {
  if (!options.thpState) {
    throw new Error('ThpStateMissing');
  }
  const channel = options.thpState.channel || THP_DEFAULT_CHANNEL;
  const {
    messageName,
    data,
    protobufEncoder,
    thpState
  } = options;
  let result;
  if (isThpMessageName(messageName)) {
    const payload = encodePayload(messageName, data, thpState);
    result = encodeThpMessage(messageName, payload, channel, options.thpState);
  } else {
    const {
      messageType,
      message
    } = protobufEncoder(messageName, data);
    result = encodeProtobufMessage(messageType, message, channel, thpState);
  }
  return result;
};
