// @trezor/protocol/libESM/protocol-thp/utils.mjs
import { ThpPairingMethod } from "./messages/index.mjs";
import { THP_CONTROL_BYTE } from "../protocol-v2/constants.mjs";
export const addAckBit = (magic, ackBit) => {
  const result = Buffer.alloc(1);
  result.writeInt8(magic | ackBit << 3);
  return result;
};
export const addSequenceBit = (magic, seqBit) => {
  const result = Buffer.alloc(1);
  result.writeInt8(magic | seqBit << 4);
  return result;
};
export const clearControlBit = magic => magic & ~(1 << 3) & ~(1 << 4);
export const getControlBit = magic => {
  const ackBit = (magic & 1 << 3) === 0 ? 0 : 1;
  const sequenceBit = (magic & 1 << 4) === 0 ? 0 : 1;
  return {
    ackBit,
    sequenceBit
  };
};
export const readThpHeader = bytes => {
  const magic = bytes.readUInt8();
  const {
    ackBit,
    sequenceBit
  } = getControlBit(magic);
  const channel = bytes.subarray(1, 3);
  return {
    magic: clearControlBit(magic),
    ackBit,
    sequenceBit,
    channel
  };
};
export const isAckExpected = bytesOrMagic => {
  const isCreateChannelMessage = magic => [THP_CONTROL_BYTE.CHANNEL_ALLOCATION_REQ, THP_CONTROL_BYTE.CHANNEL_ALLOCATION_RES].includes(magic);
  if (Array.isArray(bytesOrMagic)) {
    return !bytesOrMagic.find(n => isCreateChannelMessage(n));
  }
  return !isCreateChannelMessage(bytesOrMagic.readUInt8());
};
export const getExpectedResponses = bytes => {
  const {
    magic
  } = readThpHeader(bytes);
  if (magic === THP_CONTROL_BYTE.CHANNEL_ALLOCATION_REQ) {
    return [THP_CONTROL_BYTE.CHANNEL_ALLOCATION_RES];
  }
  if (magic === THP_CONTROL_BYTE.HANDSHAKE_INIT_REQ) {
    return [THP_CONTROL_BYTE.HANDSHAKE_INIT_RES];
  }
  if (magic === THP_CONTROL_BYTE.HANDSHAKE_COMP_REQ) {
    return [THP_CONTROL_BYTE.HANDSHAKE_COMP_RES];
  }
  if (magic === THP_CONTROL_BYTE.ENCRYPTED) {
    return [THP_CONTROL_BYTE.ENCRYPTED];
  }
  return [];
};
export const isThpMessageName = name => ['ThpCreateChannelRequest', 'ThpHandshakeInitRequest', 'ThpHandshakeCompletionRequest', 'ThpAck'].includes(name);
export const getThpPairingMethod = dm => typeof dm === 'string' ? ThpPairingMethod[dm] : dm;
