// @trezor/schema-utils/libESM/codegen.mjs
import * as Codegen from "@sinclair/typebox-codegen/typescript";
import fs from "fs";
import { pathToFileURL } from "url";
const preprocessCode = code => {
  code = code.replace(/typeof undefined/g, 'undefined');
  code = code.replace(/keyof typeof/g, 'keyof');
  return {
    code
  };
};
const runTypeBoxCodegen = code => {
  const customTypesMapping = {
    ArrayBuffer: 'Type.ArrayBuffer()',
    Buffer: 'Type.Buffer()',
    UintType: 'Type.Uint()',
    SintType: 'Type.Uint({ allowNegative: true })'
  };
  const customTypePlaceholder = Object.keys(customTypesMapping).map(t => `type ${t} = any;`);
  let output = Codegen.TypeScriptToTypeBox.Generate(customTypePlaceholder + code, {
    useTypeBoxImport: false,
    useIdentifiers: true
  });
  const lastKey = Object.keys(customTypesMapping)[Object.keys(customTypesMapping).length - 1];
  const index = output.lastIndexOf(`const ${lastKey} = `);
  const blankLineIndex = output.indexOf('\n\n', index);
  output = output.substring(blankLineIndex + 1);
  Object.entries(customTypesMapping).forEach(([key, value]) => {
    output = output.replace(new RegExp(`\\b${key}\\b`, 'g'), value);
  });
  return output;
};
const normalizeEnums = output => {
  const enums = [...output.matchAll(/enum Enum(\w+) {/g)].map(m => m[1]);
  enums.forEach(e => {
    output = output.replace(new RegExp(`\\b${e}\\b`, 'g'), `Enum${e}`);
    output = output.replace(new RegExp(`type Enum${e} = Static\\<typeof Enum${e}\\>`, 'g'), `type Enum${e} = Static<typeof Enum${e}>`);
    output = output.replace(new RegExp(`const Enum${e} = Type\\.Enum\\(Enum${e}\\)`, 'g'), `const Enum${e} = Type.Enum(${e})`);
    output = output.replace(new RegExp(`enum Enum${e} \\{`, 'g'), `enum ${e} {`);
    output = output.replace(new RegExp(`Type\\.KeyOf\\(Enum${e}(,?.*)\\)`, 'g'), `Type.KeyOfEnum(${e}$1)`);
  });
  return output;
};
export function generateTypeBox(rawCode) {
  const {
    code
  } = preprocessCode(rawCode);
  let output = runTypeBoxCodegen(code);
  output = normalizeEnums(output);
  const schemaUtilsImports = ['Type', 'Static'];
  if (/\bCloneType\b/.test(output)) {
    schemaUtilsImports.push('CloneType');
  }
  output = `import { ${schemaUtilsImports.join(', ')} } from '@trezor/schema-utils';\n\n${output}`;
  return output;
}
export function generate(code) {
  return generateTypeBox(code);
}
export function generateForFile(fileName) {
  const code = fs.readFileSync(fileName, 'utf-8');
  return generate(code);
}
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const fileName = process.argv[2];
  if (!fileName || !fs.existsSync(fileName)) {
    throw new Error('File not found');
  }
  const output = generateForFile(fileName);
  process.stdout.write(output);
}
