// @trezor/schema-utils/libESM/index.mjs
import { CloneType, JavaScriptTypeBuilder, Kind, Optional, OptionalKind } from "@sinclair/typebox";
import { Errors, ValueErrorType } from "@sinclair/typebox/errors";
import { Mixin } from "ts-mixer";
import { ArrayBufferBuilder, BufferBuilder, KeyofEnumBuilder, UintBuilder } from "./custom-types/index.mjs";
import { InvalidParameter } from "./errors.mjs";
import { setDeepValue } from "./utils.mjs";
class CustomTypeBuilder extends Mixin(JavaScriptTypeBuilder, ArrayBufferBuilder, BufferBuilder, KeyofEnumBuilder, UintBuilder) {}
export function Validate(schema, value) {
  try {
    Assert(schema, value);
    return true;
  } catch {
    return false;
  }
}
function FindErrorInUnion(error) {
  const currentValue = error.value;
  const unionMembers = error.schema.anyOf;
  const hasValidMember = unionMembers.find(unionSchema => Validate(unionSchema, currentValue));
  if (!hasValidMember) {
    const possibleMatchesByLiterals = unionMembers.filter(unionSchema => {
      if (unionSchema[Kind] !== 'Object') return false;
      return !Object.entries(unionSchema.properties).find(([property, propertySchema]) => propertySchema.const && propertySchema.const !== currentValue[property]);
    });
    if (possibleMatchesByLiterals.length === 1) {
      Assert(possibleMatchesByLiterals[0], currentValue);
    } else if (possibleMatchesByLiterals.length > 1) {
      const errorsOfPossibleMatches = possibleMatchesByLiterals.map(matchSchema => ({
        schema: matchSchema,
        errors: [...Errors(matchSchema, currentValue)]
      }));
      const sortedErrors = errorsOfPossibleMatches.sort((a, b) => a.errors.length - b.errors.length);
      const [bestMatch] = sortedErrors;
      Assert(bestMatch.schema, currentValue);
    }
    throw new InvalidParameter(error.message, error.path, error.type, error.value);
  }
}
export function Assert(schema, value) {
  const errors = [...Errors(schema, value)];
  let [error] = errors;
  while (error) {
    if (error.path === '/' && errors.length > 1) {} else if (error.value == null && error.schema[OptionalKind] === 'Optional') {} else if (error.type === ValueErrorType.Union) {
      FindErrorInUnion(error);
    } else if (error.type === ValueErrorType.Number && typeof error.value === 'string') {
      const currentValue = error.value;
      const parsedNumber = Number(currentValue);
      if (!Number.isNaN(parsedNumber) && currentValue === parsedNumber.toString()) {
        const pathParts = error.path.slice(1).split('/');
        setDeepValue(value, pathParts, parsedNumber);
      } else {
        throw new InvalidParameter(error.message, error.path, error.type, error.value);
      }
    } else {
      throw new InvalidParameter(error.message, error.path, error.type, error.value);
    }
    errors.shift();
    [error] = errors;
  }
}
export const Type = new CustomTypeBuilder();
export { Optional, CloneType };
